

<!doctype html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>alefuncs &#8212; alefuncs a documentation</title>
    
    <link rel="stylesheet" href="../_static/bizstyle.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     'a',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../_static/bizstyle.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <meta name="viewport" content="width=device-width,initial-scale=1.0">
    <!--[if lt IE 9]>
    <script type="text/javascript" src="_static/css3-mediaqueries.js"></script>
    <![endif]-->
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">alefuncs a documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="index.html" accesskey="U">Module code</a> &#187;</li> 
      </ul>
    </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for alefuncs</h1><div class="highlight"><pre>
<span></span><span class="c1">### author:  alessio.marcozzi@gmail.com</span>
<span class="c1">### version: 2017_02</span>
<span class="c1">### licence: MIT</span>

<div class="viewcode-block" id="print_sbar"><a class="viewcode-back" href="../code.html#alefuncs.print_sbar">[docs]</a><span class="k">def</span> <span class="nf">print_sbar</span><span class="p">(</span><span class="n">n</span><span class="p">,</span><span class="n">m</span><span class="p">,</span><span class="n">s</span><span class="o">=</span><span class="s1">&#39;|#.|&#39;</span><span class="p">,</span><span class="n">size</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span><span class="n">message</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;(int,int,string,int) =&gt; None</span>
<span class="sd">	Print a progress bar using the simbols in &#39;s&#39;.</span>
<span class="sd">	Example:</span>
<span class="sd">		range_limit = 1000</span>
<span class="sd">		for n in range(range_limit):</span>
<span class="sd">			print_sbar(n,m=range_limit)</span>
<span class="sd">			time.sleep(0.1)</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="kn">import</span> <span class="nn">sys</span>
	<span class="c1">#adjust to bar size</span>
	<span class="k">if</span> <span class="n">m</span> <span class="o">!=</span> <span class="n">size</span><span class="p">:</span>
		<span class="n">n</span> <span class="o">=</span><span class="p">(</span><span class="n">n</span><span class="o">*</span><span class="n">size</span><span class="p">)</span><span class="o">/</span><span class="n">m</span>
		<span class="n">m</span> <span class="o">=</span> <span class="n">size</span>
	<span class="c1">#calculate ticks</span>
	<span class="n">_a</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">n</span><span class="p">)</span><span class="o">*</span><span class="n">s</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">+</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">m</span><span class="p">)</span><span class="o">-</span><span class="nb">int</span><span class="p">(</span><span class="n">n</span><span class="p">))</span><span class="o">*</span><span class="n">s</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
	<span class="n">_b</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="n">n</span><span class="o">/</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">m</span><span class="p">))</span><span class="o">*</span><span class="mi">100</span><span class="p">,</span><span class="mi">1</span><span class="p">)</span>
	<span class="c1">#adjust overflow</span>
	<span class="k">if</span> <span class="n">_b</span> <span class="o">&gt;=</span> <span class="mi">100</span><span class="p">:</span>
		<span class="n">_b</span> <span class="o">=</span> <span class="mf">100.0</span>
	<span class="c1">#to stdout    </span>
	<span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\r</span><span class="si">{}{}{}{}</span><span class="s1"> </span><span class="si">{}</span><span class="s1">%     &#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">message</span><span class="p">,</span><span class="n">s</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="n">_a</span><span class="p">,</span><span class="n">s</span><span class="p">[</span><span class="mi">3</span><span class="p">],</span><span class="n">_b</span><span class="p">))</span>
	<span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span></div>


<div class="viewcode-block" id="hash"><a class="viewcode-back" href="../code.html#alefuncs.hash">[docs]</a><span class="k">def</span> <span class="nf">hash</span><span class="p">(</span><span class="n">a_string</span><span class="p">,</span><span class="n">algorithm</span><span class="o">=</span><span class="s1">&#39;md5&#39;</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;str =&gt; str</span>
<span class="sd">	Return the hash of a string calculated using various algorithms.</span>
<span class="sd">	Example:</span>
<span class="sd">		&gt;&gt;&gt; hash(&#39;prova&#39;,&#39;md5&#39;)</span>
<span class="sd">		&#39;189bbbb00c5f1fb7fba9ad9285f193d1&#39;</span>

<span class="sd">		&gt;&gt;&gt; hash(&#39;prova&#39;,&#39;sha256&#39;)</span>
<span class="sd">		&#39;6258a5e0eb772911d4f92be5b5db0e14511edbe01d1d0ddd1d5a2cb9db9a56ba&#39;</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="kn">import</span> <span class="nn">hashlib</span>
	<span class="k">if</span> <span class="n">algorithm</span> <span class="o">==</span> <span class="s1">&#39;md5&#39;</span><span class="p">:</span>
		<span class="k">return</span> <span class="n">hashlib</span><span class="o">.</span><span class="n">md5</span><span class="p">(</span><span class="n">a_string</span><span class="o">.</span><span class="n">encode</span><span class="p">())</span><span class="o">.</span><span class="n">hexdigest</span><span class="p">()</span>
	<span class="k">elif</span> <span class="n">algorithm</span> <span class="o">==</span> <span class="s1">&#39;sha256&#39;</span><span class="p">:</span>
		<span class="k">return</span> <span class="n">hashlib</span><span class="o">.</span><span class="n">sha256</span><span class="p">(</span><span class="n">a_string</span><span class="o">.</span><span class="n">encode</span><span class="p">())</span><span class="o">.</span><span class="n">hexdigest</span><span class="p">()</span>
	<span class="k">else</span><span class="p">:</span>
		<span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;algorithm </span><span class="si">{}</span><span class="s1"> not found&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">algorithm</span><span class="p">))</span></div>


<div class="viewcode-block" id="get_first_transcript_by_gene_name"><a class="viewcode-back" href="../code.html#alefuncs.get_first_transcript_by_gene_name">[docs]</a><span class="k">def</span> <span class="nf">get_first_transcript_by_gene_name</span><span class="p">(</span><span class="n">gene_name</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;str =&gt; str</span>
<span class="sd">	Return the id of the main trascript for a given gene.</span>
<span class="sd">	The data is from http://grch37.ensembl.org/</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="kn">from</span> <span class="nn">urllib</span> <span class="k">import</span> <span class="n">urlopen</span>
	<span class="kn">from</span> <span class="nn">pyensembl</span> <span class="k">import</span> <span class="n">EnsemblRelease</span>
	<span class="n">data</span> <span class="o">=</span> <span class="n">EnsemblRelease</span><span class="p">(</span><span class="mi">75</span><span class="p">)</span>
	<span class="n">gene</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">genes_by_name</span><span class="p">(</span><span class="n">gene_name</span><span class="p">)</span>
	<span class="n">gene_id</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">gene</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;=&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
	<span class="n">gene_location</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">gene</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;=&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s1">&#39;)&#39;</span><span class="p">)</span>
	<span class="n">url</span> <span class="o">=</span> <span class="s1">&#39;http://grch37.ensembl.org/Homo_sapiens/Gene/Summary?db=core;g=</span><span class="si">{}</span><span class="s1">;r=</span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">gene_id</span><span class="p">,</span><span class="n">gene_location</span><span class="p">)</span>
	<span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">urlopen</span><span class="p">(</span><span class="n">url</span><span class="p">):</span>
		<span class="k">if</span> <span class="s1">&#39;&lt;tbody&gt;&lt;tr&gt;&lt;td class=&quot;bold&quot;&gt;&#39;</span> <span class="ow">in</span> <span class="n">line</span><span class="p">:</span>
			<span class="k">return</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;&quot;&gt;&#39;</span><span class="p">)[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;&lt;/a&gt;&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span></div>

   
<span class="k">def</span> <span class="nf">get_exons_coord_by_gene_name</span><span class="p">(</span><span class="n">gene_name</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;str =&gt; OrderedDict({&#39;exon_id&#39;:[coordinates]})</span>
<span class="sd">	Return an OrderedDict having as k the exon_id</span>
<span class="sd">	and as value a tuple containing the genomic coordinates (&#39;chr&#39;,start,stop).</span>

<span class="sd">	Example:</span>
<span class="sd">		</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="kn">from</span> <span class="nn">pyensembl</span> <span class="k">import</span> <span class="n">EnsemblRelease</span>
	<span class="n">gene</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">genes_by_name</span><span class="p">(</span><span class="n">gene_name</span><span class="p">)</span>
	<span class="n">gene_id</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">gene</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;=&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
	<span class="n">gene_location</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">gene</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;=&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s1">&#39;)&#39;</span><span class="p">)</span>
	<span class="n">gene_transcript</span> <span class="o">=</span> <span class="n">get_first_transcript_by_gene_name</span><span class="p">(</span><span class="n">gene_name</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
	<span class="n">table</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
	<span class="k">for</span> <span class="n">exon_id</span> <span class="ow">in</span> <span class="n">data</span><span class="o">.</span><span class="n">exon_ids_of_gene_id</span><span class="p">(</span><span class="n">gene_id</span><span class="p">):</span>
		<span class="n">exon</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">exon_by_id</span><span class="p">(</span><span class="n">exon_id</span><span class="p">)</span>
		<span class="n">coordinates</span> <span class="o">=</span> <span class="p">(</span><span class="n">exon</span><span class="o">.</span><span class="n">contig</span><span class="p">,</span> <span class="n">exon</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">exon</span><span class="o">.</span><span class="n">end</span><span class="p">)</span>
		<span class="n">table</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="n">exon_id</span><span class="p">:</span><span class="n">coordinates</span><span class="p">})</span>
	<span class="k">return</span> <span class="n">table</span>


<div class="viewcode-block" id="get_exons_coord_by_gene_name"><a class="viewcode-back" href="../code.html#alefuncs.get_exons_coord_by_gene_name">[docs]</a><span class="k">def</span> <span class="nf">get_exons_coord_by_gene_name</span><span class="p">(</span><span class="n">gene_name</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	Example:</span>
<span class="sd">			table = get_exons_coord_by_gene_name(&#39;TP53&#39;)</span>
<span class="sd">			for k,v in table.iteritems():</span>
<span class="sd">				print k,v</span>

<span class="sd">		&gt;&gt;&gt; ENSE00002419584 [&#39;7,579,721&#39;, &#39;7,579,700&#39;]</span>
<span class="sd">			ENSE00003625790 [&#39;7,579,590&#39;, &#39;7,579,312&#39;]</span>
<span class="sd">			ENSE00003518480 [&#39;7,578,554&#39;, &#39;7,578,371&#39;]</span>
<span class="sd">			ENSE00003462942 [&#39;7,578,289&#39;, &#39;7,578,177&#39;]</span>
<span class="sd">			ENSE00003504863 [&#39;7,577,608&#39;, &#39;7,577,499&#39;]</span>
<span class="sd">			ENSE00003586720 [&#39;7,577,155&#39;, &#39;7,577,019&#39;]</span>
<span class="sd">			ENSE00003636029 [&#39;7,576,926&#39;, &#39;7,576,853&#39;]</span>
<span class="sd">			ENSE00003634848 [&#39;7,574,033&#39;, &#39;7,573,927&#39;]</span>
<span class="sd">			ENSE00002667911 [&#39;7,579,940&#39;, &#39;7,579,839&#39;]</span>
<span class="sd">			ENSE00002051192 [&#39;7,590,799&#39;, &#39;7,590,695&#39;]</span>
<span class="sd">			ENSE00002034209 [&#39;7,573,008&#39;, &#39;7,571,722&#39;]</span>
<span class="sd">			ENSE00003552110 [&#39;7,576,657&#39;, &#39;7,576,525&#39;]</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="kn">from</span> <span class="nn">collections</span> <span class="k">import</span> <span class="n">OrderedDict</span>
	<span class="kn">from</span> <span class="nn">pyensembl</span> <span class="k">import</span> <span class="n">EnsemblRelease</span>
	<span class="n">data</span> <span class="o">=</span> <span class="n">EnsemblRelease</span><span class="p">(</span><span class="mi">75</span><span class="p">)</span>
	<span class="n">gene</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">genes_by_name</span><span class="p">(</span><span class="n">gene_name</span><span class="p">)</span>
	<span class="n">gene_id</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">gene</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;=&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
	<span class="n">gene_location</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">gene</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;=&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s1">&#39;)&#39;</span><span class="p">)</span>
	<span class="n">gene_transcript</span> <span class="o">=</span> <span class="n">get_first_transcript_by_gene_name</span><span class="p">(</span><span class="n">gene_name</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
	<span class="n">url</span> <span class="o">=</span> <span class="s1">&#39;http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=</span><span class="si">{}</span><span class="s1">;r=</span><span class="si">{}</span><span class="s1">;t=</span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">gene_id</span><span class="p">,</span><span class="n">gene_location</span><span class="p">,</span><span class="n">gene_transcript</span><span class="p">)</span>
	<span class="n">str_html</span> <span class="o">=</span> <span class="n">get_html</span><span class="p">(</span><span class="n">url</span><span class="p">)</span>
	<span class="n">html</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
	<span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">str_html</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">):</span>
		<span class="k">try</span><span class="p">:</span>
			<span class="c1">#print line</span>
			<span class="n">html</span> <span class="o">+=</span> <span class="nb">str</span><span class="p">(</span><span class="n">line</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span>
		<span class="k">except</span> <span class="ne">UnicodeEncodeError</span><span class="p">:</span>
			<span class="k">pass</span>
	<span class="n">blocks</span> <span class="o">=</span> <span class="n">html</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
	<span class="n">table</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
	<span class="k">for</span> <span class="n">exon_id</span> <span class="ow">in</span> <span class="n">data</span><span class="o">.</span><span class="n">exon_ids_of_gene_id</span><span class="p">(</span><span class="n">gene_id</span><span class="p">):</span>
		<span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="n">txt</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">blocks</span><span class="p">):</span>
			<span class="k">if</span> <span class="n">exon_id</span> <span class="ow">in</span> <span class="n">txt</span><span class="p">:</span>
				<span class="k">if</span> <span class="n">exon_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">table</span><span class="p">:</span>
					<span class="n">table</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="n">exon_id</span><span class="p">:[]})</span>
				<span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">txt</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;&lt;td style=&quot;width:10%;text-align:left&quot;&gt;&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]:</span>
					<span class="n">table</span><span class="p">[</span><span class="n">exon_id</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">item</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;&lt;/td&gt;&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>
	<span class="k">return</span> <span class="n">table</span></div>


<div class="viewcode-block" id="split_overlap"><a class="viewcode-back" href="../code.html#alefuncs.split_overlap">[docs]</a><span class="k">def</span> <span class="nf">split_overlap</span><span class="p">(</span><span class="n">array</span><span class="p">,</span><span class="n">size</span><span class="p">,</span><span class="n">overlap</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;(list,int,int) =&gt; [[...],[...],...]</span>
<span class="sd">	Split a list into chunks of a specific size and overlap.</span>

<span class="sd">	Examples:</span>
<span class="sd">		array = list(range(10))</span>
<span class="sd">		split_overlap(array,4,2)</span>
<span class="sd">		&gt;&gt;&gt; [[0, 1, 2, 3], [2, 3, 4, 5], [4, 5, 6, 7], [6, 7, 8, 9]]</span>

<span class="sd">		array = list(range(11))</span>
<span class="sd">		split_overlap(array,4,2)</span>
<span class="sd">		&gt;&gt;&gt; [[0, 1, 2, 3], [2, 3, 4, 5], [4, 5, 6, 7], [6, 7, 8, 9], [8, 9, 10]]</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="n">result</span> <span class="o">=</span> <span class="p">[]</span>
	<span class="k">while</span> <span class="kc">True</span><span class="p">:</span>
		<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">array</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="n">size</span><span class="p">:</span>
			<span class="n">result</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">array</span><span class="p">)</span>
			<span class="k">return</span> <span class="n">result</span>
		<span class="k">else</span><span class="p">:</span>
			<span class="n">result</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">array</span><span class="p">[:</span><span class="n">size</span><span class="p">])</span>
			<span class="n">array</span> <span class="o">=</span> <span class="n">array</span><span class="p">[</span><span class="n">size</span><span class="o">-</span><span class="n">overlap</span><span class="p">:]</span></div>


<div class="viewcode-block" id="reorder_dict"><a class="viewcode-back" href="../code.html#alefuncs.reorder_dict">[docs]</a><span class="k">def</span> <span class="nf">reorder_dict</span><span class="p">(</span><span class="n">d</span><span class="p">,</span><span class="n">keys</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;(dict,list) =&gt; OrderedDict</span>
<span class="sd">	Change the order of a dictionary&#39;s keys</span>
<span class="sd">	without copying the dictionary (save RAM!).</span>
<span class="sd">	Return an OrderedDict.</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="n">tmp</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
	<span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">keys</span><span class="p">:</span>
		<span class="n">tmp</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="n">d</span><span class="p">[</span><span class="n">k</span><span class="p">]</span>
		<span class="k">del</span> <span class="n">d</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="c1">#this saves RAM</span>
	<span class="k">return</span> <span class="n">tmp</span></div>

<span class="c1">#test = OrderedDict({&#39;1&#39;:1,&#39;2&#39;:2,&#39;4&#39;:4,&#39;3&#39;:3})</span>
<span class="c1">#print(test)</span>
<span class="c1">#test2 = reorder_dict(test,[&#39;1&#39;,&#39;2&#39;,&#39;3&#39;,&#39;4&#39;])</span>
<span class="c1">#print(test)</span>
<span class="c1">#print(test2)</span>
<span class="c1">#&gt;&gt;&gt; OrderedDict([(&#39;2&#39;, 2), (&#39;3&#39;, 3), (&#39;4&#39;, 4), (&#39;1&#39;, 1)])</span>
<span class="c1">#&gt;&gt;&gt; OrderedDict()</span>
<span class="c1">#&gt;&gt;&gt; OrderedDict([(&#39;1&#39;, 1), (&#39;2&#39;, 2), (&#39;3&#39;, 3), (&#39;4&#39;, 4)])</span>


<div class="viewcode-block" id="in_between"><a class="viewcode-back" href="../code.html#alefuncs.in_between">[docs]</a><span class="k">def</span> <span class="nf">in_between</span><span class="p">(</span><span class="n">one_number</span><span class="p">,</span><span class="n">two_numbers</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;(int,list) =&gt; bool</span>
<span class="sd">	Return true if a number is in between two other numbers.</span>
<span class="sd">	Return False otherwise.</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="k">if</span> <span class="n">two_numbers</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">two_numbers</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span>
		<span class="k">pass</span>
	<span class="k">else</span><span class="p">:</span>
		<span class="n">two_numbers</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">two_numbers</span><span class="p">)</span>
	<span class="k">return</span> <span class="n">two_numbers</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">one_number</span> <span class="o">&lt;=</span> <span class="n">two_numbers</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span></div>


<div class="viewcode-block" id="is_overlapping"><a class="viewcode-back" href="../code.html#alefuncs.is_overlapping">[docs]</a><span class="k">def</span> <span class="nf">is_overlapping</span><span class="p">(</span><span class="n">svA</span><span class="p">,</span><span class="n">svB</span><span class="p">,</span><span class="n">limit</span><span class="o">=</span><span class="mf">0.9</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;(list,list,float) =&gt; bool</span>
<span class="sd">	Check if two SV ovelaps for at least 90% (limit=0.9).</span>
<span class="sd">	svX = [chr1,brk1,chr2,brk2]</span>
<span class="sd">	&#39;&#39;&#39;</span>
	
	<span class="c1"># Step 1.</span>
	<span class="c1"># Select the breaks in order to have lower coordinates first</span>
	<span class="k">if</span> <span class="nb">int</span><span class="p">(</span><span class="n">svA</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="o">&lt;=</span> <span class="nb">int</span><span class="p">(</span><span class="n">svA</span><span class="p">[</span><span class="mi">3</span><span class="p">]):</span>
		<span class="n">chr1_A</span> <span class="o">=</span> <span class="n">svA</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
		<span class="n">brk1_A</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">svA</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
		<span class="n">chr2_A</span> <span class="o">=</span> <span class="n">svA</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
		<span class="n">brk2_A</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">svA</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
	<span class="k">else</span><span class="p">:</span>
		<span class="n">chr2_A</span> <span class="o">=</span> <span class="n">svA</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
		<span class="n">brk2_A</span> <span class="o">=</span> <span class="n">svA</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
		<span class="n">chr1_A</span> <span class="o">=</span> <span class="n">svA</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
		<span class="n">brk1_A</span> <span class="o">=</span> <span class="n">svA</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>        
	
	<span class="k">if</span> <span class="nb">int</span><span class="p">(</span><span class="n">svB</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="o">&lt;=</span> <span class="nb">int</span><span class="p">(</span><span class="n">svB</span><span class="p">[</span><span class="mi">3</span><span class="p">]):</span>
		<span class="n">chr1_B</span> <span class="o">=</span> <span class="n">svB</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
		<span class="n">brk1_B</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">svB</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
		<span class="n">chr2_B</span> <span class="o">=</span> <span class="n">svB</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
		<span class="n">brk2_B</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">svB</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
	<span class="k">else</span><span class="p">:</span>
		<span class="n">chr2_B</span> <span class="o">=</span> <span class="n">svB</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
		<span class="n">brk2_B</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">svB</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
		<span class="n">chr1_B</span> <span class="o">=</span> <span class="n">svB</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
		<span class="n">brk1_B</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">svB</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
		
	<span class="c1"># Step 2.</span>
	<span class="c1"># Determine who is the longest</span>
	<span class="c1"># Return False immediately if the chromosomes are not the same.</span>
	<span class="c1"># This computation is reasonable only for sv on the same chormosome.</span>
	<span class="k">if</span> <span class="n">chr1_A</span> <span class="o">==</span> <span class="n">chr2_A</span> <span class="ow">and</span> <span class="n">chr1_B</span> <span class="o">==</span> <span class="n">chr2_B</span> <span class="ow">and</span> <span class="n">chr1_A</span> <span class="o">==</span> <span class="n">chr1_B</span><span class="p">:</span>       
		<span class="n">len_A</span> <span class="o">=</span> <span class="n">brk2_A</span> <span class="o">-</span> <span class="n">brk1_A</span>        
		<span class="n">len_B</span> <span class="o">=</span> <span class="n">brk2_B</span> <span class="o">-</span> <span class="n">brk1_B</span>
		
		<span class="k">if</span> <span class="n">len_A</span> <span class="o">&gt;=</span> <span class="n">len_B</span><span class="p">:</span>
			<span class="n">len_reference</span> <span class="o">=</span> <span class="n">len_A</span>
			<span class="n">len_sample</span> <span class="o">=</span> <span class="n">len_B</span>
		<span class="k">else</span><span class="p">:</span>
			<span class="n">len_reference</span> <span class="o">=</span> <span class="n">len_B</span>
			<span class="n">len_sample</span> <span class="o">=</span> <span class="n">len_A</span>
		
		<span class="n">limit</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="n">len_reference</span> <span class="o">*</span> <span class="n">limit</span><span class="p">)</span> <span class="c1"># this is the minimum overlap the two sv need to share</span>
									 <span class="c1"># to be considered overlapping</span>
			
		<span class="c1"># if the sample is smaller then the limit then there is no need to go further.</span>
		<span class="c1"># the sample segment will never share enough similarity with the reference.</span>
		<span class="k">if</span> <span class="n">len_sample</span> <span class="o">&lt;</span> <span class="n">limit</span><span class="p">:</span>
			<span class="k">return</span> <span class="kc">False</span>
	<span class="k">else</span><span class="p">:</span>
		<span class="k">return</span> <span class="kc">False</span>
	
	<span class="c1"># Step 3.</span>
	<span class="c1"># Determine if there is an overlap</span>
	<span class="c1"># &gt;&gt; There is an overlap if a least one of the break of an sv is in beetween the two breals of the other sv.</span>
	<span class="n">overlapping</span> <span class="o">=</span> <span class="kc">False</span>
	<span class="k">for</span> <span class="n">b</span> <span class="ow">in</span> <span class="p">[</span><span class="n">brk1_A</span><span class="p">,</span><span class="n">brk2_A</span><span class="p">]:</span>
		<span class="k">if</span> <span class="n">in_between</span><span class="p">(</span><span class="n">b</span><span class="p">,[</span><span class="n">brk1_B</span><span class="p">,</span><span class="n">brk2_B</span><span class="p">]):</span>
			<span class="n">overlapping</span> <span class="o">=</span> <span class="kc">True</span>
	<span class="k">for</span> <span class="n">b</span> <span class="ow">in</span> <span class="p">[</span><span class="n">brk1_B</span><span class="p">,</span><span class="n">brk2_B</span><span class="p">]:</span>
		<span class="k">if</span> <span class="n">in_between</span><span class="p">(</span><span class="n">b</span><span class="p">,[</span><span class="n">brk1_A</span><span class="p">,</span><span class="n">brk2_A</span><span class="p">]):</span>
			<span class="n">overlapping</span> <span class="o">=</span> <span class="kc">True</span>
			
	<span class="k">if</span> <span class="ow">not</span> <span class="n">overlapping</span><span class="p">:</span>
		<span class="k">return</span> <span class="kc">False</span>
	
	<span class="c1"># Step 4.</span>
	<span class="c1"># Determine the lenght of the ovelapping part</span>
	
	<span class="c1"># easy case: if the points are all different then, if I sort the points,</span>
	<span class="c1"># the overlap is the region between points[1] and points[2]</span>
	
	<span class="c1"># |-----------------|             |---------------------|</span>
	<span class="c1">#         |--------------|              |-------------|</span>
	<span class="n">points</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">([</span><span class="n">brk1_A</span><span class="p">,</span><span class="n">brk2_A</span><span class="p">,</span><span class="n">brk1_B</span><span class="p">,</span><span class="n">brk2_B</span><span class="p">])</span>
	<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">points</span><span class="p">))</span> <span class="o">==</span> <span class="mi">4</span><span class="p">:</span> <span class="c1"># the points are all different</span>
		<span class="n">overlap</span> <span class="o">=</span> <span class="n">points</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">-</span><span class="n">points</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
		
	<span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">points</span><span class="p">))</span> <span class="o">==</span> <span class="mi">3</span><span class="p">:</span> <span class="c1">#one point is in common</span>
		<span class="c1"># |-----------------|</span>
		<span class="c1"># |--------------|</span>
		<span class="k">if</span> <span class="n">points</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="n">points</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span>
			<span class="n">overlap</span> <span class="o">=</span> <span class="n">points</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">-</span><span class="n">points</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
			
		<span class="c1"># |---------------------|</span>
		<span class="c1">#         |-------------|</span>
		<span class="k">if</span> <span class="n">points</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">==</span> <span class="n">points</span><span class="p">[</span><span class="mi">3</span><span class="p">]:</span>
			<span class="n">overlap</span> <span class="o">=</span> <span class="n">points</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">-</span><span class="n">points</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
			
		<span class="c1"># |-----------------|</span>
		<span class="c1">#                   |-------------|</span>
		<span class="k">if</span> <span class="n">points</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="n">points</span><span class="p">[</span><span class="mi">2</span><span class="p">]:</span>
			<span class="k">return</span> <span class="kc">False</span> <span class="c1"># there is no overlap</span>
	<span class="k">else</span><span class="p">:</span>
		<span class="c1"># |-----------------|</span>
		<span class="c1"># |-----------------|</span>
		<span class="k">return</span> <span class="kc">True</span> <span class="c1"># if two points are in common, then it is the very same sv</span>
	
	<span class="k">if</span> <span class="n">overlap</span> <span class="o">&gt;=</span> <span class="n">limit</span><span class="p">:</span>
		<span class="k">return</span> <span class="kc">True</span>
	<span class="k">else</span><span class="p">:</span>
		<span class="k">return</span> <span class="kc">False</span></div>


<div class="viewcode-block" id="load_obj"><a class="viewcode-back" href="../code.html#alefuncs.load_obj">[docs]</a><span class="k">def</span> <span class="nf">load_obj</span><span class="p">(</span><span class="n">file</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	Load a pickled object.</span>
<span class="sd">	Be aware that pickle is version dependent,</span>
<span class="sd">	i.e. objects dumped in Py3 cannot be loaded with Py2.</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="kn">import</span> <span class="nn">pickle</span>
	<span class="k">try</span><span class="p">:</span>
		<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">file</span><span class="p">,</span><span class="s1">&#39;rb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
			<span class="n">obj</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
		<span class="k">return</span> <span class="n">obj</span>
	<span class="k">except</span><span class="p">:</span>
		<span class="k">return</span> <span class="kc">False</span></div>


<div class="viewcode-block" id="save_obj"><a class="viewcode-back" href="../code.html#alefuncs.save_obj">[docs]</a><span class="k">def</span> <span class="nf">save_obj</span><span class="p">(</span><span class="n">obj</span><span class="p">,</span> <span class="n">file</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	Dump an object with pickle.</span>
<span class="sd">	Be aware that pickle is version dependent,</span>
<span class="sd">	i.e. objects dumped in Py3 cannot be loaded with Py2.</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="kn">import</span> <span class="nn">pickle</span>
	<span class="k">try</span><span class="p">:</span>
		<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">file</span><span class="p">,</span><span class="s1">&#39;wb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
			<span class="n">pickle</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">obj</span><span class="p">,</span> <span class="n">f</span><span class="p">)</span>
		<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Object saved to </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">file</span><span class="p">))</span>
		<span class="k">return</span> <span class="kc">True</span>
	<span class="k">except</span><span class="p">:</span>
		<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Error: Object not saved...&#39;</span><span class="p">)</span>
		<span class="k">return</span> <span class="kc">False</span></div>
	
<span class="c1">#save_obj(hotspots_review,&#39;hotspots_review_CIS.txt&#39;)</span>


<span class="k">def</span> <span class="nf">query_encode</span><span class="p">(</span><span class="n">chromosome</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	Queries ENCODE via http://promoter.bx.psu.edu/ENCODE/search_human.php</span>
<span class="sd">	Parses the output and returns a dictionary of CIS elements found and the relative location.</span>
<span class="sd">	&#39;&#39;&#39;</span>
	
	<span class="c1">## Regex setup</span>
	<span class="n">re1</span><span class="o">=</span><span class="s1">&#39;(chr</span><span class="si">{}</span><span class="s1">)&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chromosome</span><span class="p">)</span> <span class="c1"># The specific chromosome</span>
	<span class="n">re2</span><span class="o">=</span><span class="s1">&#39;(:)&#39;</span>    <span class="c1"># Any Single Character &#39;:&#39;</span>
	<span class="n">re3</span><span class="o">=</span><span class="s1">&#39;(</span><span class="se">\\</span><span class="s1">d+)&#39;</span> <span class="c1"># Integer</span>
	<span class="n">re4</span><span class="o">=</span><span class="s1">&#39;(-)&#39;</span>    <span class="c1"># Any Single Character &#39;-&#39;</span>
	<span class="n">re5</span><span class="o">=</span><span class="s1">&#39;(</span><span class="se">\\</span><span class="s1">d+)&#39;</span> <span class="c1"># Integer</span>
	<span class="n">rg</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="n">re1</span><span class="o">+</span><span class="n">re2</span><span class="o">+</span><span class="n">re3</span><span class="o">+</span><span class="n">re4</span><span class="o">+</span><span class="n">re5</span><span class="p">,</span><span class="n">re</span><span class="o">.</span><span class="n">IGNORECASE</span><span class="o">|</span><span class="n">re</span><span class="o">.</span><span class="n">DOTALL</span><span class="p">)</span>

	<span class="c1">## Query ENCODE</span>
	<span class="n">std_link</span> <span class="o">=</span> <span class="s1">&#39;http://promoter.bx.psu.edu/ENCODE/get_human_cis_region.php?assembly=hg19&amp;&#39;</span>
	<span class="n">query</span> <span class="o">=</span> <span class="n">std_link</span> <span class="o">+</span> <span class="s1">&#39;chr=chr</span><span class="si">{}</span><span class="s1">&amp;start=</span><span class="si">{}</span><span class="s1">&amp;end=</span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chromosome</span><span class="p">,</span><span class="n">start</span><span class="p">,</span><span class="n">end</span><span class="p">)</span>
	<span class="nb">print</span><span class="p">(</span><span class="n">query</span><span class="p">)</span>
	<span class="n">html_doc</span> <span class="o">=</span> <span class="n">urlopen</span><span class="p">(</span><span class="n">query</span><span class="p">)</span>
	<span class="n">html_txt</span> <span class="o">=</span> <span class="n">BeautifulSoup</span><span class="p">(</span><span class="n">html_doc</span><span class="p">,</span> <span class="s1">&#39;html.parser&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">get_text</span><span class="p">()</span>
	<span class="n">data</span> <span class="o">=</span> <span class="n">html_txt</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>

	<span class="c1">## Parse the output</span>
	<span class="n">parsed</span> <span class="o">=</span> <span class="p">{}</span>
	<span class="n">coordinates</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">item_</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">data</span><span class="p">)</span> <span class="k">if</span> <span class="n">item_</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="o">==</span> <span class="s1">&#39;Coordinate&#39;</span><span class="p">]</span>
	<span class="n">elements</span> <span class="o">=</span> <span class="p">[</span><span class="n">data</span><span class="p">[</span><span class="n">i</span><span class="o">-</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;  &#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;: &#39;</span><span class="p">,</span><span class="s1">&#39;&#39;</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">coordinates</span><span class="p">]</span>
	<span class="n">blocks</span> <span class="o">=</span> <span class="p">[</span><span class="n">item</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">data</span> <span class="k">if</span> <span class="n">item</span><span class="p">[:</span><span class="mi">3</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;chr&#39;</span><span class="p">]</span>
	<span class="nb">print</span><span class="p">(</span><span class="n">elements</span><span class="p">)</span>
	
	<span class="k">try</span><span class="p">:</span>
		<span class="n">i</span> <span class="o">=</span> <span class="mi">0</span>
		<span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">elements</span><span class="p">:</span>
			<span class="c1">#print(i)</span>
			<span class="k">try</span><span class="p">:</span>
				<span class="n">txt</span> <span class="o">=</span> <span class="n">blocks</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
				<span class="c1">#print(txt)</span>
				<span class="n">m</span> <span class="o">=</span> <span class="n">rg</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="n">txt</span><span class="p">)</span>
				<span class="n">bins</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">item</span><span class="p">)</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">m</span><span class="p">]</span>
				<span class="n">parsed</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="n">item</span><span class="p">:</span><span class="n">bins</span><span class="p">})</span>
				<span class="n">i</span> <span class="o">+=</span> <span class="mi">1</span>
				<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;found </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">item</span><span class="p">))</span>
			<span class="k">except</span><span class="p">:</span>
				<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;the field </span><span class="si">{}</span><span class="s1"> was empty&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">item</span><span class="p">))</span>
		<span class="k">return</span> <span class="n">parsed</span>
	<span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
		<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;ENCODE query falied on chr</span><span class="si">{}</span><span class="s1">:</span><span class="si">{}</span><span class="s1">-</span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chromosome</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">))</span>
		<span class="nb">print</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>
		<span class="k">return</span> <span class="kc">False</span>


<div class="viewcode-block" id="compare"><a class="viewcode-back" href="../code.html#alefuncs.compare">[docs]</a><span class="k">def</span> <span class="nf">compare</span><span class="p">(</span><span class="n">dict_A</span><span class="p">,</span><span class="n">dict_B</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;(dict,dict) =&gt; dict, dict, dict</span>
<span class="sd">	Compares two dicts of bins.</span>
<span class="sd">	Returns the shared elements, the unique elements of A and the unique elements of B.</span>
<span class="sd">	The dicts shape is supposed to be like this:</span>
<span class="sd">		OrderedDict([(&#39;1&#39;,</span>
<span class="sd">					  [&#39;23280000-23290000&#39;,</span>
<span class="sd">					   &#39;24390000-24400000&#39;,</span>
<span class="sd">					   ...]),</span>
<span class="sd">					 (&#39;2&#39;,</span>
<span class="sd">					  [&#39;15970000-15980000&#39;,</span>
<span class="sd">					   &#39;16020000-16030000&#39;,</span>
<span class="sd">					   ...]),</span>
<span class="sd">					 (&#39;3&#39;,</span>
<span class="sd">					  [&#39;610000-620000&#39;,</span>
<span class="sd">					   &#39;3250000-3260000&#39;,</span>
<span class="sd">					   &#39;6850000-6860000&#39;,</span>
<span class="sd">					   ...])}</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="n">chrms</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">23</span><span class="p">)]</span> <span class="o">+</span> <span class="p">[</span><span class="s1">&#39;X&#39;</span><span class="p">,</span><span class="s1">&#39;Y&#39;</span><span class="p">]</span>
	
	<span class="n">shared</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
	<span class="n">unique_A</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
	<span class="n">unique_B</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
	<span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">chrms</span><span class="p">:</span>
		<span class="n">shared</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="n">k</span><span class="p">:[]})</span>
		<span class="n">unique_A</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="n">k</span><span class="p">:[]})</span>
		<span class="n">unique_B</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="n">k</span><span class="p">:[]})</span>

		<span class="k">if</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">dict_A</span> <span class="ow">and</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">dict_B</span><span class="p">:</span>
			<span class="k">for</span> <span class="n">bin_</span> <span class="ow">in</span> <span class="n">dict_A</span><span class="p">[</span><span class="n">k</span><span class="p">]:</span>
				<span class="k">if</span> <span class="n">bin_</span> <span class="ow">in</span> <span class="n">dict_B</span><span class="p">[</span><span class="n">k</span><span class="p">]:</span>
					<span class="n">shared</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">bin_</span><span class="p">)</span>
				<span class="k">else</span><span class="p">:</span>
					<span class="n">unique_A</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">bin_</span><span class="p">)</span>
			<span class="k">for</span> <span class="n">bin_</span> <span class="ow">in</span> <span class="n">dict_B</span><span class="p">[</span><span class="n">k</span><span class="p">]:</span>
				<span class="k">if</span> <span class="n">bin_</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">shared</span><span class="p">[</span><span class="n">k</span><span class="p">]:</span>
					<span class="n">unique_B</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">bin_</span><span class="p">)</span>
		<span class="k">elif</span> <span class="n">k</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">dict_A</span><span class="p">:</span>
			<span class="n">unique_B</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">bin_</span> <span class="k">for</span> <span class="n">bin_</span> <span class="ow">in</span> <span class="n">dict_B</span><span class="p">[</span><span class="n">k</span><span class="p">]]</span>
		
		<span class="k">elif</span> <span class="n">k</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">dict_B</span><span class="p">:</span>
			<span class="n">unique_A</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">bin_</span> <span class="k">for</span> <span class="n">bin_</span> <span class="ow">in</span> <span class="n">dict_A</span><span class="p">[</span><span class="n">k</span><span class="p">]]</span>
			
	<span class="k">return</span> <span class="n">shared</span><span class="p">,</span> <span class="n">unique_A</span><span class="p">,</span> <span class="n">unique_B</span></div>


<span class="c1">#To manage heavy files</span>
<span class="k">def</span> <span class="nf">yield_file</span><span class="p">(</span><span class="n">infile</span><span class="p">):</span>
	<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">infile</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
		<span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">f</span><span class="p">:</span>
			<span class="k">if</span> <span class="n">line</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;#&#39;</span><span class="p">,</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">,</span><span class="s1">&#39; &#39;</span><span class="p">,</span><span class="s1">&#39;&#39;</span><span class="p">]:</span>
				<span class="k">yield</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>


<span class="c1">#Downaload sequence from ensembl</span>
<span class="k">def</span> <span class="nf">sequence_from_coordinates</span><span class="p">(</span><span class="n">chromosome</span><span class="p">,</span><span class="n">strand</span><span class="p">,</span><span class="n">start</span><span class="p">,</span><span class="n">end</span><span class="p">,</span><span class="n">account</span><span class="o">=</span><span class="s2">&quot;a.marcozzi@umcutrecht.nl&quot;</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	Download the nucleotide sequence from the gene_name.</span>
<span class="sd">	This function works only with with GRCh37.</span>
<span class="sd">	Inputs can be str or int e.g. 1 or &#39;1&#39; </span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="kn">from</span> <span class="nn">Bio</span> <span class="k">import</span> <span class="n">Entrez</span><span class="p">,</span> <span class="n">SeqIO</span>

	<span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="n">account</span> <span class="c1"># Always tell NCBI who you are</span>

	<span class="c1">#GRCh37 from http://www.ncbi.nlm.nih.gov/assembly/GCF_000001405.25/#/def_asm_Primary_Assembly</span>
	<span class="n">NCBI_IDS</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;1&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000001.10&#39;</span><span class="p">,</span><span class="s1">&#39;2&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000002.11&#39;</span><span class="p">,</span><span class="s1">&#39;3&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000003.11&#39;</span><span class="p">,</span><span class="s1">&#39;4&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000004.11&#39;</span><span class="p">,</span>
				<span class="s1">&#39;5&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000005.9&#39;</span><span class="p">,</span><span class="s1">&#39;6&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000006.11&#39;</span><span class="p">,</span><span class="s1">&#39;7&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000007.13&#39;</span><span class="p">,</span><span class="s1">&#39;8&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000008.10&#39;</span><span class="p">,</span>
				<span class="s1">&#39;9&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000009.11&#39;</span><span class="p">,</span><span class="s1">&#39;10&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000010.10&#39;</span><span class="p">,</span><span class="s1">&#39;11&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000011.9&#39;</span><span class="p">,</span><span class="s1">&#39;12&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000012.11&#39;</span><span class="p">,</span>
				<span class="s1">&#39;13&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000013.10&#39;</span><span class="p">,</span><span class="s1">&#39;14&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000014.8&#39;</span><span class="p">,</span><span class="s1">&#39;15&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000015.9&#39;</span><span class="p">,</span><span class="s1">&#39;16&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000016.9&#39;</span><span class="p">,</span>
				<span class="s1">&#39;17&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000017.10&#39;</span><span class="p">,</span><span class="s1">&#39;18&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000018.9&#39;</span><span class="p">,</span><span class="s1">&#39;19&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000019.9&#39;</span><span class="p">,</span><span class="s1">&#39;20&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000020.10&#39;</span><span class="p">,</span>
				<span class="s1">&#39;21&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000021.8&#39;</span><span class="p">,</span><span class="s1">&#39;22&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000022.10&#39;</span><span class="p">,</span><span class="s1">&#39;X&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000023.10&#39;</span><span class="p">,</span><span class="s1">&#39;Y&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000024.9&#39;</span><span class="p">}</span>       
  
	<span class="k">try</span><span class="p">:</span>        
		<span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">efetch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s2">&quot;nucleotide&quot;</span><span class="p">,</span> 
							   <span class="nb">id</span><span class="o">=</span><span class="n">NCBI_IDS</span><span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">chromosome</span><span class="p">)],</span> 
							   <span class="n">rettype</span><span class="o">=</span><span class="s2">&quot;fasta&quot;</span><span class="p">,</span> 
							   <span class="n">strand</span><span class="o">=</span><span class="n">strand</span><span class="p">,</span> <span class="c1">#&quot;1&quot; for the plus strand and &quot;2&quot; for the minus strand.</span>
							   <span class="n">seq_start</span><span class="o">=</span><span class="n">start</span><span class="p">,</span>
							   <span class="n">seq_stop</span><span class="o">=</span><span class="n">end</span><span class="p">)</span>
		<span class="n">record</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
		<span class="n">handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
		<span class="n">sequence</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
		<span class="k">return</span> <span class="n">sequence</span>
	<span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
		<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;ValueError: no sequence found in NCBI&#39;</span><span class="p">)</span>
		<span class="k">return</span> <span class="kc">False</span>


<span class="c1">#GC content calculator    </span>
<div class="viewcode-block" id="gc_content"><a class="viewcode-back" href="../code.html#alefuncs.gc_content">[docs]</a><span class="k">def</span> <span class="nf">gc_content</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span><span class="n">percent</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	Return the GC content of a sequence.</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="n">sequence</span> <span class="o">=</span> <span class="n">sequence</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
	<span class="n">g</span> <span class="o">=</span> <span class="n">sequence</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;G&quot;</span><span class="p">)</span>
	<span class="n">c</span> <span class="o">=</span> <span class="n">sequence</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;C&quot;</span><span class="p">)</span>
	<span class="n">t</span> <span class="o">=</span> <span class="n">sequence</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;T&quot;</span><span class="p">)</span>
	<span class="n">a</span> <span class="o">=</span> <span class="n">sequence</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;A&quot;</span><span class="p">)</span>
	<span class="n">gc_count</span> <span class="o">=</span> <span class="n">g</span><span class="o">+</span><span class="n">c</span>
	<span class="n">total_bases_count</span> <span class="o">=</span> <span class="n">g</span><span class="o">+</span><span class="n">c</span><span class="o">+</span><span class="n">t</span><span class="o">+</span><span class="n">a</span>
	<span class="k">if</span> <span class="n">total_bases_count</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
		<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Error in gc_content(sequence): sequence may contain only Ns&#39;</span><span class="p">)</span>
		<span class="k">return</span> <span class="kc">None</span>
	
	<span class="k">try</span><span class="p">:</span>
		<span class="n">gc_fraction</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">gc_count</span><span class="p">)</span> <span class="o">/</span> <span class="n">total_bases_count</span>
	<span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
		<span class="nb">print</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>
		<span class="nb">print</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
	
	<span class="k">if</span> <span class="n">percent</span><span class="p">:</span>
		<span class="k">return</span> <span class="n">gc_fraction</span> <span class="o">*</span> <span class="mi">100</span>
	<span class="k">else</span><span class="p">:</span>
		<span class="k">return</span> <span class="n">gc_fraction</span></div>
	
	
	   
<span class="c1">##Flexibility calculator##  </span>
<span class="c1">#requires stabflex3.py</span>




<span class="c1">#Endpoint function to calculate the flexibility of a given sequence</span>
<div class="viewcode-block" id="dna_flex"><a class="viewcode-back" href="../code.html#alefuncs.dna_flex">[docs]</a><span class="k">def</span> <span class="nf">dna_flex</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span><span class="n">window_size</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span><span class="n">step_zize</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span><span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;(str,int,int,bool) =&gt; list_of_tuples</span>
<span class="sd">	Calculate the flexibility index of a sequence.</span>
<span class="sd">	Return a list of tuples.</span>
<span class="sd">	Each tuple contains the bin&#39;s coordinates</span>
<span class="sd">	and the calculated flexibility of that bin.</span>
<span class="sd">	Example:</span>
<span class="sd">		dna_flex(seq_a,500,100)</span>
<span class="sd">	&gt;&gt;&gt; [(&#39;0-500&#39;, 9.7),(&#39;100-600&#39;, 9.77),...]</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
		<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Algorithm window size : </span><span class="si">%d</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">window_size</span><span class="p">)</span>
		<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Algorithm window step : </span><span class="si">%d</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">step_zize</span><span class="p">)</span>
		<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Sequence has </span><span class="si">{}</span><span class="s2"> bases&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">seq</span><span class="p">)))</span>
		
	<span class="n">algorithm</span> <span class="o">=</span> <span class="n">myFlex</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span><span class="n">window_size</span><span class="p">,</span><span class="n">step_zize</span><span class="p">)</span>
	<span class="n">flexibility_result</span> <span class="o">=</span> <span class="n">algorithm</span><span class="o">.</span><span class="n">analyse</span><span class="p">(</span><span class="n">flexibility_data</span><span class="p">)</span>
	
	<span class="k">return</span> <span class="n">flexibility_result</span><span class="o">.</span><span class="n">report</span><span class="p">(</span><span class="n">verbose</span><span class="p">)</span></div>


<span class="c1">##Repeats scanner##</span>

<span class="c1">#G-quadruplex</span>
<div class="viewcode-block" id="g4_scanner"><a class="viewcode-back" href="../code.html#alefuncs.g4_scanner">[docs]</a><span class="k">def</span> <span class="nf">g4_scanner</span><span class="p">(</span><span class="n">sequence</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	G-quadruplex motif scanner.</span>
<span class="sd">	Scan a sequence for the presence of the regex motif:</span>
<span class="sd">		[G]{3,5}[ACGT]{1,7}[G]{3,5}[ACGT]{1,7}[G]{3,5}[ACGT]{1,7}[G]{3,5}</span>
<span class="sd">		Reference: http://www.ncbi.nlm.nih.gov/pmc/articles/PMC1636468/</span>
<span class="sd">	Return two callable iterators.</span>
<span class="sd">	The first one contains G4 found on the + strand.</span>
<span class="sd">	The second contains the complementary G4 found on the + strand, i.e. a G4 in the - strand.</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="c1">#forward G4</span>
	<span class="n">pattern_f</span> <span class="o">=</span> <span class="s1">&#39;[G]{3,5}[ACGT]{1,7}[G]{3,5}[ACGT]{1,7}[G]{3,5}[ACGT]{1,7}[G]{3,5}&#39;</span>
	<span class="n">result_f</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">finditer</span><span class="p">(</span><span class="n">pattern_f</span><span class="p">,</span> <span class="n">sequence</span><span class="p">)</span>
	<span class="c1">#reverse G4</span>
	<span class="n">pattern_r</span> <span class="o">=</span> <span class="s1">&#39;[C]{3,5}[ACGT]{1,7}[C]{3,5}[ACGT]{1,7}[C]{3,5}[ACGT]{1,7}[C]{3,5}&#39;</span>
	<span class="n">result_r</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">finditer</span><span class="p">(</span><span class="n">pattern_r</span><span class="p">,</span> <span class="n">sequence</span><span class="p">)</span>
	
	<span class="k">return</span> <span class="n">result_f</span><span class="p">,</span> <span class="n">result_r</span></div>


<span class="c1">#Repeat-masker</span>
<div class="viewcode-block" id="parse_RepeatMasker"><a class="viewcode-back" href="../code.html#alefuncs.parse_RepeatMasker">[docs]</a><span class="k">def</span> <span class="nf">parse_RepeatMasker</span><span class="p">(</span><span class="n">infile</span><span class="o">=</span><span class="s2">&quot;RepeatMasker.txt&quot;</span><span class="p">,</span><span class="n">rep_type</span><span class="o">=</span><span class="s1">&#39;class&#39;</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	Parse RepeatMasker.txt and return a dict of bins for each chromosome</span>
<span class="sd">	and a set of repeats found on that bin.</span>
<span class="sd">	</span>
<span class="sd">	dict = {&#39;chromosome&#39;:{&#39;bin&#39;:set(repeats)}}</span>
<span class="sd">	&#39;&#39;&#39;</span>
	
	<span class="n">chromosomes</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">c</span><span class="p">)</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">23</span><span class="p">)]</span><span class="o">+</span><span class="p">[</span><span class="s1">&#39;X&#39;</span><span class="p">,</span><span class="s1">&#39;Y&#39;</span><span class="p">]</span>
	<span class="n">result</span> <span class="o">=</span> <span class="p">{}</span>
	
	<span class="k">if</span> <span class="n">rep_type</span> <span class="o">==</span> <span class="s1">&#39;name&#39;</span><span class="p">:</span>
		<span class="n">idx</span> <span class="o">=</span> <span class="mi">10</span> <span class="c1">#repName</span>
	<span class="k">elif</span> <span class="n">rep_type</span> <span class="o">==</span> <span class="s1">&#39;class&#39;</span><span class="p">:</span>
		<span class="n">idx</span> <span class="o">=</span> <span class="mi">11</span> <span class="c1">#repClass</span>
	<span class="k">elif</span> <span class="n">rep_type</span> <span class="o">==</span> <span class="s1">&#39;family&#39;</span><span class="p">:</span>
		<span class="n">idx</span> <span class="o">=</span> <span class="mi">12</span> <span class="c1">#repFamily</span>
	<span class="k">else</span><span class="p">:</span>
		<span class="k">raise</span> <span class="ne">NameError</span><span class="p">(</span><span class="s1">&#39;Invalid rep_type &quot;</span><span class="si">{}</span><span class="s1">&quot;. Expected &quot;class&quot;,&quot;family&quot; or &quot;name&quot;&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">rep_type</span><span class="p">))</span>
			
	<span class="c1">#RepeatMasker.txt is around 500MB!</span>
	<span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">yield_file</span><span class="p">(</span><span class="n">infile</span><span class="p">):</span>
		<span class="n">data</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
		<span class="n">chromosome</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;chr&#39;</span><span class="p">,</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
		<span class="n">start</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span>
		<span class="n">end</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span>
		<span class="n">bin_</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="si">{}</span><span class="s1">-</span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">start</span><span class="p">,</span><span class="n">end</span><span class="p">)</span>
		<span class="n">repeat</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;?&#39;</span><span class="p">,</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
		
		<span class="k">if</span> <span class="n">chromosome</span> <span class="ow">in</span> <span class="n">chromosomes</span><span class="p">:</span>
			<span class="k">if</span> <span class="n">chromosome</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">result</span><span class="p">:</span>
				<span class="n">result</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="n">chromosome</span><span class="p">:{</span><span class="n">bin_</span><span class="p">:</span><span class="nb">set</span><span class="p">([</span><span class="n">repeat</span><span class="p">])}})</span>
			<span class="k">else</span><span class="p">:</span>
				<span class="k">if</span> <span class="n">bin_</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">result</span><span class="p">[</span><span class="n">chromosome</span><span class="p">]:</span>
					<span class="n">result</span><span class="p">[</span><span class="n">chromosome</span><span class="p">]</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="n">bin_</span><span class="p">:</span><span class="nb">set</span><span class="p">([</span><span class="n">repeat</span><span class="p">])})</span>
				<span class="k">else</span><span class="p">:</span>
					<span class="n">result</span><span class="p">[</span><span class="n">chromosome</span><span class="p">][</span><span class="n">bin_</span><span class="p">]</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">repeat</span><span class="p">)</span>
				
	<span class="k">return</span> <span class="n">result</span></div>


<div class="viewcode-block" id="next_day"><a class="viewcode-back" href="../code.html#alefuncs.next_day">[docs]</a><span class="k">def</span> <span class="nf">next_day</span><span class="p">(</span><span class="n">d</span><span class="o">=</span><span class="s1">&#39;2012-12-04&#39;</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;Return the next day in the calendar.&#39;&#39;&#39;</span>
	<span class="n">Y</span><span class="p">,</span><span class="n">M</span><span class="p">,</span><span class="n">D</span> <span class="o">=</span> <span class="n">d</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)</span>
	<span class="n">t</span> <span class="o">=</span> <span class="n">datetime</span><span class="o">.</span><span class="n">date</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">Y</span><span class="p">),</span><span class="nb">int</span><span class="p">(</span><span class="n">M</span><span class="p">),</span><span class="nb">int</span><span class="p">(</span><span class="n">D</span><span class="p">))</span>
	<span class="n">_next</span> <span class="o">=</span> <span class="n">t</span> <span class="o">+</span> <span class="n">datetime</span><span class="o">.</span><span class="n">timedelta</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
	<span class="k">return</span> <span class="nb">str</span><span class="p">(</span><span class="n">_next</span><span class="p">)</span></div>
<span class="c1"># next_day(&#39;2012-12-31&#39;)</span>
<span class="c1"># &gt;&gt;&gt; &#39;2013-01-01&#39;</span>


<div class="viewcode-block" id="previous_day"><a class="viewcode-back" href="../code.html#alefuncs.previous_day">[docs]</a><span class="k">def</span> <span class="nf">previous_day</span><span class="p">(</span><span class="n">d</span><span class="o">=</span><span class="s1">&#39;2012-12-04&#39;</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;Return the previous day in the calendar.&#39;&#39;&#39;</span>
	<span class="n">Y</span><span class="p">,</span><span class="n">M</span><span class="p">,</span><span class="n">D</span> <span class="o">=</span> <span class="n">d</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)</span>
	<span class="n">t</span> <span class="o">=</span> <span class="n">datetime</span><span class="o">.</span><span class="n">date</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">Y</span><span class="p">),</span><span class="nb">int</span><span class="p">(</span><span class="n">M</span><span class="p">),</span><span class="nb">int</span><span class="p">(</span><span class="n">D</span><span class="p">))</span>
	<span class="n">_prev</span> <span class="o">=</span> <span class="n">t</span> <span class="o">+</span> <span class="n">datetime</span><span class="o">.</span><span class="n">timedelta</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span>
	<span class="k">return</span> <span class="nb">str</span><span class="p">(</span><span class="n">_prev</span><span class="p">)</span></div>
<span class="c1"># previous_day(&#39;2013-01-01&#39;)</span>
<span class="c1"># &gt;&gt;&gt; &#39;2012-12-31&#39;</span>


<div class="viewcode-block" id="intersect"><a class="viewcode-back" href="../code.html#alefuncs.intersect">[docs]</a><span class="k">def</span> <span class="nf">intersect</span><span class="p">(</span><span class="n">list1</span><span class="p">,</span><span class="n">list2</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;(list,list) =&gt; list</span>
<span class="sd">	Return the intersection of two lists, i.e. the item in common.</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="k">return</span> <span class="p">[</span><span class="n">item</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">list2</span> <span class="k">if</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">list1</span><span class="p">]</span></div>


<div class="viewcode-block" id="annotate_fusion_genes"><a class="viewcode-back" href="../code.html#alefuncs.annotate_fusion_genes">[docs]</a><span class="k">def</span> <span class="nf">annotate_fusion_genes</span><span class="p">(</span><span class="n">dataset_file</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	Uses FusionGenes_Annotation.pl to find fusion genes in the dataset.</span>
<span class="sd">	Generates a new file containing all the annotations.</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="kn">import</span> <span class="nn">time</span>
	<span class="n">start</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
	<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;annotating&#39;</span><span class="p">,</span> <span class="n">dataset_file</span><span class="p">,</span> <span class="s1">&#39;...&#39;</span><span class="p">)</span>
	<span class="n">raw_output</span> <span class="o">=</span> <span class="n">run_perl</span><span class="p">(</span><span class="s1">&#39;FusionGenes_Annotation.pl&#39;</span><span class="p">,</span> <span class="n">dataset_file</span><span class="p">)</span>
	<span class="n">raw_list</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">raw_output</span><span class="p">)[</span><span class="mi">2</span><span class="p">:]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\\</span><span class="s1">n&#39;</span><span class="p">)</span>
	<span class="n">outfile</span> <span class="o">=</span> <span class="n">dataset_file</span><span class="p">[:</span><span class="o">-</span><span class="mi">4</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_annotated.txt&#39;</span>
	<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">outfile</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">outfile</span><span class="p">:</span>
		<span class="n">line_counter</span> <span class="o">=</span> <span class="mi">0</span>
		<span class="n">header</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;##ID&#39;</span><span class="p">,</span> <span class="s1">&#39;ChrA&#39;</span><span class="p">,</span> <span class="s1">&#39;StartA&#39;</span><span class="p">,</span> <span class="s1">&#39;EndA&#39;</span><span class="p">,</span> <span class="s1">&#39;ChrB&#39;</span><span class="p">,</span> <span class="s1">&#39;StartB&#39;</span><span class="p">,</span> <span class="s1">&#39;EndB&#39;</span><span class="p">,</span> <span class="s1">&#39;CnvType&#39;</span><span class="p">,</span> <span class="s1">&#39;Orientation&#39;</span><span class="p">,</span>
				  <span class="s1">&#39;GeneA&#39;</span><span class="p">,</span> <span class="s1">&#39;StrandA&#39;</span><span class="p">,</span> <span class="s1">&#39;LastExonA&#39;</span><span class="p">,</span> <span class="s1">&#39;TotalExonsA&#39;</span><span class="p">,</span> <span class="s1">&#39;PhaseA&#39;</span><span class="p">,</span>
				  <span class="s1">&#39;GeneB&#39;</span><span class="p">,</span> <span class="s1">&#39;StrandB&#39;</span><span class="p">,</span> <span class="s1">&#39;LastExonB&#39;</span><span class="p">,</span> <span class="s1">&#39;TotalExonsB&#39;</span><span class="p">,</span> <span class="s1">&#39;PhaseB&#39;</span><span class="p">,</span>
				  <span class="s1">&#39;InFrame&#39;</span><span class="p">,</span> <span class="s1">&#39;InPhase&#39;</span><span class="p">]</span>
		<span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">list_to_line</span><span class="p">(</span><span class="n">header</span><span class="p">,</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
		<span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">raw_list</span><span class="p">:</span>
			<span class="n">cleaned_item</span> <span class="o">=</span> <span class="n">item</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\\</span><span class="s1">t&#39;</span><span class="p">)</span>
			<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">cleaned_item</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">10</span><span class="p">:</span> <span class="c1"># FusionGenes_Annotation.pl return the data twice. We kepp the annotated one.</span>
				<span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">list_to_line</span><span class="p">(</span><span class="n">cleaned_item</span><span class="p">,</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
				<span class="n">line_counter</span> <span class="o">+=</span> <span class="mi">1</span>
	<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;succesfully annotated&#39;</span><span class="p">,</span><span class="n">line_counter</span><span class="p">,</span> <span class="s1">&#39;breakpoints from&#39;</span><span class="p">,</span> <span class="n">dataset_file</span><span class="p">,</span> <span class="s1">&#39;in&#39;</span><span class="p">,</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span><span class="o">-</span><span class="n">start</span><span class="p">,</span> <span class="s1">&#39;seconds&#39;</span><span class="p">)</span> 
	<span class="c1"># track threads</span>
	<span class="k">try</span><span class="p">:</span>
		<span class="k">global</span> <span class="n">running_threads</span>
		<span class="n">running_threads</span> <span class="o">-=</span> <span class="mi">1</span>
	<span class="k">except</span><span class="p">:</span>
		<span class="k">pass</span></div>
<span class="c1"># dataset_file = &#39;/home/amarcozz/Documents/Projects/Fusion Genes/Scripts/test_datasets/public/breaks/Decipher-DeletionsOnly.txt&#39;</span>
<span class="c1"># annotate_fusion_genes(dataset_file)</span>


<div class="viewcode-block" id="blastn"><a class="viewcode-back" href="../code.html#alefuncs.blastn">[docs]</a><span class="k">def</span> <span class="nf">blastn</span><span class="p">(</span><span class="n">input_fasta_file</span><span class="p">,</span><span class="n">db_path</span><span class="o">=</span><span class="s1">&#39;/Users/amarcozzi/Desktop/BLAST_DB/&#39;</span><span class="p">,</span><span class="n">db_name</span><span class="o">=</span><span class="s1">&#39;human_genomic&#39;</span><span class="p">,</span><span class="n">out_file</span><span class="o">=</span><span class="s1">&#39;blastn_out.xml&#39;</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	Run blastn on the local machine using a local database.</span>
<span class="sd">	Requires NCBI BLAST+ to be installed. http://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&amp;PAGE_TYPE=BlastDocs&amp;DOC_TYPE=Download</span>
<span class="sd">	Takes a fasta file as input and writes the output in an XML file.</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="kn">from</span> <span class="nn">Bio.Blast.Applications</span> <span class="k">import</span> <span class="n">NcbiblastnCommandline</span>
	<span class="n">db</span> <span class="o">=</span> <span class="n">db_path</span> <span class="o">+</span> <span class="n">db_name</span>
	<span class="n">blastn_cline</span> <span class="o">=</span> <span class="n">NcbiblastnCommandline</span><span class="p">(</span><span class="n">query</span><span class="o">=</span><span class="n">input_fasta_file</span><span class="p">,</span> <span class="n">db</span><span class="o">=</span><span class="n">db</span><span class="p">,</span> <span class="n">evalue</span><span class="o">=</span><span class="mf">0.001</span><span class="p">,</span> <span class="n">outfmt</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">out</span><span class="o">=</span><span class="n">out_file</span><span class="p">)</span>
	<span class="nb">print</span><span class="p">(</span><span class="n">blastn_cline</span><span class="p">)</span>
	<span class="n">stdout</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">blastn_cline</span><span class="p">()</span></div>
<span class="c1"># to be tested</span>


<div class="viewcode-block" id="check_line"><a class="viewcode-back" href="../code.html#alefuncs.check_line">[docs]</a><span class="k">def</span> <span class="nf">check_line</span><span class="p">(</span><span class="n">line</span><span class="p">,</span><span class="n">unexpected_char</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">,</span><span class="s1">&#39;&#39;</span><span class="p">,</span><span class="s1">&#39; &#39;</span><span class="p">,</span><span class="s1">&#39;#&#39;</span><span class="p">]):</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	Check if the line starts with an unexpected character.</span>
<span class="sd">	If so, return False, else True</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">unexpected_char</span><span class="p">:</span>
		<span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="n">item</span><span class="p">):</span>
			<span class="k">return</span> <span class="kc">False</span>
	<span class="k">return</span> <span class="kc">True</span></div>


<div class="viewcode-block" id="dice_coefficient"><a class="viewcode-back" href="../code.html#alefuncs.dice_coefficient">[docs]</a><span class="k">def</span> <span class="nf">dice_coefficient</span><span class="p">(</span><span class="n">sequence_a</span><span class="p">,</span><span class="n">sequence_b</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;(str, str) =&gt; float</span>
<span class="sd">	Return the dice cofficient of two sequences.</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="n">a</span> <span class="o">=</span> <span class="n">sequence_a</span>
	<span class="n">b</span> <span class="o">=</span> <span class="n">sequence_b</span>
	<span class="k">if</span> <span class="ow">not</span> <span class="nb">len</span><span class="p">(</span><span class="n">a</span><span class="p">)</span> <span class="ow">or</span> <span class="ow">not</span> <span class="nb">len</span><span class="p">(</span><span class="n">b</span><span class="p">):</span> <span class="k">return</span> <span class="mf">0.0</span>
	<span class="c1"># quick case for true duplicates</span>
	<span class="k">if</span> <span class="n">a</span> <span class="o">==</span> <span class="n">b</span><span class="p">:</span> <span class="k">return</span> <span class="mf">1.0</span>
	<span class="c1"># if a != b, and a or b are single chars, then they can&#39;t possibly match</span>
	<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">a</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span> <span class="ow">or</span> <span class="nb">len</span><span class="p">(</span><span class="n">b</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span> <span class="k">return</span> <span class="mf">0.0</span>
	
	<span class="c1"># list comprehension, preferred over list.append() &#39;&#39;&#39;</span>
	<span class="n">a_bigram_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">a</span><span class="p">[</span><span class="n">i</span><span class="p">:</span><span class="n">i</span><span class="o">+</span><span class="mi">2</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">a</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span><span class="p">)]</span>
	<span class="n">b_bigram_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">b</span><span class="p">[</span><span class="n">i</span><span class="p">:</span><span class="n">i</span><span class="o">+</span><span class="mi">2</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">b</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span><span class="p">)]</span>
	
	<span class="n">a_bigram_list</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
	<span class="n">b_bigram_list</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
	
	<span class="c1"># assignments to save function calls</span>
	<span class="n">lena</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">a_bigram_list</span><span class="p">)</span>
	<span class="n">lenb</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">b_bigram_list</span><span class="p">)</span>
	<span class="c1"># initialize match counters</span>
	<span class="n">matches</span> <span class="o">=</span> <span class="n">i</span> <span class="o">=</span> <span class="n">j</span> <span class="o">=</span> <span class="mi">0</span>
	<span class="k">while</span> <span class="p">(</span><span class="n">i</span> <span class="o">&lt;</span> <span class="n">lena</span> <span class="ow">and</span> <span class="n">j</span> <span class="o">&lt;</span> <span class="n">lenb</span><span class="p">):</span>
		<span class="k">if</span> <span class="n">a_bigram_list</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">==</span> <span class="n">b_bigram_list</span><span class="p">[</span><span class="n">j</span><span class="p">]:</span>
			<span class="n">matches</span> <span class="o">+=</span> <span class="mi">2</span>
			<span class="n">i</span> <span class="o">+=</span> <span class="mi">1</span>
			<span class="n">j</span> <span class="o">+=</span> <span class="mi">1</span>
		<span class="k">elif</span> <span class="n">a_bigram_list</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">b_bigram_list</span><span class="p">[</span><span class="n">j</span><span class="p">]:</span>
			<span class="n">i</span> <span class="o">+=</span> <span class="mi">1</span>
		<span class="k">else</span><span class="p">:</span>
			<span class="n">j</span> <span class="o">+=</span> <span class="mi">1</span>
	
	<span class="n">score</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">matches</span><span class="p">)</span><span class="o">/</span><span class="nb">float</span><span class="p">(</span><span class="n">lena</span> <span class="o">+</span> <span class="n">lenb</span><span class="p">)</span>
	<span class="k">return</span> <span class="n">score</span></div>


<div class="viewcode-block" id="find_path"><a class="viewcode-back" href="../code.html#alefuncs.find_path">[docs]</a><span class="k">def</span> <span class="nf">find_path</span><span class="p">(</span><span class="n">graph</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">path</span><span class="o">=</span><span class="p">[]):</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	Find a path between two nodes in a graph.</span>
<span class="sd">	Works on graphs like this:</span>
<span class="sd">			graph ={&#39;A&#39;: [&#39;B&#39;, &#39;C&#39;],</span>
<span class="sd">					&#39;B&#39;: [&#39;C&#39;, &#39;D&#39;],</span>
<span class="sd">					&#39;C&#39;: [&#39;D&#39;],</span>
<span class="sd">					&#39;D&#39;: [&#39;C&#39;],</span>
<span class="sd">					&#39;E&#39;: [&#39;F&#39;],</span>
<span class="sd">					&#39;F&#39;: [&#39;C&#39;]}</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="n">path</span> <span class="o">=</span> <span class="n">path</span> <span class="o">+</span> <span class="p">[</span><span class="n">start</span><span class="p">]</span>
	<span class="k">if</span> <span class="n">start</span> <span class="o">==</span> <span class="n">end</span><span class="p">:</span>
		<span class="k">return</span> <span class="n">path</span>
	<span class="k">if</span> <span class="ow">not</span> <span class="n">graph</span><span class="o">.</span><span class="n">has_key</span><span class="p">(</span><span class="n">start</span><span class="p">):</span>
		<span class="k">return</span> <span class="kc">None</span>
	<span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">graph</span><span class="p">[</span><span class="n">start</span><span class="p">]:</span>
		<span class="k">if</span> <span class="n">node</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">path</span><span class="p">:</span>
			<span class="n">newpath</span> <span class="o">=</span> <span class="n">find_path</span><span class="p">(</span><span class="n">graph</span><span class="p">,</span> <span class="n">node</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">path</span><span class="p">)</span>
		<span class="k">if</span> <span class="n">newpath</span><span class="p">:</span> <span class="k">return</span> <span class="n">newpath</span>
	<span class="k">return</span> <span class="kc">None</span></div>


<div class="viewcode-block" id="find_all_paths"><a class="viewcode-back" href="../code.html#alefuncs.find_all_paths">[docs]</a><span class="k">def</span> <span class="nf">find_all_paths</span><span class="p">(</span><span class="n">graph</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">path</span><span class="o">=</span><span class="p">[]):</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	Find all paths between two nodes of a graph.</span>
<span class="sd">	Works on graphs like this:</span>
<span class="sd">			graph ={&#39;A&#39;: [&#39;B&#39;, &#39;C&#39;],</span>
<span class="sd">					&#39;B&#39;: [&#39;C&#39;, &#39;D&#39;],</span>
<span class="sd">					&#39;C&#39;: [&#39;D&#39;],</span>
<span class="sd">					&#39;D&#39;: [&#39;C&#39;],</span>
<span class="sd">					&#39;E&#39;: [&#39;F&#39;],</span>
<span class="sd">					&#39;F&#39;: [&#39;C&#39;]}</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="n">path</span> <span class="o">=</span> <span class="n">path</span> <span class="o">+</span> <span class="p">[</span><span class="n">start</span><span class="p">]</span>
	<span class="k">if</span> <span class="n">start</span> <span class="o">==</span> <span class="n">end</span><span class="p">:</span>
		<span class="k">return</span> <span class="p">[</span><span class="n">path</span><span class="p">]</span>
	<span class="k">if</span> <span class="ow">not</span> <span class="n">graph</span><span class="o">.</span><span class="n">has_key</span><span class="p">(</span><span class="n">start</span><span class="p">):</span>
		<span class="k">return</span> <span class="p">[]</span>
	<span class="n">paths</span> <span class="o">=</span> <span class="p">[]</span>
	<span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">graph</span><span class="p">[</span><span class="n">start</span><span class="p">]:</span>
		<span class="k">if</span> <span class="n">node</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">path</span><span class="p">:</span>
			<span class="n">newpaths</span> <span class="o">=</span> <span class="n">find_all_paths</span><span class="p">(</span><span class="n">graph</span><span class="p">,</span> <span class="n">node</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">path</span><span class="p">)</span>
			<span class="k">for</span> <span class="n">newpath</span> <span class="ow">in</span> <span class="n">newpaths</span><span class="p">:</span>
				<span class="n">paths</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">newpath</span><span class="p">)</span>
	<span class="k">return</span> <span class="n">paths</span></div>


<div class="viewcode-block" id="find_shortest_path"><a class="viewcode-back" href="../code.html#alefuncs.find_shortest_path">[docs]</a><span class="k">def</span> <span class="nf">find_shortest_path</span><span class="p">(</span><span class="n">graph</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">path</span><span class="o">=</span><span class="p">[]):</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	Find the shortest path between two nodes of a graph.</span>
<span class="sd">	Works on graphs like this:</span>
<span class="sd">			graph ={&#39;A&#39;: [&#39;B&#39;, &#39;C&#39;],</span>
<span class="sd">					&#39;B&#39;: [&#39;C&#39;, &#39;D&#39;],</span>
<span class="sd">					&#39;C&#39;: [&#39;D&#39;],</span>
<span class="sd">					&#39;D&#39;: [&#39;C&#39;],</span>
<span class="sd">					&#39;E&#39;: [&#39;F&#39;],</span>
<span class="sd">					&#39;F&#39;: [&#39;C&#39;]}</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="n">path</span> <span class="o">=</span> <span class="n">path</span> <span class="o">+</span> <span class="p">[</span><span class="n">start</span><span class="p">]</span>
	<span class="k">if</span> <span class="n">start</span> <span class="o">==</span> <span class="n">end</span><span class="p">:</span>
		<span class="k">return</span> <span class="n">path</span>
	<span class="k">if</span> <span class="ow">not</span> <span class="n">graph</span><span class="o">.</span><span class="n">has_key</span><span class="p">(</span><span class="n">start</span><span class="p">):</span>
		<span class="k">return</span> <span class="kc">None</span>
	<span class="n">shortest</span> <span class="o">=</span> <span class="kc">None</span>
	<span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">graph</span><span class="p">[</span><span class="n">start</span><span class="p">]:</span>
		<span class="k">if</span> <span class="n">node</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">path</span><span class="p">:</span>
			<span class="n">newpath</span> <span class="o">=</span> <span class="n">find_shortest_path</span><span class="p">(</span><span class="n">graph</span><span class="p">,</span> <span class="n">node</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">path</span><span class="p">)</span>
			<span class="k">if</span> <span class="n">newpath</span><span class="p">:</span>
				<span class="k">if</span> <span class="ow">not</span> <span class="n">shortest</span> <span class="ow">or</span> <span class="nb">len</span><span class="p">(</span><span class="n">newpath</span><span class="p">)</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">shortest</span><span class="p">):</span>
					<span class="n">shortest</span> <span class="o">=</span> <span class="n">newpath</span>
	<span class="k">return</span> <span class="n">shortest</span></div>


<span class="c1"># ##</span>
<span class="c1"># graph = {&#39;A&#39;: [&#39;B&#39;, &#39;C&#39;],</span>
<span class="c1">#        &#39;B&#39;: [&#39;C&#39;, &#39;D&#39;],</span>
<span class="c1">#        &#39;C&#39;: [&#39;D&#39;],</span>
<span class="c1">#        &#39;D&#39;: [&#39;C&#39;],</span>
<span class="c1">#        &#39;E&#39;: [&#39;F&#39;],</span>
<span class="c1">#        &#39;F&#39;: [&#39;C&#39;]}</span>

<span class="c1"># &gt;&gt;&gt; find_path(graph, &#39;A&#39;, &#39;D&#39;)</span>
<span class="c1">#     [&#39;A&#39;, &#39;B&#39;, &#39;C&#39;, &#39;D&#39;]</span>

<span class="c1"># &gt;&gt;&gt; find_all_paths(graph, &#39;A&#39;, &#39;D&#39;)</span>
<span class="c1">#     [[&#39;A&#39;, &#39;B&#39;, &#39;C&#39;, &#39;D&#39;], [&#39;A&#39;, &#39;B&#39;, &#39;D&#39;], [&#39;A&#39;, &#39;C&#39;, &#39;D&#39;]]</span>

<span class="c1"># &gt;&gt;&gt; find_shortest_path(graph, &#39;A&#39;, &#39;D&#39;)</span>
<span class="c1">#     [&#39;A&#39;, &#39;C&#39;, &#39;D&#39;]</span>

<div class="viewcode-block" id="gen_rnd_string"><a class="viewcode-back" href="../code.html#alefuncs.gen_rnd_string">[docs]</a><span class="k">def</span> <span class="nf">gen_rnd_string</span><span class="p">(</span><span class="n">length</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	Return a string of uppercase and lowercase ascii letters.</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="kn">import</span> <span class="nn">random</span><span class="o">,</span> <span class="nn">string</span>
	<span class="n">s</span> <span class="o">=</span> <span class="p">[</span><span class="n">l</span> <span class="k">for</span> <span class="n">l</span> <span class="ow">in</span> <span class="n">string</span><span class="o">.</span><span class="n">ascii_letters</span><span class="p">]</span>
	<span class="n">random</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
	<span class="n">s</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">s</span><span class="p">[:</span><span class="n">length</span><span class="p">])</span>
	<span class="k">return</span> <span class="n">s</span></div>

<span class="k">def</span> <span class="nf">gene_synonyms</span><span class="p">(</span><span class="n">gene_name</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;str =&gt; list()</span>
<span class="sd">	Queries http://rest.genenames.org and returns a list of synonyms of gene_name.</span>
<span class="sd">	Returns None if no synonym was found.</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="kn">import</span> <span class="nn">httplib2</span> <span class="k">as</span> <span class="nn">http</span>
	<span class="kn">import</span> <span class="nn">json</span>
	<span class="kn">from</span> <span class="nn">urllib.parse</span> <span class="k">import</span> <span class="n">urlparse</span>

	<span class="n">result</span> <span class="o">=</span> <span class="p">[]</span>
	<span class="n">headers</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;Accept&#39;</span><span class="p">:</span> <span class="s1">&#39;application/json&#39;</span><span class="p">}</span>

	<span class="n">uri</span> <span class="o">=</span> <span class="s1">&#39;http://rest.genenames.org&#39;</span>
	<span class="n">path</span> <span class="o">=</span> <span class="s1">&#39;/search/</span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">gene_name</span><span class="p">)</span>

	<span class="n">target</span> <span class="o">=</span> <span class="n">urlparse</span><span class="p">(</span><span class="n">uri</span><span class="o">+</span><span class="n">path</span><span class="p">)</span>
	<span class="n">method</span> <span class="o">=</span> <span class="s1">&#39;GET&#39;</span>
	<span class="n">body</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>

	<span class="n">h</span> <span class="o">=</span> <span class="n">http</span><span class="o">.</span><span class="n">Http</span><span class="p">()</span>

	<span class="n">response</span><span class="p">,</span> <span class="n">content</span> <span class="o">=</span> <span class="n">h</span><span class="o">.</span><span class="n">request</span><span class="p">(</span>
									<span class="n">target</span><span class="o">.</span><span class="n">geturl</span><span class="p">(),</span>
									<span class="n">method</span><span class="p">,</span>
									<span class="n">body</span><span class="p">,</span>
									<span class="n">headers</span> <span class="p">)</span>

	<span class="k">if</span> <span class="n">response</span><span class="p">[</span><span class="s1">&#39;status&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;200&#39;</span><span class="p">:</span>
		<span class="c1"># assume that content is a json reply</span>
		<span class="c1"># parse content with the json module </span>
		<span class="n">data</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">loads</span><span class="p">(</span><span class="n">content</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s1">&#39;utf8&#39;</span><span class="p">))</span>
		<span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;response&#39;</span><span class="p">][</span><span class="s1">&#39;docs&#39;</span><span class="p">]:</span>
			<span class="n">result</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">item</span><span class="p">[</span><span class="s1">&#39;symbol&#39;</span><span class="p">])</span>
		<span class="k">return</span> <span class="n">result</span>
	 
	<span class="k">else</span><span class="p">:</span>
		<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Error detected: &#39;</span> <span class="o">+</span> <span class="n">response</span><span class="p">[</span><span class="s1">&#39;status&#39;</span><span class="p">])</span>
		<span class="k">return</span> <span class="kc">None</span>
<span class="c1">#print(gene_synonyms(&#39;MLL3&#39;))</span>

<div class="viewcode-block" id="string_to_number"><a class="viewcode-back" href="../code.html#alefuncs.string_to_number">[docs]</a><span class="k">def</span> <span class="nf">string_to_number</span><span class="p">(</span><span class="n">s</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	Convert a bytes string into a single number.</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="k">return</span> <span class="nb">int</span><span class="o">.</span><span class="n">from_bytes</span><span class="p">(</span><span class="n">s</span><span class="o">.</span><span class="n">encode</span><span class="p">(),</span> <span class="s1">&#39;little&#39;</span><span class="p">)</span></div>
<span class="c1">#string_to_number(&#39;foo bar baz&#39;)</span>
<span class="c1">#&gt;&gt;&gt; 147948829660780569073512294</span>

<div class="viewcode-block" id="number_to_string"><a class="viewcode-back" href="../code.html#alefuncs.number_to_string">[docs]</a><span class="k">def</span> <span class="nf">number_to_string</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	Convert a number into a bytes string.</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="kn">import</span> <span class="nn">math</span>
	<span class="k">return</span> <span class="n">n</span><span class="o">.</span><span class="n">to_bytes</span><span class="p">(</span><span class="n">math</span><span class="o">.</span><span class="n">ceil</span><span class="p">(</span><span class="n">n</span><span class="o">.</span><span class="n">bit_length</span><span class="p">()</span> <span class="o">/</span> <span class="mi">8</span><span class="p">),</span> <span class="s1">&#39;little&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">decode</span><span class="p">()</span></div>
<span class="c1">#x = 147948829660780569073512294</span>
<span class="c1">#number_to_string(x)</span>
<span class="c1">#&gt;&gt;&gt; &#39;foo bar baz&#39;</span>
 
<div class="viewcode-block" id="determine_average_breaks_distance"><a class="viewcode-back" href="../code.html#alefuncs.determine_average_breaks_distance">[docs]</a><span class="k">def</span> <span class="nf">determine_average_breaks_distance</span><span class="p">(</span><span class="n">dataset</span><span class="p">):</span> <span class="c1"># tested only for deletion/duplication</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	Evaluate the average distance among breaks in a dataset.</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="n">data</span> <span class="o">=</span> <span class="n">extract_data</span><span class="p">(</span><span class="n">dataset</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">4</span><span class="p">,</span><span class="mi">5</span><span class="p">],</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
	<span class="n">to_average</span> <span class="o">=</span> <span class="p">[]</span>
	<span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">data</span><span class="p">:</span>
		<span class="k">if</span> <span class="n">item</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="n">item</span><span class="p">[</span><span class="mi">2</span><span class="p">]:</span>
			<span class="n">to_average</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">item</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span><span class="o">-</span><span class="nb">int</span><span class="p">(</span><span class="n">item</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>
	<span class="k">return</span> <span class="nb">sum</span><span class="p">(</span><span class="n">to_average</span><span class="p">)</span><span class="o">/</span><span class="nb">len</span><span class="p">(</span><span class="n">to_average</span><span class="p">)</span></div>
<span class="c1">#print(determine_average_breaks_distance(&#39;/home/amarcozz/Documents/Projects/Fusion Genes/Scripts/test_datasets/random/sorted/rnd_dataset_100_annotated_sorted.txt&#39;))</span>


<div class="viewcode-block" id="dict_overview"><a class="viewcode-back" href="../code.html#alefuncs.dict_overview">[docs]</a><span class="k">def</span> <span class="nf">dict_overview</span><span class="p">(</span><span class="n">dictionary</span><span class="p">,</span><span class="n">how_many_keys</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	Prints out how_many_elements of the target dictionary.</span>
<span class="sd">	Useful to have a quick look at the structure of a dictionary.</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="kn">from</span> <span class="nn">itertools</span> <span class="k">import</span> <span class="n">islice</span>
	<span class="n">ks</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">islice</span><span class="p">(</span><span class="n">dictionary</span><span class="p">,</span> <span class="n">how_many_keys</span><span class="p">))</span>
	<span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">ks</span><span class="p">:</span>
		<span class="nb">print</span><span class="p">(</span><span class="n">k</span><span class="p">)</span>
		<span class="nb">print</span><span class="p">(</span><span class="n">dictionary</span><span class="p">[</span><span class="n">k</span><span class="p">])</span></div>


<div class="viewcode-block" id="download_human_genome"><a class="viewcode-back" href="../code.html#alefuncs.download_human_genome">[docs]</a><span class="k">def</span> <span class="nf">download_human_genome</span><span class="p">(</span><span class="n">build</span><span class="o">=</span><span class="s1">&#39;GRCh37&#39;</span><span class="p">,</span> <span class="n">entrez_usr_email</span><span class="o">=</span><span class="s2">&quot;A.E.vanvlimmeren@students.uu.nl&quot;</span><span class="p">):</span> <span class="c1">#beta: works properly only forGRCh37</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	Download the Human genome from enterez.</span>
<span class="sd">	Save each chromosome in a separate txt file.</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="kn">from</span> <span class="nn">Bio</span> <span class="k">import</span> <span class="n">Entrez</span><span class="p">,</span> <span class="n">SeqIO</span>

	<span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="n">entrez_usr_email</span>

	<span class="c1">#Last available version</span>
	<span class="n">NCBI_IDS</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;1&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000001&quot;</span><span class="p">,</span> <span class="s1">&#39;2&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000002&quot;</span><span class="p">,</span><span class="s1">&#39;3&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000003&quot;</span><span class="p">,</span><span class="s1">&#39;4&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000004&quot;</span><span class="p">,</span>
				<span class="s1">&#39;5&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000005&quot;</span><span class="p">,</span><span class="s1">&#39;6&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000006&quot;</span><span class="p">,</span><span class="s1">&#39;7&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000007&quot;</span><span class="p">,</span> <span class="s1">&#39;8&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000008&quot;</span><span class="p">,</span>
				<span class="s1">&#39;9&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000009&quot;</span><span class="p">,</span> <span class="s1">&#39;10&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000010&quot;</span><span class="p">,</span> <span class="s1">&#39;11&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000011&quot;</span><span class="p">,</span> <span class="s1">&#39;12&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000012&quot;</span><span class="p">,</span>
				<span class="s1">&#39;13&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000013&quot;</span><span class="p">,</span><span class="s1">&#39;14&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000014&quot;</span><span class="p">,</span> <span class="s1">&#39;15&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000015&quot;</span><span class="p">,</span> <span class="s1">&#39;16&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000016&quot;</span><span class="p">,</span> 
				<span class="s1">&#39;17&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000017&quot;</span><span class="p">,</span> <span class="s1">&#39;18&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000018&quot;</span><span class="p">,</span> <span class="s1">&#39;19&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000019&quot;</span><span class="p">,</span> <span class="s1">&#39;20&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000020&quot;</span><span class="p">,</span>
				<span class="s1">&#39;21&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000021&quot;</span><span class="p">,</span> <span class="s1">&#39;22&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000022&quot;</span><span class="p">,</span> <span class="s1">&#39;X&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000023&quot;</span><span class="p">,</span> <span class="s1">&#39;Y&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000024&quot;</span><span class="p">}</span>


	<span class="c1">#GRCh37 from http://www.ncbi.nlm.nih.gov/assembly/GCF_000001405.25/#/def_asm_Primary_Assembly</span>
	<span class="n">NCBI_IDS_GRCh37</span> <span class="o">=</span> <span class="p">{</span> <span class="s1">&#39;NC_000001.10&#39;</span><span class="p">,</span><span class="s1">&#39;NC_000002.11&#39;</span><span class="p">,</span><span class="s1">&#39;NC_000003.11&#39;</span><span class="p">,</span><span class="s1">&#39;NC_000004.11&#39;</span><span class="p">,</span>
						<span class="s1">&#39;NC_000005.9&#39;</span><span class="p">,</span><span class="s1">&#39;NC_000006.11&#39;</span><span class="p">,</span><span class="s1">&#39;NC_000007.13&#39;</span><span class="p">,</span><span class="s1">&#39;NC_000008.10&#39;</span><span class="p">,</span>
						<span class="s1">&#39;NC_000009.11&#39;</span><span class="p">,</span><span class="s1">&#39;NC_000010.10&#39;</span><span class="p">,</span><span class="s1">&#39;NC_000011.9&#39;</span><span class="p">,</span><span class="s1">&#39;NC_000012.11&#39;</span><span class="p">,</span>
						<span class="s1">&#39;NC_000013.10&#39;</span><span class="p">,</span><span class="s1">&#39;NC_000014.8&#39;</span><span class="p">,</span><span class="s1">&#39;NC_000015.9&#39;</span><span class="p">,</span><span class="s1">&#39;NC_000016.9&#39;</span><span class="p">,</span>
						<span class="s1">&#39;NC_000017.10&#39;</span><span class="p">,</span><span class="s1">&#39;NC_000018.9&#39;</span><span class="p">,</span><span class="s1">&#39;NC_000019.9&#39;</span><span class="p">,</span><span class="s1">&#39;NC_000020.10&#39;</span><span class="p">,</span>
						<span class="s1">&#39;NC_000021.8&#39;</span><span class="p">,</span><span class="s1">&#39;NC_000022.10&#39;</span><span class="p">,</span><span class="s1">&#39;NC_000023.10&#39;</span><span class="p">,</span><span class="s1">&#39;NC_000024.9&#39;</span><span class="p">}</span>
	

	<span class="n">CHR_LENGTHS_GRCh37</span> <span class="o">=</span> <span class="p">{</span>  <span class="s1">&#39;1&#39;</span><span class="p">:</span><span class="mi">249250621</span><span class="p">,</span><span class="s1">&#39;2&#39;</span> <span class="p">:</span><span class="mi">243199373</span><span class="p">,</span><span class="s1">&#39;3&#39;</span> <span class="p">:</span><span class="mi">198022430</span><span class="p">,</span><span class="s1">&#39;4&#39;</span> <span class="p">:</span><span class="mi">191154276</span><span class="p">,</span>
							<span class="s1">&#39;5&#39;</span> <span class="p">:</span><span class="mi">180915260</span><span class="p">,</span><span class="s1">&#39;6&#39;</span> <span class="p">:</span><span class="mi">171115067</span><span class="p">,</span><span class="s1">&#39;7&#39;</span> <span class="p">:</span><span class="mi">159138663</span><span class="p">,</span><span class="s1">&#39;8&#39;</span> <span class="p">:</span><span class="mi">146364022</span><span class="p">,</span>
							<span class="s1">&#39;9&#39;</span> <span class="p">:</span><span class="mi">141213431</span><span class="p">,</span><span class="s1">&#39;10&#39;</span><span class="p">:</span><span class="mi">135534747</span><span class="p">,</span><span class="s1">&#39;11&#39;</span><span class="p">:</span><span class="mi">135006516</span><span class="p">,</span><span class="s1">&#39;12&#39;</span><span class="p">:</span><span class="mi">133851895</span><span class="p">,</span>
							<span class="s1">&#39;13&#39;</span><span class="p">:</span><span class="mi">115169878</span><span class="p">,</span><span class="s1">&#39;14&#39;</span><span class="p">:</span><span class="mi">107349540</span><span class="p">,</span><span class="s1">&#39;15&#39;</span><span class="p">:</span><span class="mi">102531392</span><span class="p">,</span><span class="s1">&#39;16&#39;</span><span class="p">:</span><span class="mi">90354753</span><span class="p">,</span>
							<span class="s1">&#39;17&#39;</span><span class="p">:</span><span class="mi">81195210</span><span class="p">,</span><span class="s1">&#39;18&#39;</span><span class="p">:</span><span class="mi">78077248</span><span class="p">,</span><span class="s1">&#39;19&#39;</span><span class="p">:</span><span class="mi">59128983</span><span class="p">,</span><span class="s1">&#39;20&#39;</span><span class="p">:</span><span class="mi">63025520</span><span class="p">,</span>
							<span class="s1">&#39;21&#39;</span><span class="p">:</span><span class="mi">48129895</span><span class="p">,</span><span class="s1">&#39;22&#39;</span><span class="p">:</span><span class="mi">51304566</span><span class="p">,</span><span class="s1">&#39;X&#39;</span> <span class="p">:</span><span class="mi">155270560</span><span class="p">,</span><span class="s1">&#39;Y&#39;</span> <span class="p">:</span><span class="mi">59373566</span><span class="p">}</span>


	<span class="k">if</span> <span class="n">build</span> <span class="o">==</span> <span class="s1">&#39;GRCh37&#39;</span><span class="p">:</span>
		<span class="n">NCBI_IDS</span> <span class="o">=</span> <span class="n">NCBI_IDS_GRCh37</span>
		<span class="n">CHR_LENGTHS</span> <span class="o">=</span> <span class="n">CHR_LENGTHS_GRCh37</span>
	<span class="k">else</span><span class="p">:</span>
		<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;This function only work with genome build GRCh37 fow now...&#39;</span><span class="p">)</span>
		<span class="k">return</span> <span class="kc">False</span>


	<span class="n">idx</span> <span class="o">=</span> <span class="mi">0</span>
	<span class="k">for</span> <span class="n">target_chromosome</span> <span class="ow">in</span> <span class="n">NCBI_IDS</span><span class="p">:</span>
		<span class="n">length</span> <span class="o">=</span> <span class="n">CHR_LENGTHS</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span>
		<span class="n">idx</span> <span class="o">+=</span> <span class="mi">1</span>

		<span class="n">sequence</span> <span class="o">=</span> <span class="kc">False</span>

		<span class="k">try</span><span class="p">:</span>
				 <span class="c1"># Always tell NCBI who you are</span>
			<span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">efetch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s2">&quot;nucleotide&quot;</span><span class="p">,</span> 
								   <span class="nb">id</span><span class="o">=</span><span class="n">target_chromosome</span><span class="p">,</span> 
								   <span class="n">rettype</span><span class="o">=</span><span class="s2">&quot;fasta&quot;</span><span class="p">,</span> 
								   <span class="n">strand</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> 
								   <span class="n">seq_start</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="c1">#this is to obtain actual start coordinates from the index</span>
								   <span class="n">seq_stop</span><span class="o">=</span><span class="n">length</span><span class="p">)</span> <span class="c1"># this is the end of the chromosome</span>
			<span class="n">record</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
			<span class="n">handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
			<span class="n">sequence</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>

		<span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
			<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;ValueError: no sequence found in NCBI&#39;</span><span class="p">)</span>

		<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;sequence_</span><span class="si">{}</span><span class="s1">.txt&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">target_chromosome</span><span class="p">),</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
			<span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>   </div>


<div class="viewcode-block" id="exponential_range"><a class="viewcode-back" href="../code.html#alefuncs.exponential_range">[docs]</a><span class="k">def</span> <span class="nf">exponential_range</span><span class="p">(</span><span class="n">start</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span><span class="n">end</span><span class="o">=</span><span class="mi">10000</span><span class="p">,</span><span class="n">base</span><span class="o">=</span><span class="mi">10</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	Generates a range of integer that grow exponentially.</span>
<span class="sd">	Example: list(exp_range(0,100000,2))</span>
<span class="sd">	Output :[0,</span>
<span class="sd">			 2,</span>
<span class="sd">			 4,</span>
<span class="sd">			 8,</span>
<span class="sd">			 16,</span>
<span class="sd">			 32,</span>
<span class="sd">			 64,</span>
<span class="sd">			 128,</span>
<span class="sd">			 256,</span>
<span class="sd">			 512,</span>
<span class="sd">			 1024,</span>
<span class="sd">			 2048,</span>
<span class="sd">			 4096,</span>
<span class="sd">			 8192,</span>
<span class="sd">			 16384,</span>
<span class="sd">			 32768,</span>
<span class="sd">			 65536]</span>
<span class="sd">	&#39;&#39;&#39;</span>

	<span class="k">if</span> <span class="n">end</span><span class="o">/</span><span class="n">base</span> <span class="o">&lt;</span> <span class="n">base</span><span class="p">:</span>
		<span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;&quot;end&quot; must be at least &quot;base**2&quot;&#39;</span><span class="p">)</span>
	<span class="n">result</span> <span class="o">=</span> <span class="p">[]</span>
	 
	<span class="n">new_start</span> <span class="o">=</span> <span class="n">start</span>
	<span class="n">new_end</span> <span class="o">=</span> <span class="n">base</span><span class="o">**</span><span class="mi">2</span>
	<span class="n">new_base</span> <span class="o">=</span> <span class="n">base</span>
	
	<span class="k">while</span> <span class="n">new_start</span> <span class="o">&lt;</span> <span class="n">end</span><span class="p">:</span>
		<span class="n">result</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">new_start</span><span class="p">,</span><span class="n">new_end</span><span class="p">,</span><span class="n">new_base</span><span class="p">))</span>
		
		<span class="n">new_start</span> <span class="o">=</span> <span class="n">new_end</span>
		<span class="n">new_end</span> <span class="o">=</span> <span class="n">new_start</span><span class="o">*</span><span class="n">base</span>
		<span class="n">new_base</span> <span class="o">=</span> <span class="n">new_base</span><span class="o">*</span><span class="n">base</span>
			  
	<span class="c1">#print(result)</span>
	<span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">result</span><span class="p">:</span>    
		<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">item</span><span class="p">:</span>
			<span class="k">yield</span> <span class="n">i</span></div>
<span class="c1">##list(exp_range(0,100000,10))</span>


<div class="viewcode-block" id="extract_data"><a class="viewcode-back" href="../code.html#alefuncs.extract_data">[docs]</a><span class="k">def</span> <span class="nf">extract_data</span><span class="p">(</span><span class="n">infile</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="mi">3</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">5</span><span class="p">],</span> <span class="n">header</span><span class="o">=</span><span class="s1">&#39;##&#39;</span><span class="p">,</span> <span class="n">skip_lines_starting_with</span><span class="o">=</span><span class="s1">&#39;#&#39;</span><span class="p">,</span> <span class="n">data_separator</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span> <span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	Extract data from a file. Returns a list of tuples. </span>
<span class="sd">	Each tuple contains the data extracted from one line of the file</span>
<span class="sd">	in the indicated columns and with the indicated order.</span>
<span class="sd">	&#39;&#39;&#39;</span>
	
	<span class="n">extracted_data</span> <span class="o">=</span> <span class="p">[]</span>
	<span class="n">header_list</span> <span class="o">=</span> <span class="p">[]</span>
	<span class="n">header_flag</span> <span class="o">=</span> <span class="mi">0</span>
	<span class="n">line_counter</span> <span class="o">=</span> <span class="mi">0</span>

	<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">infile</span><span class="p">)</span> <span class="k">as</span> <span class="n">infile</span><span class="p">:</span>
		<span class="n">lines</span> <span class="o">=</span> <span class="n">infile</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>

	<span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">lines</span><span class="p">:</span> <span class="c1"># yield_file(infile) can be used instead</span>
		<span class="n">line_counter</span> <span class="o">+=</span> <span class="mi">1</span>

		<span class="k">if</span> <span class="n">line</span><span class="p">[:</span><span class="nb">len</span><span class="p">(</span><span class="n">header</span><span class="p">)]</span> <span class="o">==</span> <span class="n">header</span><span class="p">:</span> <span class="c1"># checks the header</span>
			<span class="n">header_list</span> <span class="o">=</span> <span class="n">line_to_list</span><span class="p">(</span><span class="n">line</span><span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="n">header</span><span class="p">):],</span> <span class="n">data_separator</span><span class="p">)</span>
			<span class="n">header_flag</span> <span class="o">+=</span> <span class="mi">1</span>
			<span class="k">if</span> <span class="n">header_flag</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
				<span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;More than one line seems to contain the header identificator &quot;&#39;</span> <span class="o">+</span> <span class="n">header</span> <span class="o">+</span> <span class="s1">&#39;&quot;.&#39;</span><span class="p">)</span>
		<span class="k">elif</span> <span class="n">line</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="n">skip_lines_starting_with</span> <span class="ow">or</span> <span class="n">line</span> <span class="o">==</span> <span class="s1">&#39;&#39;</span> <span class="ow">or</span> <span class="n">line</span> <span class="o">==</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">:</span> <span class="c1"># skips comments and blank lines</span>
			<span class="k">pass</span>
		<span class="k">else</span><span class="p">:</span>
			<span class="n">list_</span> <span class="o">=</span> <span class="n">line_to_list</span><span class="p">(</span><span class="n">line</span><span class="p">,</span> <span class="n">data_separator</span><span class="p">)</span>
			<span class="n">reduced_list</span><span class="o">=</span><span class="p">[]</span>
			<span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">columns</span><span class="p">:</span>
				<span class="n">reduced_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">list_</span><span class="p">[</span><span class="n">item</span><span class="p">])</span>
			<span class="n">extracted_data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">tuple</span><span class="p">(</span><span class="n">reduced_list</span><span class="p">))</span>

	<span class="k">if</span> <span class="n">verbose</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span> <span class="c1"># Prints out a brief summary</span>
		<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Data extracted from&#39;</span><span class="p">,</span> <span class="n">infile</span><span class="p">)</span>
		<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Header =&#39;</span><span class="p">,</span> <span class="n">header_list</span><span class="p">)</span>
		<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Total lines =&#39;</span><span class="p">,</span> <span class="n">line_counter</span><span class="p">)</span>

	<span class="k">return</span> <span class="n">extracted_data</span></div>
<span class="c1"># extract_data(&#39;tables/clinvarCnv.txt&#39;, columns=[3,0,1,2,5], header=&#39;##&#39;, skip_lines_starting_with=&#39;#&#39;, data_separator=&#39;\t&#39;, verbose=True)</span>


<div class="viewcode-block" id="extract_Toronto"><a class="viewcode-back" href="../code.html#alefuncs.extract_Toronto">[docs]</a><span class="k">def</span> <span class="nf">extract_Toronto</span><span class="p">(</span><span class="n">infile</span><span class="p">,</span> <span class="n">outfile</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	Ad hoc function to extract deletions and duplications out of the Toronto Genetic Variants Database.</span>
<span class="sd">	Returns a file ready to be annotated with FusionGenes_Annotation.pl .</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="c1"># Extract data from infile</span>
	<span class="c1"># Columns are: ID, Chr, Start, End, CNV_Type</span>
	<span class="n">raw_data</span> <span class="o">=</span> <span class="n">extract_data</span><span class="p">(</span><span class="n">infile</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">5</span><span class="p">],</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span> <span class="p">)</span>

	<span class="c1"># Take only deletions and duplications</span>
	<span class="n">filtered_data</span> <span class="o">=</span> <span class="p">[]</span>
	<span class="k">for</span> <span class="n">data</span> <span class="ow">in</span> <span class="n">raw_data</span><span class="p">:</span>
		<span class="k">if</span> <span class="s2">&quot;deletion&quot;</span> <span class="ow">in</span> <span class="n">data</span> <span class="ow">or</span> <span class="s1">&#39;duplication&#39;</span> <span class="ow">in</span> <span class="n">data</span><span class="p">:</span>
			<span class="n">filtered_data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
	<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;len(row_data)      :&#39;</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">raw_data</span><span class="p">))</span>
	<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;len(filtered_data) :&#39;</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">filtered_data</span><span class="p">))</span>


	<span class="c1"># Write filtered_data to a text file</span>
	<span class="n">header</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;##ID&#39;</span><span class="p">,</span><span class="s1">&#39;ChrA&#39;</span><span class="p">,</span><span class="s1">&#39;StartA&#39;</span><span class="p">,</span><span class="s1">&#39;EndA&#39;</span><span class="p">,</span><span class="s1">&#39;ChrB&#39;</span><span class="p">,</span><span class="s1">&#39;StartB&#39;</span><span class="p">,</span><span class="s1">&#39;EndB&#39;</span><span class="p">,</span><span class="s1">&#39;CnvType&#39;</span><span class="p">,</span><span class="s1">&#39;Orientation&#39;</span><span class="p">]</span>
	<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">outfile</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">outfile</span><span class="p">:</span>
		<span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">list_to_line</span><span class="p">(</span><span class="n">header</span><span class="p">,</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
		<span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">filtered_data</span><span class="p">:</span>
			<span class="k">if</span> <span class="n">item</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;duplication&#39;</span><span class="p">:</span>
				<span class="n">orientation</span> <span class="o">=</span> <span class="s1">&#39;HT&#39;</span>
			<span class="k">elif</span> <span class="n">item</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;deletion&#39;</span><span class="p">:</span>
				<span class="n">orientation</span> <span class="o">=</span> <span class="s1">&#39;TH&#39;</span>
			<span class="k">else</span><span class="p">:</span>
				<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;ERROR: unable to determine &quot;Orientation&quot;...&#39;</span><span class="p">)</span>
			<span class="n">list_</span> <span class="o">=</span> <span class="p">[</span><span class="n">item</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="n">item</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span><span class="n">item</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span><span class="n">item</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span><span class="n">item</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span><span class="n">item</span><span class="p">[</span><span class="mi">3</span><span class="p">],</span><span class="n">item</span><span class="p">[</span><span class="mi">3</span><span class="p">],</span><span class="n">item</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">(),</span><span class="n">orientation</span><span class="p">]</span>
			<span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">list_to_line</span><span class="p">(</span><span class="n">list_</span><span class="p">,</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
	<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Done&#39;</span><span class="p">)</span></div>
<span class="c1"># infile = &#39;/home/amarcozz/Documents/Projects/Fusion Genes/Scripts/test_datasets/public/GRCh37_hg19_variants_2014-10-16.txt&#39;</span>
<span class="c1"># outfile = infile[:-4]+&#39;_DelDupOnly.txt&#39;</span>
<span class="c1"># extract_Toronto(infile, outfile)</span>


<div class="viewcode-block" id="extract_Decipher"><a class="viewcode-back" href="../code.html#alefuncs.extract_Decipher">[docs]</a><span class="k">def</span> <span class="nf">extract_Decipher</span><span class="p">(</span><span class="n">infile</span><span class="p">,</span> <span class="n">outfile</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	Ad hoc function to extract deletions and duplications out of the Decipher Database.</span>
<span class="sd">	Returns a file ready to be annotated with FusionGenes_Annotation.pl .</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="c1"># Extract data from infile</span>
	<span class="c1"># Columns are: ID, Chr, Start, End, CNV_Type(here expressed as &quot;mean_ratio&quot;)</span>
	<span class="n">raw_data</span> <span class="o">=</span> <span class="n">extract_data</span><span class="p">(</span><span class="n">infile</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span> <span class="p">)</span>
	<span class="n">header</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;##ID&#39;</span><span class="p">,</span><span class="s1">&#39;ChrA&#39;</span><span class="p">,</span><span class="s1">&#39;StartA&#39;</span><span class="p">,</span><span class="s1">&#39;EndA&#39;</span><span class="p">,</span><span class="s1">&#39;ChrB&#39;</span><span class="p">,</span><span class="s1">&#39;StartB&#39;</span><span class="p">,</span><span class="s1">&#39;EndB&#39;</span><span class="p">,</span><span class="s1">&#39;CnvType&#39;</span><span class="p">,</span><span class="s1">&#39;Orientation&#39;</span><span class="p">]</span>
	<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">outfile</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">outfile</span><span class="p">:</span>
		<span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">list_to_line</span><span class="p">(</span><span class="n">header</span><span class="p">,</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
		<span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">raw_data</span><span class="p">:</span>
			<span class="c1"># Convert mean_ratio to CnvType</span>
			<span class="k">if</span> <span class="nb">float</span><span class="p">(</span><span class="n">item</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
				<span class="n">CnvType</span> <span class="o">=</span> <span class="s1">&#39;DUPLICATION&#39;</span>
				<span class="n">orientation</span> <span class="o">=</span> <span class="s1">&#39;HT&#39;</span>
			<span class="k">elif</span> <span class="nb">float</span><span class="p">(</span><span class="n">item</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
				<span class="n">CnvType</span> <span class="o">=</span> <span class="s1">&#39;DELETION&#39;</span>
				<span class="n">orientation</span> <span class="o">=</span> <span class="s1">&#39;TH&#39;</span>
			<span class="k">else</span><span class="p">:</span>
				<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;ERROR: unable to determine &quot;Orientation&quot;...&#39;</span><span class="p">)</span>
			<span class="c1"># Write output</span>
			<span class="n">list_</span> <span class="o">=</span> <span class="p">[</span><span class="n">item</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="n">item</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span><span class="n">item</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span><span class="n">item</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span><span class="n">item</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span><span class="n">item</span><span class="p">[</span><span class="mi">3</span><span class="p">],</span><span class="n">item</span><span class="p">[</span><span class="mi">3</span><span class="p">],</span><span class="n">CnvType</span><span class="p">,</span><span class="n">orientation</span><span class="p">]</span>
			<span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">list_to_line</span><span class="p">(</span><span class="n">list_</span><span class="p">,</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
	<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Done&#39;</span><span class="p">)</span></div>
<span class="c1"># infile = &#39;/home/amarcozz/Documents/Projects/Fusion Genes/Scripts/test_datasets/public/decipher-hg19_15-01-30.txt&#39;</span>
<span class="c1"># outfile = infile[:-4]+&#39;_DelDupOnly.txt&#39;</span>
<span class="c1"># extract_Decipher(infile, outfile)</span>


<div class="viewcode-block" id="extract_dgvMerged"><a class="viewcode-back" href="../code.html#alefuncs.extract_dgvMerged">[docs]</a><span class="k">def</span> <span class="nf">extract_dgvMerged</span><span class="p">(</span><span class="n">infile</span><span class="p">,</span> <span class="n">outfile</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	Ad hoc function to extract deletions and losses out of the dgvMerged database.</span>
<span class="sd">	Returns a file ready to be annotated with FusionGenes_Annotation.pl .</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="c1">#original_header = &#39;##bin   chrom   chromStart  chromEnd    name    score   strand  thickStart  thickEnd    itemRgb varType reference   pubMedId    method  platform    mergedVariants  supportingVariants  sampleSize  observedGains   observedLosses  cohortDescription   genes   samples&#39;</span>
					<span class="c1">#   [0]     [1]     [2]         [3]         [4]     [5]     [6]     [7]         [8]         [9]     [10]    [11]        [12]        [13]    [14]        [15]            [16]                [17]        [18]            [19]            [20]                [21]    [22]</span>
	<span class="n">raw_data</span> <span class="o">=</span> <span class="n">extract_data</span><span class="p">(</span><span class="n">infile</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="mi">4</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">10</span><span class="p">],</span> <span class="n">header</span><span class="o">=</span><span class="s1">&#39;##&#39;</span><span class="p">,</span> <span class="n">skip_lines_starting_with</span><span class="o">=</span><span class="s1">&#39;#&#39;</span><span class="p">,</span> <span class="n">data_separator</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span> <span class="p">)</span>

	<span class="c1"># Take only deletions and losses</span>
	<span class="n">filtered_data</span> <span class="o">=</span> <span class="p">[]</span>
	<span class="k">for</span> <span class="n">data</span> <span class="ow">in</span> <span class="n">raw_data</span><span class="p">:</span>
		<span class="k">if</span> <span class="s2">&quot;Deletion&quot;</span> <span class="ow">in</span> <span class="n">data</span> <span class="ow">or</span> <span class="s1">&#39;Loss&#39;</span> <span class="ow">in</span> <span class="n">data</span><span class="p">:</span>
			<span class="n">filtered_data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
	<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;len(row_data)      :&#39;</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">raw_data</span><span class="p">))</span>
	<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;len(filtered_data) :&#39;</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">filtered_data</span><span class="p">))</span>

	<span class="c1"># Write filtered_data to a text file</span>
	<span class="n">header</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;##ID&#39;</span><span class="p">,</span><span class="s1">&#39;ChrA&#39;</span><span class="p">,</span><span class="s1">&#39;StartA&#39;</span><span class="p">,</span><span class="s1">&#39;EndA&#39;</span><span class="p">,</span><span class="s1">&#39;ChrB&#39;</span><span class="p">,</span><span class="s1">&#39;StartB&#39;</span><span class="p">,</span><span class="s1">&#39;EndB&#39;</span><span class="p">,</span><span class="s1">&#39;CnvType&#39;</span><span class="p">,</span><span class="s1">&#39;Orientation&#39;</span><span class="p">]</span>
	<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">outfile</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">outfile</span><span class="p">:</span>
		<span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">list_to_line</span><span class="p">(</span><span class="n">header</span><span class="p">,</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
		<span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">filtered_data</span><span class="p">:</span>
			<span class="k">if</span> <span class="n">item</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Deletion&#39;</span> <span class="ow">or</span> <span class="n">item</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Loss&#39;</span><span class="p">:</span>
				<span class="n">cnv_type</span> <span class="o">=</span> <span class="s1">&#39;DELETION&#39;</span>
				<span class="n">orientation</span> <span class="o">=</span> <span class="s1">&#39;HT&#39;</span>
			<span class="c1"># elif item[-1] == &#39;deletion&#39;:</span>
			<span class="c1">#   orientation = &#39;TH&#39;</span>
			<span class="k">else</span><span class="p">:</span>
				<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;ERROR: unable to determine &quot;Orientation&quot;...&#39;</span><span class="p">)</span>
			<span class="n">list_</span> <span class="o">=</span> <span class="p">[</span><span class="n">item</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="n">item</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="mi">3</span><span class="p">:],</span><span class="n">item</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span><span class="n">item</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span><span class="n">item</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="mi">3</span><span class="p">:],</span><span class="n">item</span><span class="p">[</span><span class="mi">3</span><span class="p">],</span><span class="n">item</span><span class="p">[</span><span class="mi">3</span><span class="p">],</span><span class="n">cnv_type</span><span class="p">,</span><span class="n">orientation</span><span class="p">]</span>
			<span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">list_to_line</span><span class="p">(</span><span class="n">list_</span><span class="p">,</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
	<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Done&#39;</span><span class="p">)</span></div>
<span class="c1"># ## Extract deletions and Losses from dgvMerged</span>
<span class="c1"># folder = &#39;/home/amarcozz/Documents/Projects/Fusion Genes/Scripts/test_datasets/public/breaks&#39;</span>
<span class="c1"># file_name = &#39;dgvMerged.txt&#39;</span>
<span class="c1"># infile = folder + &#39;/&#39; + file_name</span>
<span class="c1"># outfile = folder + &#39;/&#39; + &#39;dgvMerged-DeletionsOnly.txt&#39;</span>
<span class="c1"># extract_dgvMerged(infile, outfile)</span>
<span class="c1"># ## annotate</span>
<span class="c1"># dataset_file = &#39;/home/amarcozz/Documents/Projects/Fusion Genes/Scripts/test_datasets/public/breaks/dgvMerged-DeletionsOnly.txt&#39;</span>
<span class="c1"># annotate_fusion_genes(dataset_file)</span>


<div class="viewcode-block" id="fill_and_sort"><a class="viewcode-back" href="../code.html#alefuncs.fill_and_sort">[docs]</a><span class="k">def</span> <span class="nf">fill_and_sort</span><span class="p">(</span><span class="n">pandas_chrSeries</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;incomplete pandas.Series =&gt; complete and sorted pandas.Series</span>
<span class="sd">	Given a pandas.Series in which the first argument is the chromosome name</span>
<span class="sd">	and the second argument is a count &quot; [(&#39;1&#39;, 61), (&#39;3&#39;, 28), ..., (&#39;X&#39;, 29)]&quot;</span>
<span class="sd">	This function returns a new (sorted by chromosome) series with the missing chromosome included as (&#39;Chr_name&#39;,0).</span>
<span class="sd">	</span>
<span class="sd">	This is useful when creating series out of subsets grouped by Chr.</span>
<span class="sd">	If the Chr does not contains any event, then it will be excluded from the subset.</span>
<span class="sd">	However, expecially for plotting reasons, you may want to have (&#39;Chr&#39;,0) in you list instead of a missing Chr.</span>
<span class="sd">	</span>
<span class="sd">	Example.</span>
<span class="sd">	&gt; series = [(&#39;1&#39;, 61), (&#39;3&#39;, 28), ..., (&#39;X&#39;, 29)] # in this Series Chr_2 and Chr_Y are missing.</span>
<span class="sd">	&gt; fill_and_sort(series)</span>
<span class="sd">	&gt;&gt;&gt; [(&#39;1&#39;, 61), (&#39;2&#39;,0), (&#39;3&#39;, 28), ..., (&#39;X&#39;, 29), (&#39;Y&#39;,0)] # this Series have all the chromosomes</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="kn">import</span> <span class="nn">pandas</span>
	<span class="c1"># add missing ChrA</span>
	<span class="n">CHROMOSOMES</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">c</span><span class="p">)</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">23</span><span class="p">)]</span><span class="o">+</span><span class="p">[</span><span class="s1">&#39;X&#39;</span><span class="p">,</span><span class="s1">&#39;Y&#39;</span><span class="p">]</span>
	<span class="n">chr_list</span> <span class="o">=</span> <span class="n">CHROMOSOMES</span><span class="p">[:]</span>
	<span class="n">complete_series</span> <span class="o">=</span> <span class="p">[]</span>
	<span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">pandas_chrSeries</span><span class="o">.</span><span class="n">iteritems</span><span class="p">():</span>
		<span class="n">chr_list</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">item</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
		<span class="n">complete_series</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">item</span><span class="p">)</span>
	<span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">chr_list</span><span class="p">:</span>
		<span class="n">complete_series</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">item</span><span class="p">,</span><span class="mi">0</span><span class="p">))</span>
	
	<span class="c1"># sort by chromosome</span>
	<span class="n">sorted_</span> <span class="o">=</span> <span class="p">[]</span>
	<span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">CHROMOSOMES</span><span class="p">:</span>
		<span class="k">for</span> <span class="n">_item</span> <span class="ow">in</span> <span class="n">complete_series</span><span class="p">:</span>
			<span class="k">if</span> <span class="n">_item</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">==</span><span class="n">item</span><span class="p">:</span>
				<span class="n">sorted_</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">_item</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
	<span class="k">return</span> <span class="n">pandas</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">sorted_</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">CHROMOSOMES</span><span class="p">)</span></div>
<span class="c1"># counts = [50,9,45,6]</span>
<span class="c1"># pandas_chrSeries = pandas.Series(counts, index=[&#39;1&#39;,&#39;4&#39;,&#39;X&#39;,&#39;10&#39;])</span>
<span class="c1"># print(pandas_chrSeries)</span>
<span class="c1"># good_series = fill_and_sort(pandas_chrSeries)</span>
<span class="c1"># print(good_series)</span>


<div class="viewcode-block" id="find"><a class="viewcode-back" href="../code.html#alefuncs.find">[docs]</a><span class="k">def</span> <span class="nf">find</span><span class="p">(</span><span class="n">string</span><span class="p">,</span> <span class="n">char</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	Looks for a character in a sctring and returns its index.</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="c1"># Compared to string.find(), it returns ALL the indexes, not only the first one.</span>
	<span class="k">return</span> <span class="p">[</span><span class="n">index</span> <span class="k">for</span> <span class="n">index</span><span class="p">,</span> <span class="n">letter</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">string</span><span class="p">)</span> <span class="k">if</span> <span class="n">letter</span> <span class="o">==</span> <span class="n">char</span><span class="p">]</span></div>
<span class="c1"># print(find(&#39;alessio&#39;, &#39;s&#39;))</span>

<div class="viewcode-block" id="filter_out"><a class="viewcode-back" href="../code.html#alefuncs.filter_out">[docs]</a><span class="k">def</span> <span class="nf">filter_out</span><span class="p">(</span><span class="n">word</span><span class="p">,</span> <span class="n">infile</span><span class="p">,</span> <span class="n">outfile</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	Reads a file line by line</span>
<span class="sd">	and writes an output file containing only</span>
<span class="sd">	the lines that DO NOT contains &#39;word&#39;.</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Filtering out lines containing&#39;</span><span class="p">,</span><span class="n">word</span><span class="p">,</span><span class="s1">&#39;...&#39;</span><span class="p">)</span>
	<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">infile</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">infile</span><span class="p">:</span>
		<span class="n">lines</span> <span class="o">=</span> <span class="n">infile</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
	<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">outfile</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">outfile</span><span class="p">:</span>
		<span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">lines</span><span class="p">:</span> <span class="c1"># yield_file(infile) can be used instead</span>
			<span class="k">if</span> <span class="n">word</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">line</span><span class="p">:</span>
				<span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
	<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Done&#39;</span><span class="p">)</span></div>
<span class="c1"># infile = &#39;/home/amarcozz/Documents/Projects/Fusion Genes/Scripts/test_datasets/public/breaks/Decipher_DelDupOnly.txt&#39;</span>
<span class="c1"># outfile = &#39;/home/amarcozz/Documents/Projects/Fusion Genes/Scripts/test_datasets/public/breaks/Decipher-DeletionsOnly.txt&#39;</span>
<span class="c1"># filter_out(&#39;DUPLICATION&#39;,infile, outfile)</span>

<div class="viewcode-block" id="flatten2"><a class="viewcode-back" href="../code.html#alefuncs.flatten2">[docs]</a><span class="k">def</span> <span class="nf">flatten2</span><span class="p">(</span><span class="n">l</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	Flat an irregular iterable to a list.</span>
<span class="sd">	Python &gt;= 2.6 version.</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">l</span><span class="p">:</span>
		<span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">item</span><span class="p">,</span> <span class="n">collections</span><span class="o">.</span><span class="n">Iterable</span><span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">item</span><span class="p">,</span> <span class="n">basestring</span><span class="p">):</span>
			<span class="k">for</span> <span class="n">sub</span> <span class="ow">in</span> <span class="n">flatten</span><span class="p">(</span><span class="n">item</span><span class="p">):</span>
				<span class="k">yield</span> <span class="n">sub</span>
		<span class="k">else</span><span class="p">:</span>
			<span class="k">yield</span> <span class="n">item</span></div>


<div class="viewcode-block" id="flatten"><a class="viewcode-back" href="../code.html#alefuncs.flatten">[docs]</a><span class="k">def</span> <span class="nf">flatten</span><span class="p">(</span><span class="n">l</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	Flat an irregular iterable to a list.</span>
<span class="sd">	Python &gt;= 3.3 version.</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">l</span><span class="p">:</span>
		<span class="k">try</span><span class="p">:</span>
			<span class="k">yield from</span> <span class="n">flatten</span><span class="p">(</span><span class="n">item</span><span class="p">)</span>
		<span class="k">except</span> <span class="ne">TypeError</span><span class="p">:</span>
			<span class="k">yield</span> <span class="n">item</span></div>

<div class="viewcode-block" id="gene_synonyms"><a class="viewcode-back" href="../code.html#alefuncs.gene_synonyms">[docs]</a><span class="k">def</span> <span class="nf">gene_synonyms</span><span class="p">(</span><span class="n">gene_name</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;str =&gt; list()</span>
<span class="sd">	Queries http://rest.genenames.org and http://www.ncbi.nlm.nih.gov/ to figure out the best synonym of gene_name.</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="kn">import</span> <span class="nn">httplib2</span> <span class="k">as</span> <span class="nn">http</span>
	<span class="kn">import</span> <span class="nn">json</span>
	<span class="kn">from</span> <span class="nn">urllib.parse</span> <span class="k">import</span> <span class="n">urlparse</span>
	<span class="kn">from</span> <span class="nn">urllib.request</span> <span class="k">import</span> <span class="n">urlopen</span>
	<span class="kn">from</span> <span class="nn">bs4</span> <span class="k">import</span> <span class="n">BeautifulSoup</span>
	

	<span class="n">result</span> <span class="o">=</span> <span class="p">[]</span>
	<span class="n">tmp</span> <span class="o">=</span> <span class="p">[]</span>
	<span class="n">headers</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;Accept&#39;</span><span class="p">:</span> <span class="s1">&#39;application/json&#39;</span><span class="p">}</span>

	<span class="n">uri</span> <span class="o">=</span> <span class="s1">&#39;http://rest.genenames.org&#39;</span>
	<span class="n">path</span> <span class="o">=</span> <span class="s1">&#39;/search/</span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">gene_name</span><span class="p">)</span>
	<span class="n">html_doc</span> <span class="o">=</span> <span class="n">urlopen</span><span class="p">(</span><span class="s1">&#39;http://www.ncbi.nlm.nih.gov/gene/?term=</span><span class="si">{}</span><span class="s1">[sym]&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">gene_name</span><span class="p">))</span>
	<span class="n">html_txt</span> <span class="o">=</span> <span class="n">BeautifulSoup</span><span class="p">(</span><span class="n">html_doc</span><span class="p">,</span> <span class="s1">&#39;html.parser&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">get_text</span><span class="p">()</span>


	<span class="n">target</span> <span class="o">=</span> <span class="n">urlparse</span><span class="p">(</span><span class="n">uri</span><span class="o">+</span><span class="n">path</span><span class="p">)</span>
	<span class="n">method</span> <span class="o">=</span> <span class="s1">&#39;GET&#39;</span>
	<span class="n">body</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>

	<span class="n">h</span> <span class="o">=</span> <span class="n">http</span><span class="o">.</span><span class="n">Http</span><span class="p">()</span>

	<span class="n">response</span><span class="p">,</span> <span class="n">content</span> <span class="o">=</span> <span class="n">h</span><span class="o">.</span><span class="n">request</span><span class="p">(</span>
									<span class="n">target</span><span class="o">.</span><span class="n">geturl</span><span class="p">(),</span>
									<span class="n">method</span><span class="p">,</span>
									<span class="n">body</span><span class="p">,</span>
									<span class="n">headers</span> <span class="p">)</span>

	<span class="k">if</span> <span class="n">response</span><span class="p">[</span><span class="s1">&#39;status&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;200&#39;</span><span class="p">:</span>
		<span class="c1"># assume that content is a json reply</span>
		<span class="c1"># parse content with the json module </span>
		<span class="n">data</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">loads</span><span class="p">(</span><span class="n">content</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s1">&#39;utf8&#39;</span><span class="p">))</span>
		<span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;response&#39;</span><span class="p">][</span><span class="s1">&#39;docs&#39;</span><span class="p">]:</span>
			<span class="n">tmp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">item</span><span class="p">[</span><span class="s1">&#39;symbol&#39;</span><span class="p">])</span>
	 
	<span class="k">else</span><span class="p">:</span>
		<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Error detected: &#39;</span> <span class="o">+</span> <span class="n">response</span><span class="p">[</span><span class="s1">&#39;status&#39;</span><span class="p">])</span>
		<span class="k">return</span> <span class="kc">None</span>

	<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">tmp</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
		<span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">tmp</span><span class="p">:</span>
			<span class="k">if</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">html_txt</span><span class="p">:</span>
				<span class="n">result</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
		<span class="k">return</span> <span class="n">result</span>
	<span class="k">else</span><span class="p">:</span>
		<span class="k">return</span> <span class="n">tmp</span></div>
<span class="c1">#print(gene_synonyms(&#39;MLL3&#39;))</span>

<span class="k">def</span> <span class="nf">gen_controls</span><span class="p">(</span><span class="n">how_many</span><span class="p">,</span><span class="n">chromosome</span><span class="p">,</span><span class="n">GapTable_file</span><span class="p">,</span><span class="n">outfile</span><span class="p">):</span>
	<span class="k">global</span> <span class="n">running_threads</span> <span class="c1"># in case of multithreading</span>
	<span class="n">list_brkps</span> <span class="o">=</span> <span class="n">gen_rnd_single_break</span><span class="p">(</span><span class="n">how_many</span><span class="p">,</span> <span class="n">chromosome</span><span class="p">,</span> <span class="n">GapTable_file</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
	<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">outfile</span><span class="p">,</span><span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
		<span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">list_brkps</span><span class="p">:</span>
			<span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">list_to_line</span><span class="p">(</span><span class="n">item</span><span class="p">,</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
	<span class="n">running_threads</span> <span class="o">-=</span> <span class="mi">1</span> <span class="c1"># in case of multithreading</span>
<span class="c1"># # Generate controls</span>
<span class="c1"># import time</span>
<span class="c1"># from threading import Thread</span>
<span class="c1"># threads = 0</span>
<span class="c1"># running_threads = 0</span>
<span class="c1"># max_simultaneous_threads = 20</span>
<span class="c1"># how_many=9045</span>
<span class="c1"># chromosome=&#39;9&#39;</span>
<span class="c1"># GapTable_file=&#39;/Users/alec/Desktop/UMCU_Backup/Projects/Anne_Project/current_brkps_DB/out_ALL_gap.txt&#39;</span>
<span class="c1"># while threads &lt; 100:</span>
<span class="c1">#   while running_threads &gt;= max_simultaneous_threads:</span>
<span class="c1">#       time.sleep(1)</span>
<span class="c1">#   running_threads += 1</span>
<span class="c1">#   outfile = &#39;/Users/alec/Desktop/UMCU_Backup/Projects/Anne_Project/current_brkps_DB/out_chr9_control_&#39;+str(threads)+&#39;.txt&#39;</span>
<span class="c1">#   print(&#39;thread&#39;, threads, &#39;|&#39;, &#39;running threads:&#39;,running_threads)</span>
<span class="c1">#   Thread(target=gen_controls, args=(how_many,chromosome,GapTable_file,outfile)).start()</span>
<span class="c1">#   threads += 1</span>

<div class="viewcode-block" id="gen_control_dataset"><a class="viewcode-back" href="../code.html#alefuncs.gen_control_dataset">[docs]</a><span class="k">def</span> <span class="nf">gen_control_dataset</span><span class="p">(</span><span class="n">real_dataset</span><span class="p">,</span><span class="n">suffix</span><span class="o">=</span><span class="s1">&#39;_control.txt&#39;</span><span class="p">):</span><span class="c1"># tested only for deletion/duplication</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	Generates a control dataset ad hoc.</span>
<span class="sd">	Takes as input an existing dataset and generates breaks</span>
<span class="sd">	in the same chromosomes and with the same distance (+-1bp),</span>
<span class="sd">	the position are however randomized.</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="n">real_data</span> <span class="o">=</span> <span class="n">extract_data</span><span class="p">(</span><span class="n">real_dataset</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">4</span><span class="p">,</span><span class="mi">5</span><span class="p">,</span><span class="mi">7</span><span class="p">,</span><span class="mi">8</span><span class="p">],</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
	<span class="n">control_data</span> <span class="o">=</span> <span class="p">[]</span>
	<span class="n">_id_list</span> <span class="o">=</span> <span class="p">[]</span>
	<span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">real_data</span><span class="p">:</span>
		<span class="k">if</span> <span class="n">item</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="n">item</span><span class="p">[</span><span class="mi">2</span><span class="p">]:</span> <span class="c1"># ChrA == ChrB</span>
			
			<span class="c1"># generate a unique id</span>
			<span class="n">_id</span> <span class="o">=</span> <span class="n">gen_rnd_id</span><span class="p">(</span><span class="mi">16</span><span class="p">)</span>
			<span class="k">while</span> <span class="n">_id</span> <span class="ow">in</span> <span class="n">_id_list</span><span class="p">:</span>
					<span class="n">_id</span> <span class="o">=</span> <span class="n">gen_rnd_id</span><span class="p">(</span><span class="mi">16</span><span class="p">)</span>
			<span class="n">_id_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">_id</span><span class="p">)</span>

			<span class="n">chromosome</span> <span class="o">=</span> <span class="n">item</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
			<span class="n">distance</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">item</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span><span class="o">-</span><span class="nb">int</span><span class="p">(</span><span class="n">item</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="c1"># </span>
			<span class="n">cnv_type</span> <span class="o">=</span> <span class="n">item</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span>
			<span class="n">orientation</span> <span class="o">=</span> <span class="n">item</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span>
			<span class="n">breaks</span> <span class="o">=</span> <span class="n">gen_rnd_breaks</span><span class="p">(</span><span class="n">how_many</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">chromosome</span><span class="o">=</span><span class="n">chromosome</span><span class="p">,</span>
									<span class="n">min_distance</span><span class="o">=</span><span class="n">distance</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="n">max_distance</span><span class="o">=</span><span class="n">distance</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span>
									<span class="n">GapTable_file</span><span class="o">=</span><span class="s1">&#39;tables/gap.txt&#39;</span><span class="p">)</span>
			<span class="nb">print</span><span class="p">(</span><span class="n">breaks</span><span class="p">)</span>
			<span class="n">control_data</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">_id</span><span class="p">,</span><span class="n">chromosome</span><span class="p">,</span><span class="n">breaks</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">],</span><span class="n">breaks</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">],</span><span class="n">chromosome</span><span class="p">,</span><span class="n">breaks</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">2</span><span class="p">],</span>
								 <span class="n">breaks</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">2</span><span class="p">],</span><span class="n">cnv_type</span><span class="p">,</span><span class="n">orientation</span><span class="p">])</span>
		<span class="k">else</span><span class="p">:</span>
			<span class="nb">print</span><span class="p">(</span><span class="n">item</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="s1">&#39;is no equal to&#39;</span><span class="p">,</span><span class="n">item</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span><span class="s1">&#39;I am skipping these breaks&#39;</span><span class="p">)</span>
	
	<span class="n">header</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;##ID&#39;</span><span class="p">,</span> <span class="s1">&#39;ChrA&#39;</span><span class="p">,</span> <span class="s1">&#39;StartA&#39;</span><span class="p">,</span> <span class="s1">&#39;EndA&#39;</span><span class="p">,</span> <span class="s1">&#39;ChrB&#39;</span><span class="p">,</span> <span class="s1">&#39;StartB&#39;</span><span class="p">,</span> <span class="s1">&#39;EndB&#39;</span><span class="p">,</span> <span class="s1">&#39;CnvType&#39;</span><span class="p">,</span> <span class="s1">&#39;Orientation&#39;</span><span class="p">]</span>
	
	<span class="n">filename</span> <span class="o">=</span> <span class="n">real_dataset</span><span class="p">[:</span><span class="o">-</span><span class="mi">4</span><span class="p">]</span><span class="o">+</span> <span class="n">suffix</span>
	<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span><span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">outfile</span><span class="p">:</span>
		<span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">list_to_line</span><span class="p">(</span><span class="n">header</span><span class="p">,</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
		<span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">control_data</span><span class="p">:</span>
			<span class="n">line</span> <span class="o">=</span> <span class="n">list_to_line</span><span class="p">(</span><span class="n">item</span><span class="p">,</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
			<span class="nb">print</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
			<span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">line</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>

	<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Data written in&#39;</span><span class="p">,</span><span class="n">filename</span><span class="p">)</span></div>
<span class="c1"># gen_control_dataset(&#39;/home/amarcozz/Documents/Projects/Fusion Genes/Scripts/test_datasets/raw/clinvarCnv-DeletionsOnly.txt&#39;)</span>

<div class="viewcode-block" id="gen_gap_table"><a class="viewcode-back" href="../code.html#alefuncs.gen_gap_table">[docs]</a><span class="k">def</span> <span class="nf">gen_gap_table</span><span class="p">(</span><span class="n">infile</span><span class="o">=</span><span class="s1">&#39;/Users/amarcozzi/Desktop/All_breakpoints_HG19_final.txt&#39;</span><span class="p">,</span> <span class="n">outfile</span><span class="o">=</span><span class="s1">&#39;/Users/amarcozzi/Desktop/All_breakpoints_HG19_gap.txt&#39;</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mi">10000</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	Generates a file containing a list of coordinates </span>
<span class="sd">	for wich no brakpoints have been found in the input file.</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="c1"># Global constants</span>
	<span class="n">CHROMOSOMES</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">c</span><span class="p">)</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">23</span><span class="p">)]</span><span class="o">+</span><span class="p">[</span><span class="s1">&#39;X&#39;</span><span class="p">,</span><span class="s1">&#39;Y&#39;</span><span class="p">]</span>
	<span class="c1"># length of chromosomes based on GRCh37 (Data source: Ensembl genome browser release 68, July 2012)</span>
	<span class="c1"># http://jul2012.archive.ensembl.org/Homo_sapiens/Location/Chromosome?r=1:1-1000000</span>
	<span class="c1"># http://grch37.ensembl.org/Homo_sapiens/Location/Chromosome?r=1:24626643-24726643</span>
	<span class="n">CHR_LENGTHS</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;1&#39;</span><span class="p">:</span><span class="mi">249250621</span><span class="p">,</span><span class="s1">&#39;2&#39;</span> <span class="p">:</span><span class="mi">243199373</span><span class="p">,</span><span class="s1">&#39;3&#39;</span> <span class="p">:</span><span class="mi">198022430</span><span class="p">,</span><span class="s1">&#39;4&#39;</span> <span class="p">:</span><span class="mi">191154276</span><span class="p">,</span>
				   <span class="s1">&#39;5&#39;</span> <span class="p">:</span><span class="mi">180915260</span><span class="p">,</span><span class="s1">&#39;6&#39;</span> <span class="p">:</span><span class="mi">171115067</span><span class="p">,</span><span class="s1">&#39;7&#39;</span> <span class="p">:</span><span class="mi">159138663</span><span class="p">,</span><span class="s1">&#39;8&#39;</span> <span class="p">:</span><span class="mi">146364022</span><span class="p">,</span>
				   <span class="s1">&#39;9&#39;</span> <span class="p">:</span><span class="mi">141213431</span><span class="p">,</span><span class="s1">&#39;10&#39;</span><span class="p">:</span><span class="mi">135534747</span><span class="p">,</span><span class="s1">&#39;11&#39;</span><span class="p">:</span><span class="mi">135006516</span><span class="p">,</span><span class="s1">&#39;12&#39;</span><span class="p">:</span><span class="mi">133851895</span><span class="p">,</span>
				   <span class="s1">&#39;13&#39;</span><span class="p">:</span><span class="mi">115169878</span><span class="p">,</span><span class="s1">&#39;14&#39;</span><span class="p">:</span><span class="mi">107349540</span><span class="p">,</span><span class="s1">&#39;15&#39;</span><span class="p">:</span><span class="mi">102531392</span><span class="p">,</span><span class="s1">&#39;16&#39;</span><span class="p">:</span><span class="mi">90354753</span><span class="p">,</span>
				   <span class="s1">&#39;17&#39;</span><span class="p">:</span><span class="mi">81195210</span><span class="p">,</span><span class="s1">&#39;18&#39;</span><span class="p">:</span><span class="mi">78077248</span><span class="p">,</span><span class="s1">&#39;19&#39;</span><span class="p">:</span><span class="mi">59128983</span><span class="p">,</span><span class="s1">&#39;20&#39;</span><span class="p">:</span><span class="mi">63025520</span><span class="p">,</span>
				   <span class="s1">&#39;21&#39;</span><span class="p">:</span><span class="mi">48129895</span><span class="p">,</span><span class="s1">&#39;22&#39;</span><span class="p">:</span><span class="mi">51304566</span><span class="p">,</span><span class="s1">&#39;X&#39;</span> <span class="p">:</span><span class="mi">155270560</span><span class="p">,</span><span class="s1">&#39;Y&#39;</span> <span class="p">:</span><span class="mi">59373566</span><span class="p">}</span>
	<span class="n">gap_list</span> <span class="o">=</span> <span class="p">[]</span>
	<span class="k">for</span> <span class="n">Chr</span> <span class="ow">in</span> <span class="n">CHROMOSOMES</span><span class="p">:</span>
		<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;-----------------------------------------------------&#39;</span><span class="p">)</span>
		<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Analyzing breakpoints in chromosome&#39;</span><span class="p">,</span><span class="n">Chr</span><span class="p">)</span>
		<span class="n">length</span> <span class="o">=</span> <span class="n">CHR_LENGTHS</span><span class="p">[</span><span class="n">Chr</span><span class="p">]</span>
		<span class="c1"># determine the intervals given the chromosome length and the resolution</span>
		<span class="n">x_ax</span> <span class="o">=</span> <span class="p">[]</span> <span class="c1"># data holder</span>
		<span class="n">y_ax</span> <span class="o">=</span> <span class="p">[]</span> <span class="c1"># stores breakpoint counts per inteval</span>
		<span class="n">breakpoint_list</span> <span class="o">=</span> <span class="p">[]</span>
		
		<span class="c1"># # Extract data from infile, chromosome by chromosome</span>
		<span class="c1"># with open(infile, &#39;r&#39;) as f:</span>
		<span class="c1">#   lines = f.readlines()</span>
		<span class="c1">#   for line in lines: # yield_file(infile) can be used instead</span>
		<span class="c1">#       if line.startswith(&#39;chr&#39;+Chr+&#39;:&#39;):</span>
		<span class="c1">#           tmp = line.split(&#39;:&#39;)</span>
		<span class="c1">#           breakpoint = tmp[1].split(&#39;-&#39;)[0]</span>
		<span class="c1">#           breakpoint_list.append(int(breakpoint))</span>
		<span class="c1"># print(len(breakpoint_list),&#39;breakpoints found...&#39;)</span>

		<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">infile</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
			<span class="c1">#lines = f.readlines()</span>
			<span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">f</span><span class="p">:</span><span class="c1">#lines: # yield_file(infile) can be used instead</span>
				<span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="n">Chr</span><span class="o">+</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">):</span>
					<span class="n">tmp</span> <span class="o">=</span> <span class="n">line_to_list</span><span class="p">(</span><span class="n">line</span><span class="p">,</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
					<span class="n">breakpoint</span> <span class="o">=</span> <span class="n">tmp</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
					<span class="n">breakpoint_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">breakpoint</span><span class="p">))</span>
		<span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">breakpoint_list</span><span class="p">),</span><span class="s1">&#39;breakpoints found...&#39;</span><span class="p">)</span>

		<span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">resolution</span><span class="p">,</span><span class="n">length</span><span class="o">+</span><span class="n">resolution</span><span class="p">,</span><span class="n">resolution</span><span class="p">):</span>
			<span class="n">x_ax</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">item</span><span class="p">)</span>
		<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Interval list:&#39;</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">x_ax</span><span class="p">),</span> <span class="s1">&#39;at&#39;</span><span class="p">,</span><span class="n">resolution</span><span class="p">,</span><span class="s1">&#39;bases resolution&#39;</span><span class="p">)</span>

		<span class="k">for</span> <span class="n">interval</span> <span class="ow">in</span> <span class="n">x_ax</span><span class="p">:</span>
			<span class="n">count</span> <span class="o">=</span> <span class="mi">0</span>
			<span class="n">to_remove</span> <span class="o">=</span> <span class="p">[]</span>
			<span class="k">for</span> <span class="n">breakpoint</span> <span class="ow">in</span> <span class="n">breakpoint_list</span><span class="p">:</span>
				<span class="k">if</span> <span class="n">breakpoint</span> <span class="o">&lt;=</span> <span class="n">interval</span><span class="p">:</span>
					<span class="n">count</span> <span class="o">+=</span> <span class="mi">1</span>
					<span class="n">to_remove</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">breakpoint</span><span class="p">)</span>
			<span class="n">y_ax</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">count</span><span class="p">)</span>

			<span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">to_remove</span><span class="p">:</span>
				<span class="k">try</span><span class="p">:</span>
					<span class="n">breakpoint_list</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">item</span><span class="p">)</span>
				<span class="k">except</span><span class="p">:</span>
					<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Error&#39;</span><span class="p">,</span><span class="n">item</span><span class="p">)</span>

		<span class="n">counter</span> <span class="o">=</span> <span class="mi">0</span>
		<span class="k">for</span> <span class="n">idx</span><span class="p">,</span><span class="n">count_</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">y_ax</span><span class="p">):</span>
			<span class="k">if</span> <span class="n">count_</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
				<span class="n">gap</span> <span class="o">=</span> <span class="n">x_ax</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span>
				<span class="n">gap_list</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">Chr</span><span class="p">,</span><span class="n">gap</span><span class="p">))</span>
				<span class="n">counter</span> <span class="o">+=</span> <span class="mi">1</span>
		<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Found&#39;</span><span class="p">,</span> <span class="n">counter</span><span class="p">,</span><span class="s1">&#39;gaps in chromosome&#39;</span><span class="p">,</span><span class="n">Chr</span><span class="p">,</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>

	<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">outfile</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
		<span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;#Gap table at &#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">resolution</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39; bases resolution based on &#39;</span><span class="o">+</span><span class="n">infile</span><span class="o">+</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
		<span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;##chrom&#39;</span><span class="o">+</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="o">+</span><span class="s1">&#39;chromStart&#39;</span><span class="o">+</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="o">+</span><span class="s1">&#39;chromEnd&#39;</span><span class="o">+</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>

		<span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">gap_list</span><span class="p">:</span>
			<span class="n">line</span> <span class="o">=</span> <span class="s1">&#39;chr&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">item</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span><span class="o">+</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">item</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">-</span><span class="n">resolution</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">item</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
			<span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">line</span><span class="o">+</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span></div>
<span class="c1"># import time</span>
<span class="c1"># start = time.time()</span>
<span class="c1"># gen_gap_table()</span>
<span class="c1"># print(&#39;Done in&#39;,time.time()-start,&#39;seconds&#39;)</span>
<span class="c1">## Generate a gap table file</span>
<span class="c1"># import time</span>
<span class="c1"># start = time.time()</span>
<span class="c1"># gen_gap_table(infile=&#39;/Users/amarcozzi/Desktop/current_brkps_DB/out_ALL.txt&#39;, outfile=&#39;/Users/amarcozzi/Desktop/current_brkps_DB/out_ALL_gap.txt&#39;, resolution=10000)</span>
<span class="c1"># print(&#39;Done in&#39;,time.time()-start,&#39;seconds&#39;)</span>

<div class="viewcode-block" id="gen_multiple_controls"><a class="viewcode-back" href="../code.html#alefuncs.gen_multiple_controls">[docs]</a><span class="k">def</span> <span class="nf">gen_multiple_controls</span><span class="p">(</span><span class="n">real_dataset</span><span class="p">,</span><span class="n">how_many</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	Generates how_many control datasets.</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="n">n</span><span class="o">=</span><span class="mi">0</span>
	<span class="k">while</span> <span class="n">n</span> <span class="o">&lt;</span> <span class="n">how_many</span><span class="p">:</span>
		<span class="n">suffix</span> <span class="o">=</span> <span class="s1">&#39;_control_&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">n</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;.txt&#39;</span>
		<span class="c1">#real_dataset = &#39;/home/amarcozz/Documents/Projects/Fusion Genes/Scripts/test_datasets/raw/dataset_1b.txt&#39;</span>
		<span class="n">gen_control_dataset</span><span class="p">(</span><span class="n">real_dataset</span><span class="p">,</span><span class="n">suffix</span><span class="p">)</span>
		<span class="n">n</span><span class="o">+=</span><span class="mi">1</span>
	<span class="nb">print</span><span class="p">(</span><span class="n">n</span><span class="p">,</span><span class="s1">&#39;datasets have been generated&#39;</span><span class="p">)</span></div>
<span class="c1"># gen_multiple_controls(&#39;/home/amarcozz/Documents/Projects/Fusion Genes/Scripts/test_datasets/raw/dataset_4.txt&#39;,1000)</span>
<span class="c1"># ## Generate multiple controls of datasets found in a folder</span>
<span class="c1"># folder = &#39;/home/amarcozz/Documents/Projects/Fusion Genes/Scripts/test_datasets/random&#39;</span>
<span class="c1"># for item in list_of_files(folder,&#39;txt&#39;):</span>
<span class="c1">#   gen_multiple_controls(item,1000)</span>

<div class="viewcode-block" id="gen_deletion_dataset_from_breaks"><a class="viewcode-back" href="../code.html#alefuncs.gen_deletion_dataset_from_breaks">[docs]</a><span class="k">def</span> <span class="nf">gen_deletion_dataset_from_breaks</span><span class="p">(</span><span class="n">list_of_breaks</span><span class="p">,</span> <span class="n">outfile</span><span class="p">,</span> <span class="n">ID_already</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;Genrates a proper deletion dataset file out of a list of breaks &#39;&#39;&#39;</span>
	<span class="c1"># Var names are not pythonic but I think it is better for readibility</span>
	<span class="n">header</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;##ID&#39;</span><span class="p">,</span> <span class="s1">&#39;ChrA&#39;</span><span class="p">,</span> <span class="s1">&#39;StartA&#39;</span><span class="p">,</span> <span class="s1">&#39;EndA&#39;</span><span class="p">,</span> <span class="s1">&#39;ChrB&#39;</span><span class="p">,</span> <span class="s1">&#39;StartB&#39;</span><span class="p">,</span> <span class="s1">&#39;EndB&#39;</span><span class="p">,</span> <span class="s1">&#39;CnvType&#39;</span><span class="p">,</span> <span class="s1">&#39;Orientation&#39;</span><span class="p">]</span>
	<span class="n">ID_list</span> <span class="o">=</span> <span class="p">[]</span> <span class="c1"># to check if the ID is already present</span>
	<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;writing breakpoints to&#39;</span><span class="p">,</span> <span class="n">outfile</span><span class="p">,</span> <span class="s1">&#39;..........&#39;</span><span class="p">)</span>
	<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">outfile</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">outfile</span><span class="p">:</span>
		<span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">list_to_line</span><span class="p">(</span><span class="n">header</span><span class="p">,</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
		<span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">list_of_breaks</span><span class="p">:</span>
			<span class="k">if</span> <span class="n">ID_already</span> <span class="o">==</span> <span class="kc">False</span><span class="p">:</span> <span class="c1"># the braks do not have an ID</span>
				<span class="k">while</span> <span class="kc">True</span><span class="p">:</span> <span class="c1"># checks ID</span>
					<span class="n">ID</span> <span class="o">=</span> <span class="n">gen_rnd_id</span><span class="p">(</span><span class="mi">8</span><span class="p">)</span>
					<span class="k">if</span> <span class="n">ID</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">ID_list</span><span class="p">:</span>
						<span class="n">ID_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ID</span><span class="p">)</span>
						<span class="k">break</span>
				<span class="n">ChrA</span> <span class="o">=</span> <span class="n">ChrB</span> <span class="o">=</span> <span class="n">item</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">3</span><span class="p">:]</span>
				<span class="n">StartA</span> <span class="o">=</span> <span class="n">EndA</span> <span class="o">=</span> <span class="n">item</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
				<span class="n">StartB</span> <span class="o">=</span> <span class="n">EndB</span> <span class="o">=</span> <span class="n">item</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
			<span class="k">else</span><span class="p">:</span> <span class="c1"># the break do have an ID</span>
				<span class="n">ID</span> <span class="o">=</span> <span class="n">item</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="c1"># the ID is supposed to be the first entry</span>
				<span class="n">ChrA</span> <span class="o">=</span> <span class="n">ChrB</span> <span class="o">=</span> <span class="n">item</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="mi">3</span><span class="p">:]</span>
				<span class="n">StartA</span> <span class="o">=</span> <span class="n">EndA</span> <span class="o">=</span> <span class="n">item</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
				<span class="n">StartB</span> <span class="o">=</span> <span class="n">EndB</span> <span class="o">=</span> <span class="n">item</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>
			<span class="n">CnvType</span> <span class="o">=</span> <span class="s1">&#39;DELETION&#39;</span>
			<span class="n">Orientation</span> <span class="o">=</span> <span class="s1">&#39;TH&#39;</span>

			<span class="n">line</span> <span class="o">=</span> <span class="n">list_to_line</span><span class="p">([</span><span class="n">ID</span><span class="p">,</span> <span class="n">ChrA</span><span class="p">,</span> <span class="n">StartA</span><span class="p">,</span> <span class="n">EndA</span><span class="p">,</span> <span class="n">ChrB</span><span class="p">,</span> <span class="n">StartB</span><span class="p">,</span> <span class="n">EndB</span><span class="p">,</span> <span class="n">CnvType</span><span class="p">,</span> <span class="n">Orientation</span><span class="p">],</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
			<span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">line</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
	<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;OK&#39;</span><span class="p">)</span></div>
<span class="c1"># list_of_breaks = gen_rnd_breaks(how_many=100, chromosome=&#39;Y&#39;, min_distance=1000, max_distance=15000, GapTable_file=&#39;tables/gap.txt&#39;)</span>
<span class="c1"># gen_deletion_dataset_from_breaks(list_of_breaks, &#39;test_deletion_dataset.txt&#39;)</span>
<span class="c1"># ## Generate (m) RANDOM datasets of different length (n)</span>
<span class="c1"># for m in range(1000):</span>
<span class="c1">#   for n in [100,1000,10000,100000,1000000]:</span>
<span class="c1">#       outfile = &#39;rnd_dataset_&#39;+ str(n)+&#39;_&#39;+str(m)+&#39;.txt&#39;</span>
<span class="c1">#       breaks = list()</span>
<span class="c1">#       for chromosome in CHROMOSOMES:      </span>
<span class="c1">#           breaks.extend(gen_rnd_breaks(how_many=500, chromosome=chromosome, min_distance=0, max_distance=n))</span>
<span class="c1">#       gen_deletion_dataset_from_breaks(breaks, outfile)</span>

<div class="viewcode-block" id="gen_rnd_breaks"><a class="viewcode-back" href="../code.html#alefuncs.gen_rnd_breaks">[docs]</a><span class="k">def</span> <span class="nf">gen_rnd_breaks</span><span class="p">(</span><span class="n">how_many</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">chromosome</span><span class="o">=</span><span class="s1">&#39;Y&#39;</span><span class="p">,</span> <span class="n">min_distance</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">max_distance</span><span class="o">=</span><span class="mi">15000</span><span class="p">,</span> <span class="n">GapTable_file</span><span class="o">=</span><span class="s1">&#39;tables/gap.txt&#39;</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;Returns tuples containing 1)the chromosome, 2)first breakpoint, 3)second breakpoint</span>
<span class="sd">	Keeps only the points that do not appear in te gap table.</span>
<span class="sd">	gen_rnd_breaks(int, string, int, int, filepath) =&gt; [(chrX, int, int), ...]</span>
<span class="sd">	valid chromosomes inputs are &quot;1&quot; to &quot;22&quot; ; &quot;Y&quot; ; &quot;X&quot;</span>
<span class="sd">	The chromosome length is based on the build GRCh37/hg19.&#39;&#39;&#39;</span>

	<span class="kn">import</span> <span class="nn">random</span>
	<span class="c1"># CHR_LENGTHS is based on GRCh37</span>
	<span class="n">CHR_LENGTHS</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;1&#39;</span><span class="p">:</span><span class="mi">249250621</span><span class="p">,</span><span class="s1">&#39;2&#39;</span> <span class="p">:</span><span class="mi">243199373</span><span class="p">,</span><span class="s1">&#39;3&#39;</span> <span class="p">:</span><span class="mi">198022430</span><span class="p">,</span><span class="s1">&#39;4&#39;</span> <span class="p">:</span><span class="mi">191154276</span><span class="p">,</span>
			   <span class="s1">&#39;5&#39;</span> <span class="p">:</span><span class="mi">180915260</span><span class="p">,</span><span class="s1">&#39;6&#39;</span> <span class="p">:</span><span class="mi">171115067</span><span class="p">,</span><span class="s1">&#39;7&#39;</span> <span class="p">:</span><span class="mi">159138663</span><span class="p">,</span><span class="s1">&#39;8&#39;</span> <span class="p">:</span><span class="mi">146364022</span><span class="p">,</span>
			   <span class="s1">&#39;9&#39;</span> <span class="p">:</span><span class="mi">141213431</span><span class="p">,</span><span class="s1">&#39;10&#39;</span><span class="p">:</span><span class="mi">135534747</span><span class="p">,</span><span class="s1">&#39;11&#39;</span><span class="p">:</span><span class="mi">135006516</span><span class="p">,</span><span class="s1">&#39;12&#39;</span><span class="p">:</span><span class="mi">133851895</span><span class="p">,</span>
			   <span class="s1">&#39;13&#39;</span><span class="p">:</span><span class="mi">115169878</span><span class="p">,</span><span class="s1">&#39;14&#39;</span><span class="p">:</span><span class="mi">107349540</span><span class="p">,</span><span class="s1">&#39;15&#39;</span><span class="p">:</span><span class="mi">102531392</span><span class="p">,</span><span class="s1">&#39;16&#39;</span><span class="p">:</span><span class="mi">90354753</span><span class="p">,</span>
			   <span class="s1">&#39;17&#39;</span><span class="p">:</span><span class="mi">81195210</span><span class="p">,</span><span class="s1">&#39;18&#39;</span><span class="p">:</span><span class="mi">78077248</span><span class="p">,</span><span class="s1">&#39;19&#39;</span><span class="p">:</span><span class="mi">59128983</span><span class="p">,</span><span class="s1">&#39;20&#39;</span><span class="p">:</span><span class="mi">63025520</span><span class="p">,</span>
			   <span class="s1">&#39;21&#39;</span><span class="p">:</span><span class="mi">48129895</span><span class="p">,</span><span class="s1">&#39;22&#39;</span><span class="p">:</span><span class="mi">51304566</span><span class="p">,</span><span class="s1">&#39;X&#39;</span> <span class="p">:</span><span class="mi">155270560</span><span class="p">,</span><span class="s1">&#39;Y&#39;</span> <span class="p">:</span><span class="mi">59373566</span><span class="p">}</span>

	<span class="c1"># Genrates a chromosome-specific gap list</span>
	<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;generating&#39;</span><span class="p">,</span> <span class="n">how_many</span><span class="p">,</span> <span class="s1">&#39;breakpoints in Chr&#39;</span><span class="p">,</span> <span class="n">chromosome</span><span class="p">,</span> <span class="s1">&#39;..........&#39;</span><span class="p">)</span>
	<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">GapTable_file</span><span class="p">,</span><span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">infile</span><span class="p">:</span>
		<span class="n">lines</span> <span class="o">=</span> <span class="n">infile</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
	
	<span class="n">full_gap_list</span> <span class="o">=</span> <span class="p">[]</span>
	<span class="n">chr_specific_gap</span> <span class="o">=</span> <span class="p">[]</span>
	<span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">lines</span><span class="p">:</span>
		<span class="k">if</span> <span class="s1">&#39;#&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">line</span><span class="p">:</span> <span class="c1"># skip comments</span>
			<span class="n">full_gap_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">line_to_list</span><span class="p">(</span><span class="n">line</span><span class="p">,</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">))</span>
	
	<span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">full_gap_list</span><span class="p">:</span>
		<span class="k">if</span> <span class="s1">&#39;chr&#39;</span> <span class="o">+</span> <span class="n">chromosome</span> <span class="ow">in</span> <span class="n">item</span><span class="p">:</span>
			<span class="c1"># Database/browser start coordinates differ by 1 base</span>
			<span class="n">chr_specific_gap</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">item</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span><span class="n">item</span><span class="p">[</span><span class="mi">3</span><span class="p">]))</span>

	<span class="c1"># Merge contiguous gaps</span>
	<span class="n">merged_gaps</span> <span class="o">=</span> <span class="p">[]</span>
	<span class="n">n</span> <span class="o">=</span> <span class="mi">0</span>
	<span class="n">left_tick</span> <span class="o">=</span> <span class="kc">False</span>
	<span class="k">while</span> <span class="n">n</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">chr_specific_gap</span><span class="p">):</span>
		<span class="k">if</span> <span class="n">left_tick</span> <span class="o">==</span> <span class="kc">False</span><span class="p">:</span>
			<span class="n">left_tick</span> <span class="o">=</span> <span class="n">chr_specific_gap</span><span class="p">[</span><span class="n">n</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
		<span class="k">try</span><span class="p">:</span>
			<span class="k">if</span> <span class="n">chr_specific_gap</span><span class="p">[</span><span class="n">n</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="n">chr_specific_gap</span><span class="p">[</span><span class="n">n</span><span class="o">+</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">]:</span>
				<span class="n">n</span> <span class="o">+=</span> <span class="mi">1</span>
			<span class="k">else</span><span class="p">:</span>
				<span class="n">right_tick</span> <span class="o">=</span> <span class="n">chr_specific_gap</span><span class="p">[</span><span class="n">n</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>
				<span class="n">merged_gaps</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">left_tick</span><span class="p">,</span><span class="n">right_tick</span><span class="p">))</span>
				<span class="n">left_tick</span> <span class="o">=</span> <span class="kc">False</span>
				<span class="n">n</span> <span class="o">+=</span> <span class="mi">1</span>
		<span class="k">except</span><span class="p">:</span>
			<span class="n">n</span> <span class="o">+=</span> <span class="mi">1</span>

	<span class="c1"># Genrates breakpoint list</span>
	<span class="n">list_of_breakpoints</span> <span class="o">=</span> <span class="p">[]</span>
	<span class="k">while</span> <span class="nb">len</span><span class="p">(</span><span class="n">list_of_breakpoints</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">how_many</span><span class="p">:</span>
		<span class="k">try</span><span class="p">:</span>
			<span class="n">start</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">CHR_LENGTHS</span><span class="p">[</span><span class="n">chromosome</span><span class="p">])</span>
		<span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
			<span class="k">if</span> <span class="n">chromosome</span> <span class="o">==</span> <span class="s1">&#39;23&#39;</span><span class="p">:</span>
				<span class="n">chromosome</span> <span class="o">=</span> <span class="s1">&#39;X&#39;</span>
				<span class="n">start</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">CHR_LENGTHS</span><span class="p">[</span><span class="n">chromosome</span><span class="p">])</span>
			<span class="k">elif</span> <span class="n">chromosome</span> <span class="o">==</span> <span class="s1">&#39;24&#39;</span><span class="p">:</span>
				<span class="n">chromosome</span> <span class="o">=</span> <span class="s1">&#39;Y&#39;</span>
				<span class="n">start</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">CHR_LENGTHS</span><span class="p">[</span><span class="n">chromosome</span><span class="p">])</span>
			<span class="k">else</span><span class="p">:</span>
				<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;ERROR: Wrong chromosome name!!&#39;</span><span class="p">)</span>


		<span class="n">end</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="n">start</span><span class="o">+</span><span class="n">min_distance</span><span class="p">,</span> <span class="n">start</span><span class="o">+</span><span class="n">max_distance</span><span class="p">)</span>
		<span class="n">are_points_ok</span> <span class="o">=</span> <span class="kc">True</span> <span class="c1"># assumes that the points are ok</span>
		
		<span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">merged_gaps</span><span class="p">:</span>
			<span class="c1"># checks whether the points are ok for real</span>
			<span class="k">if</span> <span class="n">start</span> <span class="o">&lt;</span> <span class="nb">int</span><span class="p">(</span><span class="n">item</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="ow">or</span> <span class="n">start</span> <span class="o">&gt;</span> <span class="nb">int</span><span class="p">(</span><span class="n">item</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
				<span class="k">if</span> <span class="n">end</span> <span class="o">&lt;</span> <span class="nb">int</span><span class="p">(</span><span class="n">item</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="ow">or</span> <span class="n">end</span> <span class="o">&gt;</span> <span class="nb">int</span><span class="p">(</span><span class="n">item</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
					<span class="k">pass</span>
				<span class="k">else</span><span class="p">:</span> <span class="n">are_points_ok</span> <span class="o">=</span> <span class="kc">False</span>
			<span class="k">else</span><span class="p">:</span> <span class="n">are_points_ok</span> <span class="o">=</span> <span class="kc">False</span>

		<span class="k">if</span> <span class="n">are_points_ok</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>           
			<span class="n">list_of_breakpoints</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="s1">&#39;chr&#39;</span><span class="o">+</span><span class="n">chromosome</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">))</span>
	<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;OK&#39;</span><span class="p">)</span>
	<span class="k">return</span> <span class="n">list_of_breakpoints</span></div>
<span class="c1"># print(gen_rnd_breaks(how_many=100, chromosome=&#39;Y&#39;, min_distance=1000, max_distance=15000, GapTable_file=&#39;tables/gap.txt&#39;))</span>

<span class="k">def</span> <span class="nf">gen_rnd_id</span><span class="p">(</span><span class="n">length</span><span class="p">):</span>
	<span class="kn">import</span> <span class="nn">random</span>
	<span class="sd">&#39;&#39;&#39;Generates a random string made by uppercase ascii chars and digits&#39;&#39;&#39;</span>
	<span class="n">chars</span> <span class="o">=</span> <span class="n">string</span><span class="o">.</span><span class="n">ascii_uppercase</span> <span class="o">+</span> <span class="n">string</span><span class="o">.</span><span class="n">digits</span>
	<span class="k">return</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">chars</span><span class="p">)</span> <span class="k">for</span> <span class="n">char</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">length</span><span class="p">))</span>
<span class="c1"># print(gen_rnd_id(16))</span>

<span class="c1">#@profile</span>
<div class="viewcode-block" id="gen_rnd_single_break"><a class="viewcode-back" href="../code.html#alefuncs.gen_rnd_single_break">[docs]</a><span class="k">def</span> <span class="nf">gen_rnd_single_break</span><span class="p">(</span><span class="n">how_many</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">chromosome</span><span class="o">=</span><span class="s1">&#39;1&#39;</span><span class="p">,</span> <span class="n">GapTable_file</span><span class="o">=</span><span class="s1">&#39;/Users/amarcozzi/Desktop/All_breakpoints_HG19_gap_10k.txt&#39;</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;Returns tuples containing 1)the chromosome, 2)the breakpoint</span>
<span class="sd">	Keeps only the points that do not appear in te gap table.</span>
<span class="sd">	gen_rnd_breaks(int, string, filepath) =&gt; [(chrX, int), ...]</span>
<span class="sd">	valid chromosomes inputs are &quot;1&quot; to &quot;22&quot; ; &quot;Y&quot; ; &quot;X&quot;</span>
<span class="sd">	The chromosome length is based on the build GRCh37/hg19.</span>
<span class="sd">	Prerequisites: The gap_list file is in the form:</span>
<span class="sd">														##chrom chromStart  chromEnd</span>
<span class="sd">														chr1    0   10000</span>
<span class="sd">														chr1    30000   40000</span>
<span class="sd">														chr1    40000   50000</span>
<span class="sd">														chr1    50000   60000</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="kn">import</span> <span class="nn">random</span>
	<span class="k">if</span> <span class="n">verbose</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
		<span class="kn">import</span> <span class="nn">time</span>
		<span class="n">start_time</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>

	<span class="c1"># CHR_LENGTHS is based on GRCh37</span>
	<span class="n">CHR_LENGTHS</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;1&#39;</span><span class="p">:</span><span class="mi">249250621</span><span class="p">,</span><span class="s1">&#39;2&#39;</span> <span class="p">:</span><span class="mi">243199373</span><span class="p">,</span><span class="s1">&#39;3&#39;</span> <span class="p">:</span><span class="mi">198022430</span><span class="p">,</span><span class="s1">&#39;4&#39;</span> <span class="p">:</span><span class="mi">191154276</span><span class="p">,</span>
			   <span class="s1">&#39;5&#39;</span> <span class="p">:</span><span class="mi">180915260</span><span class="p">,</span><span class="s1">&#39;6&#39;</span> <span class="p">:</span><span class="mi">171115067</span><span class="p">,</span><span class="s1">&#39;7&#39;</span> <span class="p">:</span><span class="mi">159138663</span><span class="p">,</span><span class="s1">&#39;8&#39;</span> <span class="p">:</span><span class="mi">146364022</span><span class="p">,</span>
			   <span class="s1">&#39;9&#39;</span> <span class="p">:</span><span class="mi">141213431</span><span class="p">,</span><span class="s1">&#39;10&#39;</span><span class="p">:</span><span class="mi">135534747</span><span class="p">,</span><span class="s1">&#39;11&#39;</span><span class="p">:</span><span class="mi">135006516</span><span class="p">,</span><span class="s1">&#39;12&#39;</span><span class="p">:</span><span class="mi">133851895</span><span class="p">,</span>
			   <span class="s1">&#39;13&#39;</span><span class="p">:</span><span class="mi">115169878</span><span class="p">,</span><span class="s1">&#39;14&#39;</span><span class="p">:</span><span class="mi">107349540</span><span class="p">,</span><span class="s1">&#39;15&#39;</span><span class="p">:</span><span class="mi">102531392</span><span class="p">,</span><span class="s1">&#39;16&#39;</span><span class="p">:</span><span class="mi">90354753</span><span class="p">,</span>
			   <span class="s1">&#39;17&#39;</span><span class="p">:</span><span class="mi">81195210</span><span class="p">,</span><span class="s1">&#39;18&#39;</span><span class="p">:</span><span class="mi">78077248</span><span class="p">,</span><span class="s1">&#39;19&#39;</span><span class="p">:</span><span class="mi">59128983</span><span class="p">,</span><span class="s1">&#39;20&#39;</span><span class="p">:</span><span class="mi">63025520</span><span class="p">,</span>
			   <span class="s1">&#39;21&#39;</span><span class="p">:</span><span class="mi">48129895</span><span class="p">,</span><span class="s1">&#39;22&#39;</span><span class="p">:</span><span class="mi">51304566</span><span class="p">,</span><span class="s1">&#39;X&#39;</span> <span class="p">:</span><span class="mi">155270560</span><span class="p">,</span><span class="s1">&#39;Y&#39;</span> <span class="p">:</span><span class="mi">59373566</span><span class="p">}</span>

	<span class="c1"># Genrates a chromosome-specific gap list</span>
	<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">GapTable_file</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">infile</span><span class="p">:</span>
		<span class="n">lines</span> <span class="o">=</span> <span class="n">infile</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
	
	<span class="n">full_gap_list</span> <span class="o">=</span> <span class="p">[]</span>
	<span class="n">chr_specific_gap</span> <span class="o">=</span> <span class="p">[]</span>
	<span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">lines</span><span class="p">:</span>
		<span class="k">if</span> <span class="s1">&#39;#&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">line</span><span class="p">:</span> <span class="c1"># skip comments</span>
			<span class="n">full_gap_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">line_to_list</span><span class="p">(</span><span class="n">line</span><span class="p">,</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">))</span>
	
	<span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">full_gap_list</span><span class="p">:</span>
		<span class="k">if</span> <span class="s1">&#39;chr&#39;</span> <span class="o">+</span> <span class="n">chromosome</span> <span class="ow">in</span> <span class="n">item</span><span class="p">:</span>
			<span class="n">chr_specific_gap</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">item</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span><span class="n">item</span><span class="p">[</span><span class="mi">2</span><span class="p">]))</span>

	<span class="c1"># Merge contiguous gaps</span>
	<span class="n">merged_gaps</span> <span class="o">=</span> <span class="n">merge_gaps</span><span class="p">(</span><span class="n">chr_specific_gap</span><span class="p">)</span>
	<span class="c1"># merged_gaps = []</span>
	<span class="c1"># while len(chr_specific_gap) &gt; 0:</span>
	<span class="c1">#   try:</span>
	<span class="c1">#       if chr_specific_gap[0][1] == chr_specific_gap[1][0]:</span>
	<span class="c1">#           tmp = (chr_specific_gap[0][0],chr_specific_gap[1][1])</span>
	<span class="c1">#           chr_specific_gap.pop(0)</span>
	<span class="c1">#           chr_specific_gap[0] = tmp</span>
	<span class="c1">#       else:</span>
	<span class="c1">#           merged_gaps.append(chr_specific_gap.pop(0))</span>
	<span class="c1">#   except:</span>
	<span class="c1">#       merged_gaps.append(chr_specific_gap.pop(0))</span>

	<span class="c1"># Genrates breakpoint list</span>
	<span class="k">if</span> <span class="n">verbose</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span> <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;generating&#39;</span><span class="p">,</span> <span class="n">how_many</span><span class="p">,</span> <span class="s1">&#39;breakpoints in Chr&#39;</span><span class="p">,</span> <span class="n">chromosome</span><span class="p">)</span>
	<span class="n">list_of_breakpoints</span> <span class="o">=</span> <span class="p">[]</span>
	<span class="k">while</span> <span class="nb">len</span><span class="p">(</span><span class="n">list_of_breakpoints</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">how_many</span><span class="p">:</span>
		<span class="k">try</span><span class="p">:</span>
			<span class="n">start</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">CHR_LENGTHS</span><span class="p">[</span><span class="n">chromosome</span><span class="p">])</span>
			<span class="c1"># if verbose == True: print(start)</span>
		<span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
			<span class="k">if</span> <span class="n">chromosome</span> <span class="o">==</span> <span class="s1">&#39;23&#39;</span><span class="p">:</span>
				<span class="n">chromosome</span> <span class="o">=</span> <span class="s1">&#39;X&#39;</span>
				<span class="n">start</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">CHR_LENGTHS</span><span class="p">[</span><span class="n">chromosome</span><span class="p">])</span>
			<span class="k">elif</span> <span class="n">chromosome</span> <span class="o">==</span> <span class="s1">&#39;24&#39;</span><span class="p">:</span>
				<span class="n">chromosome</span> <span class="o">=</span> <span class="s1">&#39;Y&#39;</span>
				<span class="n">start</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">CHR_LENGTHS</span><span class="p">[</span><span class="n">chromosome</span><span class="p">])</span>
			<span class="k">else</span><span class="p">:</span>
				<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;ERROR: Wrong chromosome name!!&#39;</span><span class="p">)</span>

		<span class="c1">#end = random.randint(start+min_distance, start+max_distance)</span>
		<span class="n">are_points_ok</span> <span class="o">=</span> <span class="kc">True</span> <span class="c1"># assumes that the points are ok</span>
		
		<span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">merged_gaps</span><span class="p">:</span>
			<span class="c1"># checks whether the points are ok for real</span>
			<span class="k">if</span> <span class="n">start</span> <span class="o">&lt;=</span> <span class="nb">int</span><span class="p">(</span><span class="n">item</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="ow">or</span> <span class="n">start</span> <span class="o">&gt;=</span> <span class="nb">int</span><span class="p">(</span><span class="n">item</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
				<span class="k">pass</span>
			<span class="k">else</span><span class="p">:</span>
				<span class="n">are_points_ok</span> <span class="o">=</span> <span class="kc">False</span>
				<span class="k">if</span> <span class="n">verbose</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span> <span class="nb">print</span><span class="p">(</span><span class="n">start</span><span class="p">,</span><span class="s1">&#39;is in a gap and will be discarded&#39;</span><span class="p">)</span>
			
		<span class="k">if</span> <span class="n">are_points_ok</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>           
			<span class="n">list_of_breakpoints</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">chromosome</span><span class="p">,</span> <span class="n">start</span><span class="p">))</span>
			<span class="k">if</span> <span class="n">verbose</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span> <span class="nb">print</span><span class="p">(</span><span class="n">start</span><span class="p">,</span><span class="s1">&#39;is OK&#39;</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">list_of_breakpoints</span><span class="p">),</span><span class="s1">&#39;good breaks generated out of&#39;</span><span class="p">,</span><span class="n">how_many</span><span class="p">)</span>

	<span class="k">if</span> <span class="n">verbose</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span> <span class="nb">print</span><span class="p">(</span><span class="n">how_many</span><span class="p">,</span><span class="s1">&#39;breakpoint have been generated in chromosome&#39;</span><span class="p">,</span><span class="n">chromosome</span><span class="p">,</span><span class="s1">&#39;in&#39;</span><span class="p">,</span><span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span><span class="o">-</span><span class="n">start_time</span><span class="p">,</span><span class="s1">&#39;seconds&#39;</span><span class="p">)</span>
	<span class="k">return</span> <span class="n">list_of_breakpoints</span></div>
<span class="c1"># gen_rnd_single_break(verbose=True)</span>
<span class="c1"># ## Generate single breaks dataset</span>
<span class="c1"># import time</span>
<span class="c1"># start = time.time()</span>
<span class="c1"># breaks_on_1 = gen_rnd_single_break(how_many=19147,verbose=False)</span>
<span class="c1"># for item in breaks_on_1:</span>
<span class="c1">#   print(str(item[0])+&#39;\t&#39;+str(item[1]))</span>
<span class="c1"># print(&#39;Done in&#39;, time.time()-start,&#39;seconds..&#39;)</span>
<span class="c1"># ## Generate a control file</span>
<span class="c1"># list_brkps = gen_rnd_single_break(how_many=20873, chromosome=&#39;1&#39;, GapTable_file=&#39;/Users/amarcozzi/Desktop/current_brkps_DB/out_ALL_gap.txt&#39;, verbose=True)</span>
<span class="c1"># with open(&#39;/Users/amarcozzi/Desktop/current_brkps_DB/out_chr1_control.txt&#39;,&#39;w&#39;) as f:</span>
<span class="c1">#   for item in list_brkps:</span>
<span class="c1">#       f.write(list_to_line(item,&#39;\t&#39;)+&#39;\n&#39;)</span>
<span class="c1"># ## Generate multiple controls</span>
<span class="c1"># import time</span>
<span class="c1"># from threading import Thread</span>
<span class="c1"># start_time = time.time()</span>
<span class="c1"># threads = 0</span>
<span class="c1"># running_threads = 0</span>
<span class="c1"># max_simultaneous_threads = 20</span>
<span class="c1"># GapTable_file = &#39;/Users/amarcozzi/Desktop/Projects/Anne_Project/current_brkps_DB/out_ALL_gap.txt&#39;</span>
<span class="c1"># chromosome = &#39;Y&#39;</span>
<span class="c1"># infile = &#39;/Users/amarcozzi/Desktop/Projects/Anne_Project/current_brkps_DB/out_chr&#39;+chromosome+&#39;.txt&#39;</span>
<span class="c1"># how_many = 0</span>
<span class="c1"># for line in yield_file(infile):</span>
<span class="c1">#   if line.startswith(chromosome+&#39;\t&#39;):</span>
<span class="c1">#       how_many += 1</span>
<span class="c1"># print(&#39;found&#39;,how_many,&#39;breakpoints in chromosome&#39;,chromosome)</span>
<span class="c1"># while threads &lt; 100:</span>
<span class="c1">#   while running_threads &gt;= max_simultaneous_threads:</span>
<span class="c1">#       time.sleep(1)</span>
<span class="c1">#   running_threads += 1</span>
<span class="c1">#   outfile = &#39;/Users/amarcozzi/Desktop/Projects/Anne_Project/current_brkps_DB/controls/out_chr&#39;+chromosome+&#39;_control_&#39;+str(threads)+&#39;.txt&#39;</span>
<span class="c1">#   print(&#39;thread&#39;, threads, &#39;|&#39;, &#39;running threads:&#39;,running_threads)</span>
<span class="c1">#   Thread(target=gen_controls, args=(how_many,chromosome,GapTable_file,outfile)).start()</span>
<span class="c1">#   threads += 1</span>
<span class="c1"># print(&#39;Waiting for threads to finish...&#39;)</span>
<span class="c1"># while running_threads &gt; 0:</span>
<span class="c1">#   time.sleep(1)</span>
<span class="c1"># end_time = time.time()</span>
<span class="c1"># print(&#39;\nDone in&#39;,(end_time-start_time)/60,&#39;minutes&#39;)</span>

<div class="viewcode-block" id="kmers_finder"><a class="viewcode-back" href="../code.html#alefuncs.kmers_finder">[docs]</a><span class="k">def</span> <span class="nf">kmers_finder</span><span class="p">(</span><span class="n">sequence_dict</span><span class="p">,</span> <span class="n">motif_length</span><span class="p">,</span> <span class="n">min_repetition</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;(dict, int, int) =&gt; OrderedDict(sorted(list))</span>
<span class="sd">	Find all the motifs long &#39;motif_length&#39; and repeated at least &#39;min_repetition&#39; times.</span>
<span class="sd">	Return an OrderedDict having motif:repetition as key:value sorted by value. </span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="kn">from</span> <span class="nn">collections</span> <span class="k">import</span> <span class="n">OrderedDict</span> 
	<span class="kn">from</span> <span class="nn">operator</span> <span class="k">import</span> <span class="n">itemgetter</span>
	<span class="n">motif_dict</span> <span class="o">=</span> <span class="p">{}</span>
	<span class="k">for</span> <span class="n">_id</span><span class="p">,</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="n">sequence_dict</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
		<span class="c1">#populate a dictionary of motifs (motif_dict)</span>
		<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span> <span class="o">-</span> <span class="n">motif_length</span> <span class="o">+</span><span class="mi">1</span><span class="p">):</span>
			<span class="n">motif</span> <span class="o">=</span> <span class="n">sequence</span><span class="p">[</span><span class="n">i</span><span class="p">:</span><span class="n">i</span><span class="o">+</span><span class="n">motif_length</span><span class="p">]</span>
			<span class="k">if</span> <span class="n">motif</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">motif_dict</span><span class="p">:</span>
				<span class="n">motif_dict</span><span class="p">[</span><span class="n">motif</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
			<span class="k">else</span><span class="p">:</span>
				<span class="n">motif_dict</span><span class="p">[</span><span class="n">motif</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>

	<span class="c1">#remove from motif_dict all the motifs repeated less than &#39;repetition&#39; times</span>
	<span class="n">keys_to_remove</span> <span class="o">=</span> <span class="p">[</span><span class="n">key</span> <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">motif_dict</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">value</span> <span class="o">&lt;</span> <span class="n">min_repetition</span><span class="p">]</span>
	<span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">keys_to_remove</span><span class="p">:</span>
		<span class="k">del</span> <span class="n">motif_dict</span><span class="p">[</span><span class="n">key</span><span class="p">]</span>
	
	<span class="c1">#Return a sorted dictionary</span>
	<span class="k">return</span> <span class="n">OrderedDict</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">motif_dict</span><span class="o">.</span><span class="n">items</span><span class="p">(),</span> <span class="n">key</span><span class="o">=</span><span class="n">itemgetter</span><span class="p">(</span><span class="mi">1</span><span class="p">),</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span></div>

<div class="viewcode-block" id="kmers_finder_with_mismatches"><a class="viewcode-back" href="../code.html#alefuncs.kmers_finder_with_mismatches">[docs]</a><span class="k">def</span> <span class="nf">kmers_finder_with_mismatches</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">motif_length</span><span class="p">,</span> <span class="n">max_mismatches</span><span class="p">,</span> <span class="n">most_common</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;(str, int, int) =&gt; sorted(list)</span>
<span class="sd">	Find the most frequent k-mers with mismatches in a string.</span>
<span class="sd">	Input: A sequence and a pair of integers: motif_length (&lt;=12) and max_mismatch (&lt;= 3).</span>
<span class="sd">	Output: An OrderedDict containing all k-mers with up to d mismatches in string.</span>

<span class="sd">	Sample Input:   ACGTTGCATGTCGCATGATGCATGAGAGCT 4 1</span>
<span class="sd">	Sample Output:  OrderedDict([(&#39;ATGC&#39;, 5), (&#39;ATGT&#39;, 5), (&#39;GATG&#39;, 5),...])</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="kn">from</span> <span class="nn">collections</span> <span class="k">import</span> <span class="n">OrderedDict</span>
	<span class="kn">from</span> <span class="nn">operator</span> <span class="k">import</span> <span class="n">itemgetter</span>

	<span class="c1">#check passed variables</span>
	<span class="k">if</span> <span class="ow">not</span> <span class="n">motif_length</span> <span class="o">&lt;=</span> <span class="mi">12</span> <span class="ow">and</span> <span class="n">motif_length</span> <span class="o">&gt;=</span> <span class="mi">1</span><span class="p">:</span>
		<span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;motif_length must be between 0 and 12. </span><span class="si">{}</span><span class="s2"> was passed.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">motif_length</span><span class="p">))</span>
	<span class="k">if</span> <span class="ow">not</span> <span class="n">max_mismatches</span> <span class="o">&lt;=</span> <span class="mi">3</span> <span class="ow">and</span> <span class="n">max_mismatches</span> <span class="o">&gt;=</span> <span class="mi">1</span><span class="p">:</span>
		<span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;max_mismatch must be between 0 and 3. </span><span class="si">{}</span><span class="s2"> was passed.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">max_mismatches</span><span class="p">))</span>

	<span class="n">motif_dict</span> <span class="o">=</span> <span class="p">{}</span>
	<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span> <span class="o">-</span> <span class="n">motif_length</span> <span class="o">+</span><span class="mi">1</span><span class="p">):</span>
		<span class="n">motif</span> <span class="o">=</span> <span class="n">sequence</span><span class="p">[</span><span class="n">i</span><span class="p">:</span><span class="n">i</span><span class="o">+</span><span class="n">motif_length</span><span class="p">]</span>
		<span class="k">if</span> <span class="n">motif</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">motif_dict</span><span class="p">:</span>
			<span class="n">motif_dict</span><span class="p">[</span><span class="n">motif</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
		<span class="k">else</span><span class="p">:</span>
			<span class="n">motif_dict</span><span class="p">[</span><span class="n">motif</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>

	<span class="n">motif_dict_with_mismatches</span> <span class="o">=</span> <span class="p">{}</span>
	<span class="k">for</span> <span class="n">kmer</span> <span class="ow">in</span> <span class="n">motif_dict</span><span class="p">:</span>
		<span class="n">motif_dict_with_mismatches</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="n">kmer</span><span class="p">:[]})</span>
		<span class="k">for</span> <span class="n">other_kmer</span> <span class="ow">in</span> <span class="n">motif_dict</span><span class="p">:</span>
			<span class="n">mismatches</span> <span class="o">=</span> <span class="mi">0</span>
			<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">kmer</span><span class="p">)):</span>
				<span class="k">if</span> <span class="n">kmer</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">!=</span> <span class="n">other_kmer</span><span class="p">[</span><span class="n">i</span><span class="p">]:</span>
					<span class="n">mismatches</span> <span class="o">+=</span> <span class="mi">1</span>
			<span class="k">if</span> <span class="n">mismatches</span> <span class="o">&lt;=</span> <span class="n">max_mismatches</span><span class="p">:</span>
				<span class="n">motif_dict_with_mismatches</span><span class="p">[</span><span class="n">kmer</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">other_kmer</span><span class="p">,</span><span class="n">motif_dict</span><span class="p">[</span><span class="n">other_kmer</span><span class="p">]])</span>

	<span class="n">tmp</span> <span class="o">=</span> <span class="p">{}</span>
	<span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">motif_dict_with_mismatches</span><span class="p">:</span>
		<span class="n">count</span> <span class="o">=</span> <span class="mi">0</span>
		<span class="k">for</span> <span class="n">motif</span> <span class="ow">in</span> <span class="n">motif_dict_with_mismatches</span><span class="p">[</span><span class="n">item</span><span class="p">]:</span>
			<span class="n">count</span> <span class="o">+=</span> <span class="n">motif</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
		<span class="n">tmp</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="n">item</span><span class="p">:</span><span class="n">count</span><span class="p">})</span>

	<span class="n">result</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">tmp</span><span class="o">.</span><span class="n">items</span><span class="p">(),</span> <span class="n">key</span><span class="o">=</span><span class="n">itemgetter</span><span class="p">(</span><span class="mi">1</span><span class="p">),</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>
	
	<span class="k">if</span> <span class="n">most_common</span><span class="p">:</span>
		<span class="n">commons</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
		<span class="n">_max</span> <span class="o">=</span> <span class="n">result</span><span class="o">.</span><span class="n">items</span><span class="p">()[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>
		<span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">result</span><span class="p">:</span>
			<span class="k">if</span> <span class="n">result</span><span class="p">[</span><span class="n">item</span><span class="p">]</span> <span class="o">==</span> <span class="n">_max</span><span class="p">:</span>
				<span class="n">commons</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="n">item</span><span class="p">:</span><span class="n">result</span><span class="p">[</span><span class="n">item</span><span class="p">]})</span>
			<span class="k">else</span><span class="p">:</span>
				<span class="k">return</span> <span class="n">commons</span>

	<span class="k">return</span> <span class="n">result</span></div>

<div class="viewcode-block" id="line_to_list"><a class="viewcode-back" href="../code.html#alefuncs.line_to_list">[docs]</a><span class="k">def</span> <span class="nf">line_to_list</span><span class="p">(</span><span class="n">line</span><span class="p">,</span> <span class="n">char</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;Makes a list of string out of a line. Splits the word at char.&#39;&#39;&#39;</span>
	<span class="c1"># Allows for more customization compared with string.split()</span>
	<span class="n">split_indexes</span> <span class="o">=</span> <span class="n">find</span><span class="p">(</span><span class="n">line</span><span class="p">,</span> <span class="n">char</span><span class="p">)</span>
	<span class="n">list_</span> <span class="o">=</span> <span class="p">[]</span>
	<span class="n">n</span> <span class="o">=</span> <span class="mi">0</span>
	<span class="k">for</span> <span class="n">index</span> <span class="ow">in</span> <span class="n">split_indexes</span><span class="p">:</span>
		<span class="n">item</span> <span class="o">=</span> <span class="n">line</span><span class="p">[</span><span class="n">n</span><span class="p">:</span><span class="n">index</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">,</span><span class="s1">&#39;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\r</span><span class="s1">&#39;</span><span class="p">,</span><span class="s1">&#39;&#39;</span><span class="p">)</span> <span class="c1"># cleans up the line</span>
		<span class="k">if</span> <span class="n">item</span> <span class="o">!=</span> <span class="s1">&#39;&#39;</span><span class="p">:</span> <span class="c1"># skips empty &#39;cells&#39;</span>
			<span class="n">list_</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">item</span><span class="p">)</span>
		<span class="n">n</span> <span class="o">=</span> <span class="n">index</span> <span class="o">+</span> <span class="mi">1</span>
	<span class="n">list_</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">line</span><span class="p">[</span><span class="n">n</span><span class="p">:]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">,</span><span class="s1">&#39;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\r</span><span class="s1">&#39;</span><span class="p">,</span><span class="s1">&#39;&#39;</span><span class="p">))</span> <span class="c1"># append the last item </span>
	<span class="k">return</span> <span class="n">list_</span></div>
<span class="c1"># print(line_to_list(&#39;Makes a list of string out of a line. Splits the word at char.&#39;, &#39; &#39;))</span>

<div class="viewcode-block" id="list_to_line"><a class="viewcode-back" href="../code.html#alefuncs.list_to_line">[docs]</a><span class="k">def</span> <span class="nf">list_to_line</span><span class="p">(</span><span class="n">list_</span><span class="p">,</span> <span class="n">char</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;Makes a string out of a list of items&#39;&#39;&#39;</span>
	<span class="c1"># Allows for more customization compared with string.split()</span>
	<span class="n">string</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
	<span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">list_</span><span class="p">:</span>
		<span class="n">string</span> <span class="o">+=</span> <span class="nb">str</span><span class="p">(</span><span class="n">item</span><span class="p">)</span> <span class="o">+</span> <span class="n">char</span>
	<span class="k">return</span> <span class="n">string</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="n">char</span><span class="p">)</span> <span class="c1"># Removes the last char</span></div>
<span class="c1">#print(list_to_line([&#39;prova&#39;, &#39;1&#39;, &#39;2&#39;, &#39;3&#39;, &#39;prova&#39;], &#39;---&#39;))</span>

<div class="viewcode-block" id="list_of_files"><a class="viewcode-back" href="../code.html#alefuncs.list_of_files">[docs]</a><span class="k">def</span> <span class="nf">list_of_files</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">extension</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	Return a list of filepath for each file into path with the target extension.</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="kn">import</span> <span class="nn">glob</span>
	<span class="k">return</span> <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">path</span> <span class="o">+</span> <span class="s1">&#39;/*.&#39;</span> <span class="o">+</span> <span class="n">extension</span><span class="p">))</span></div>
<span class="c1"># print(list_of_files(&#39;/home/amarcozz/Documents/Projects/Fusion Genes/Scripts/test datasets&#39;, &#39;txt&#39;))</span>

<div class="viewcode-block" id="merge_gaps"><a class="viewcode-back" href="../code.html#alefuncs.merge_gaps">[docs]</a><span class="k">def</span> <span class="nf">merge_gaps</span><span class="p">(</span><span class="n">gap_list</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	Merges overlapping gaps in a gap list.</span>
<span class="sd">	The gap list is in the form: [(&#39;3&#39;,&#39;4&#39;),(&#39;5&#39;,&#39;6&#39;),(&#39;6&#39;,&#39;7&#39;),(&#39;8&#39;,&#39;9&#39;),(&#39;10&#39;,&#39;11&#39;),(&#39;15&#39;,&#39;16&#39;),(&#39;17&#39;,&#39;18&#39;),(&#39;18&#39;,&#39;19&#39;)]</span>
<span class="sd">	Returns a new list containing the merged gaps: [(&#39;3&#39;,&#39;4&#39;),(&#39;5&#39;,&#39;7&#39;),(&#39;8&#39;,&#39;9&#39;),(&#39;10&#39;,&#39;11&#39;),(&#39;15&#39;,&#39;16&#39;),(&#39;17&#39;,&#39;19&#39;)]</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="n">merged_gaps</span> <span class="o">=</span> <span class="p">[]</span>
	<span class="k">while</span> <span class="nb">len</span><span class="p">(</span><span class="n">gap_list</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
		<span class="k">try</span><span class="p">:</span>
			<span class="k">if</span> <span class="nb">int</span><span class="p">(</span><span class="n">gap_list</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">])</span> <span class="o">&gt;=</span> <span class="nb">int</span><span class="p">(</span><span class="n">gap_list</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">]):</span>
				<span class="n">tmp</span> <span class="o">=</span> <span class="p">(</span><span class="n">gap_list</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span><span class="n">gap_list</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="mi">1</span><span class="p">])</span>
				<span class="n">gap_list</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
				<span class="n">gap_list</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">tmp</span>
			<span class="k">else</span><span class="p">:</span>
				<span class="n">merged_gaps</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gap_list</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="mi">0</span><span class="p">))</span>
		<span class="k">except</span><span class="p">:</span>
			<span class="n">merged_gaps</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gap_list</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="mi">0</span><span class="p">))</span>
	<span class="k">return</span> <span class="n">merged_gaps</span></div>
<span class="c1"># gap_list = [(&#39;3&#39;,&#39;4&#39;),(&#39;5&#39;,&#39;6&#39;),(&#39;6&#39;,&#39;7&#39;),(&#39;8&#39;,&#39;9&#39;),(&#39;10&#39;,&#39;11&#39;),(&#39;15&#39;,&#39;16&#39;),(&#39;17&#39;,&#39;18&#39;),(&#39;18&#39;,&#39;19&#39;)]</span>
<span class="c1"># expected = [(&#39;3&#39;,&#39;4&#39;),(&#39;5&#39;,&#39;7&#39;),(&#39;8&#39;,&#39;9&#39;),(&#39;10&#39;,&#39;11&#39;),(&#39;15&#39;,&#39;16&#39;),(&#39;17&#39;,&#39;19&#39;)]</span>
<span class="c1"># prova = merge_gaps(gap_list)</span>
<span class="c1"># print(prova)</span>
<span class="c1"># print(expected)</span>

<div class="viewcode-block" id="merge_sort"><a class="viewcode-back" href="../code.html#alefuncs.merge_sort">[docs]</a><span class="k">def</span> <span class="nf">merge_sort</span><span class="p">(</span><span class="n">intervals</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	Merges and sorts the intervals in a list.</span>
<span class="sd">	It&#39;s an alternative of merge_gaps() that sort the list before merging.</span>
<span class="sd">	Should be faster but I haven&#39;t campared them yet.</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="n">sorted_by_lower_bound</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">intervals</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">tup</span><span class="p">:</span> <span class="n">tup</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
	<span class="n">merged</span> <span class="o">=</span> <span class="p">[]</span>

	<span class="k">for</span> <span class="n">higher</span> <span class="ow">in</span> <span class="n">sorted_by_lower_bound</span><span class="p">:</span>
		<span class="k">if</span> <span class="ow">not</span> <span class="n">merged</span><span class="p">:</span>
			<span class="n">merged</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">higher</span><span class="p">)</span>
		<span class="k">else</span><span class="p">:</span>
			<span class="n">lower</span> <span class="o">=</span> <span class="n">merged</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
			<span class="c1"># test for intersection between lower and higher:</span>
			<span class="c1"># we know via sorting that lower[0] &lt;= higher[0]</span>
			<span class="k">if</span> <span class="n">higher</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">lower</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span>
				<span class="n">upper_bound</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">lower</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">higher</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
				<span class="n">merged</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">lower</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">upper_bound</span><span class="p">)</span>  <span class="c1"># replace by merged interval</span>
			<span class="k">else</span><span class="p">:</span>
				<span class="n">merged</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">higher</span><span class="p">)</span>
	<span class="k">return</span> <span class="n">merged</span> </div>

<div class="viewcode-block" id="multi_threads_fusion_genes_annotation"><a class="viewcode-back" href="../code.html#alefuncs.multi_threads_fusion_genes_annotation">[docs]</a><span class="k">def</span> <span class="nf">multi_threads_fusion_genes_annotation</span><span class="p">(</span><span class="n">folder_path</span><span class="p">,</span> <span class="n">extension</span><span class="p">,</span> <span class="n">max_simultaneous_threads</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39; Executes annotate_fusion_genes() for each dataset file in a folder.</span>
<span class="sd">	Each execution run on a different thread.&#39;&#39;&#39;</span>
	<span class="kn">from</span> <span class="nn">threading</span> <span class="k">import</span> <span class="n">Thread</span>
	<span class="k">global</span> <span class="n">running_threads</span>
	<span class="n">dataset_files</span> <span class="o">=</span> <span class="n">list_of_files</span><span class="p">(</span><span class="n">folder_path</span><span class="p">,</span> <span class="n">extension</span><span class="p">)</span>
	<span class="n">threads</span> <span class="o">=</span> <span class="mi">0</span>
	<span class="n">running_threads</span> <span class="o">=</span> <span class="mi">0</span>
	<span class="k">for</span> <span class="n">file_</span> <span class="ow">in</span> <span class="n">dataset_files</span><span class="p">:</span>
		<span class="k">while</span> <span class="n">running_threads</span> <span class="o">&gt;=</span> <span class="n">max_simultaneous_threads</span><span class="p">:</span>
			<span class="n">time</span><span class="o">.</span><span class="n">sleep</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
		<span class="n">threads</span> <span class="o">+=</span> <span class="mi">1</span>
		<span class="n">running_threads</span> <span class="o">+=</span> <span class="mi">1</span>
		<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;thread&#39;</span><span class="p">,</span> <span class="n">threads</span><span class="p">,</span> <span class="s1">&#39;|&#39;</span><span class="p">,</span> <span class="s1">&#39;running threads:&#39;</span><span class="p">,</span><span class="n">running_threads</span><span class="p">)</span>
		<span class="n">Thread</span><span class="p">(</span><span class="n">target</span><span class="o">=</span><span class="n">annotate_fusion_genes</span><span class="p">,</span> <span class="n">args</span><span class="o">=</span><span class="p">(</span><span class="n">file_</span><span class="p">,))</span><span class="o">.</span><span class="n">start</span><span class="p">()</span> <span class="c1"># with multithreading</span></div>
<span class="c1"># folder = &#39;/home/amarcozz/Documents/Projects/Fusion Genes/Scripts/test_datasets/public&#39;</span>
<span class="c1"># multi_threads_fusion_genes_annotation(folder, &#39;txt&#39;,50)</span>

<div class="viewcode-block" id="pandize_dataset"><a class="viewcode-back" href="../code.html#alefuncs.pandize_dataset">[docs]</a><span class="k">def</span> <span class="nf">pandize_dataset</span><span class="p">(</span><span class="n">annotated_dataset</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	Prepares a dataset to be &quot;pandas ready&quot;.</span>
<span class="sd">	Takes a file path as input.</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="kn">import</span> <span class="nn">pandas</span>
	<span class="c1"># Parse</span>
	<span class="k">if</span> <span class="n">verbose</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
		<span class="n">message</span> <span class="o">=</span> <span class="s1">&#39;parsing &#39;</span> <span class="o">+</span> <span class="n">annotated_dataset</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
		<span class="n">spacer</span> <span class="o">=</span> <span class="p">(</span><span class="mi">100</span><span class="o">-</span><span class="nb">len</span><span class="p">(</span><span class="n">message</span><span class="p">))</span><span class="o">*</span><span class="s1">&#39;.&#39;</span>
		<span class="nb">print</span><span class="p">(</span><span class="n">message</span><span class="p">,</span> <span class="n">spacer</span><span class="p">)</span>

	<span class="n">dataset</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">parsers</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span><span class="n">annotated_dataset</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;ChrA&#39;</span><span class="p">:</span><span class="s1">&#39;str&#39;</span><span class="p">,</span><span class="s1">&#39;ChrB&#39;</span><span class="p">:</span><span class="s1">&#39;str&#39;</span><span class="p">},</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
	<span class="k">if</span> <span class="n">verbose</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
		<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;OK&#39;</span><span class="p">)</span>
	
	<span class="c1"># Clean</span>
	<span class="k">if</span> <span class="n">verbose</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
		<span class="n">message</span> <span class="o">=</span> <span class="s1">&#39;cleaning &#39;</span> <span class="o">+</span> <span class="n">annotated_dataset</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
		<span class="n">spacer</span> <span class="o">=</span> <span class="p">(</span><span class="mi">100</span><span class="o">-</span><span class="nb">len</span><span class="p">(</span><span class="n">message</span><span class="p">))</span><span class="o">*</span><span class="s1">&#39;.&#39;</span>
		<span class="nb">print</span><span class="p">(</span><span class="n">message</span><span class="p">,</span> <span class="n">spacer</span><span class="p">)</span>

	<span class="n">dataset</span> <span class="o">=</span> <span class="n">dataset</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;In Frame&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
	<span class="n">dataset</span> <span class="o">=</span> <span class="n">dataset</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;Not in Frame&#39;</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
	<span class="n">dataset</span> <span class="o">=</span> <span class="n">dataset</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;In Phase&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
	<span class="n">dataset</span> <span class="o">=</span> <span class="n">dataset</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;Not in Phase&#39;</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
	<span class="k">if</span> <span class="n">verbose</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
		<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;OK&#39;</span><span class="p">)</span>

	<span class="k">return</span> <span class="n">dataset</span></div>
<span class="c1"># pandize_dataset(&#39;test_data_annotated.txt&#39;)</span>
<span class="c1"># pandize_dataset(&#39;/home/amarcozz/Documents/Projects/Fusion Genes/Scripts/test_datasets/control_dataset_100-1000-150000_annotated.txt&#39;)</span>

<div class="viewcode-block" id="parse_blastXML"><a class="viewcode-back" href="../code.html#alefuncs.parse_blastXML">[docs]</a><span class="k">def</span> <span class="nf">parse_blastXML</span><span class="p">(</span><span class="n">infile</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	Parses a blast outfile (XML).</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="kn">from</span> <span class="nn">Bio.Blast</span> <span class="k">import</span> <span class="n">NCBIXML</span>
	<span class="k">for</span> <span class="n">blast_record</span> <span class="ow">in</span> <span class="n">NCBIXML</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="n">infile</span><span class="p">)):</span>
		<span class="k">for</span> <span class="n">alignment</span> <span class="ow">in</span> <span class="n">blast_record</span><span class="o">.</span><span class="n">alignments</span><span class="p">:</span>
			<span class="k">for</span> <span class="n">hsp</span> <span class="ow">in</span> <span class="n">alignment</span><span class="o">.</span><span class="n">hsps</span><span class="p">:</span>
				<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;*****Alignment****&quot;</span><span class="p">)</span>
				<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;sequence:&quot;</span><span class="p">,</span> <span class="n">alignment</span><span class="o">.</span><span class="n">title</span><span class="p">)</span>
				<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;length:&quot;</span><span class="p">,</span> <span class="n">alignment</span><span class="o">.</span><span class="n">length</span><span class="p">)</span>
				<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;e-value:&quot;</span><span class="p">,</span> <span class="n">hsp</span><span class="o">.</span><span class="n">expect</span><span class="p">)</span>
				<span class="nb">print</span><span class="p">(</span><span class="n">hsp</span><span class="o">.</span><span class="n">query</span><span class="p">)</span>
				<span class="nb">print</span><span class="p">(</span><span class="n">hsp</span><span class="o">.</span><span class="n">match</span><span class="p">)</span>
				<span class="nb">print</span><span class="p">(</span><span class="n">hsp</span><span class="o">.</span><span class="n">sbjct</span><span class="p">)</span></div>
<span class="c1"># to be tested</span>


<span class="k">def</span> <span class="nf">reverse</span><span class="p">(</span><span class="n">sequence</span><span class="p">):</span>
	<span class="n">r</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
	<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">xrange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">),</span><span class="mi">0</span><span class="p">,</span><span class="o">-</span><span class="mi">1</span><span class="p">):</span>
		<span class="n">r</span> <span class="o">+=</span> <span class="n">sequence</span><span class="p">[</span><span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
	<span class="k">return</span> <span class="n">r</span>

<span class="k">def</span> <span class="nf">complement</span><span class="p">(</span><span class="n">sequence</span><span class="p">):</span>
	<span class="n">d</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;A&#39;</span><span class="p">:</span><span class="s1">&#39;T&#39;</span><span class="p">,</span>
		 <span class="s1">&#39;T&#39;</span><span class="p">:</span><span class="s1">&#39;A&#39;</span><span class="p">,</span>
		 <span class="s1">&#39;C&#39;</span><span class="p">:</span><span class="s1">&#39;G&#39;</span><span class="p">,</span>
		 <span class="s1">&#39;G&#39;</span><span class="p">:</span><span class="s1">&#39;C&#39;</span><span class="p">}</span>
	<span class="n">r</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
	<span class="k">for</span> <span class="n">b</span> <span class="ow">in</span> <span class="n">sequence</span><span class="o">.</span><span class="n">upper</span><span class="p">():</span>
		<span class="n">r</span> <span class="o">+=</span> <span class="n">d</span><span class="p">[</span><span class="n">b</span><span class="p">]</span>
	<span class="k">return</span> <span class="n">r</span>

<span class="k">def</span> <span class="nf">get_mismatches</span><span class="p">(</span><span class="n">template</span><span class="p">,</span><span class="n">primer</span><span class="p">,</span><span class="n">maxerr</span><span class="p">,</span><span class="n">overlapped</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
	<span class="kn">import</span> <span class="nn">regex</span>
	<span class="n">error</span> <span class="o">=</span> <span class="s1">&#39;e&lt;=</span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">maxerr</span><span class="p">)</span>
	<span class="k">return</span> <span class="n">regex</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="s1">&#39;(</span><span class="si">{}</span><span class="s1">){{</span><span class="si">{}</span><span class="s1">}}&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">primer</span><span class="p">,</span><span class="n">error</span><span class="p">),</span> <span class="n">template</span><span class="p">,</span> <span class="n">overlapped</span><span class="o">=</span><span class="n">overlapped</span><span class="p">)</span>


<span class="k">def</span> <span class="nf">pcr</span><span class="p">(</span><span class="n">template</span><span class="p">,</span><span class="n">primer_F</span><span class="p">,</span><span class="n">primer_R</span><span class="p">,</span><span class="n">circular</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
	<span class="kn">import</span> <span class="nn">re</span>
	<span class="k">if</span> <span class="n">circular</span><span class="p">:</span> <span class="c1">##works only with primers without 5&#39; overhang</span>
		<span class="n">i</span> <span class="o">=</span> <span class="n">template</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="n">primer_F</span><span class="o">.</span><span class="n">upper</span><span class="p">())</span>
		<span class="n">template</span> <span class="o">=</span> <span class="n">template</span><span class="p">[</span><span class="n">i</span><span class="p">:]</span><span class="o">+</span><span class="n">template</span><span class="p">[:</span><span class="n">i</span><span class="p">]</span>
	
	<span class="c1">#Find primer_F, or the largest 3&#39;part of it, in template</span>
	<span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">primer_F</span><span class="p">)):</span>
		<span class="n">ix_F</span> <span class="o">=</span> <span class="p">[</span><span class="n">m</span><span class="o">.</span><span class="n">end</span><span class="p">()</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">re</span><span class="o">.</span><span class="n">finditer</span><span class="p">(</span><span class="n">primer_F</span><span class="p">[</span><span class="n">n</span><span class="p">:]</span><span class="o">.</span><span class="n">upper</span><span class="p">(),</span>
										   <span class="n">template</span><span class="o">.</span><span class="n">upper</span><span class="p">())]</span>
		<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">ix_F</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span> <span class="c1">#it&#39;s unique</span>
			<span class="c1">#print(ix_F)</span>
			<span class="c1">#print(primer_F[n:])</span>
			<span class="k">break</span>
		<span class="n">n</span> <span class="o">+=</span> <span class="mi">1</span>
	<span class="c1">#print(ix_F)</span>
	<span class="c1">#Find primer_R, or the largest 5&#39;part of it, in template</span>
	<span class="n">rc_R</span> <span class="o">=</span> <span class="n">reverse</span><span class="p">(</span><span class="n">complement</span><span class="p">(</span><span class="n">primer_R</span><span class="p">))</span>
	<span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">primer_R</span><span class="p">)):</span>
		<span class="n">ix</span> <span class="o">=</span> <span class="p">[</span><span class="n">m</span><span class="o">.</span><span class="n">start</span><span class="p">()</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">re</span><span class="o">.</span><span class="n">finditer</span><span class="p">(</span><span class="n">rc_R</span><span class="p">[:</span><span class="n">n</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">(),</span>
										   <span class="n">template</span><span class="o">.</span><span class="n">upper</span><span class="p">())]</span>
		<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">ix</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span> <span class="c1">#it&#39;s unique</span>
			<span class="n">ix_R</span> <span class="o">=</span> <span class="n">ix</span><span class="p">[:]</span>

		<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">ix</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span> <span class="c1">#it&#39;s the largest possible</span>
			<span class="c1">#print(ix_R)</span>
			<span class="c1">#print(rc_R[:n])</span>
			<span class="k">break</span>
		<span class="n">n</span> <span class="o">+=</span> <span class="mi">1</span>
	<span class="c1">#Build the product</span>
	<span class="k">return</span> <span class="n">primer_F</span> <span class="o">+</span> <span class="n">template</span><span class="p">[</span><span class="n">ix_F</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">ix_R</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">+</span> <span class="n">rc_R</span>
<span class="c1">##template = &#39;CTAGAGAGGGCCTATTTCCCATGATT--something--GCCAATTCTGCAGACAAATGGGGTACCCG&#39;</span>
<span class="c1">##primer_F = &#39;GACAAATGGCTCTAGAGAGGGCCTATTTCCCATGATT&#39;</span>
<span class="c1">##primer_R = &#39;TTATGTAACGGGTACCCCATTTGTCTGCAGAATTGGC&#39;</span>
<span class="c1">##product = pcr(template,primer_F,primer_R)</span>
<span class="c1">##expected = &#39;GACAAATGGCTCTAGAGAGGGCCTATTTCCCATGATT--something--GCCAATTCTGCAGACAAATGGGGTACCCGTTACATAA&#39;</span>
<span class="c1">##expected == result</span>

<div class="viewcode-block" id="pip_upgrade_all"><a class="viewcode-back" href="../code.html#alefuncs.pip_upgrade_all">[docs]</a><span class="k">def</span> <span class="nf">pip_upgrade_all</span><span class="p">():</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	Upgrades all pip-installed packages.</span>
<span class="sd">	Requires a bash shell. (Unix)</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="kn">from</span> <span class="nn">subprocess</span> <span class="k">import</span> <span class="n">call</span>
	<span class="c1">#pip</span>
	<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;upgrading pip...&#39;</span><span class="p">)</span>
	<span class="n">call</span><span class="p">(</span><span class="s1">&#39;pip install --upgrade pip&#39;</span><span class="p">,</span> <span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
	<span class="n">call</span><span class="p">(</span><span class="s2">&quot;pip freeze --local | grep -v &#39;^\-e&#39; | cut -d = -f 1 | xargs -n1 pip install -U&quot;</span><span class="p">,</span> <span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
	<span class="c1">#pip2</span>
	<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;upgrading pip2...&#39;</span><span class="p">)</span>
	<span class="n">call</span><span class="p">(</span><span class="s1">&#39;pip2 install --upgrade pip2&#39;</span><span class="p">,</span> <span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
	<span class="n">call</span><span class="p">(</span><span class="s2">&quot;pip2 freeze --local | grep -v &#39;^\-e&#39; | cut -d = -f 1 | xargs -n1 pip2 install -U&quot;</span><span class="p">,</span> <span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
	<span class="c1">#pip3</span>
	<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;upgrading pip3...&#39;</span><span class="p">)</span>
	<span class="n">call</span><span class="p">(</span><span class="s1">&#39;pip3 install --upgrade pip3&#39;</span><span class="p">,</span> <span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
	<span class="n">call</span><span class="p">(</span><span class="s2">&quot;pip3 freeze --local | grep -v &#39;^\-e&#39; | cut -d = -f 1 | xargs -n1 pip3 install -U&quot;</span><span class="p">,</span> <span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
	<span class="c1">#pypy</span>
	<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;upgrading pypy-pip...&#39;</span><span class="p">)</span>
	<span class="n">call</span><span class="p">(</span><span class="s1">&#39;pypy -m pip install --upgrade pip&#39;</span><span class="p">,</span><span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
	<span class="n">call</span><span class="p">(</span><span class="s2">&quot;pypy -m pip freeze --local | grep -v &#39;^\-e&#39; | cut -d = -f 1 | xargs -n1 pypy -m pip install -U&quot;</span><span class="p">,</span> <span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></div>

<div class="viewcode-block" id="probability"><a class="viewcode-back" href="../code.html#alefuncs.probability">[docs]</a><span class="k">def</span> <span class="nf">probability</span><span class="p">(</span><span class="n">p</span><span class="p">,</span><span class="n">n</span><span class="p">,</span><span class="n">k</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	Simple probability calculator.</span>
<span class="sd">	Calculates what is the probability that k events occur in n trials.</span>
<span class="sd">	Each event have p probability of occurring once.</span>
<span class="sd">	Example: What is the probability of having 3 Heads by flipping a coin 10 times?</span>
<span class="sd">	probability = prob(0.5,10,3)</span>
<span class="sd">	print(probability) =&gt; (15/128) = 0.1171875</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="kn">from</span> <span class="nn">math</span> <span class="k">import</span> <span class="n">factorial</span>
	<span class="n">p</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>
	<span class="n">n</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">n</span><span class="p">)</span>
	<span class="n">k</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">k</span><span class="p">)</span>
	<span class="n">C</span> <span class="o">=</span> <span class="n">factorial</span><span class="p">(</span><span class="n">n</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span> <span class="n">factorial</span><span class="p">(</span><span class="n">k</span><span class="p">)</span> <span class="o">*</span> <span class="n">factorial</span><span class="p">(</span><span class="n">n</span><span class="o">-</span><span class="n">k</span><span class="p">)</span> <span class="p">)</span>
	<span class="n">probability</span> <span class="o">=</span> <span class="n">C</span> <span class="o">*</span> <span class="p">(</span><span class="n">p</span><span class="o">**</span><span class="n">k</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span><span class="o">-</span><span class="n">p</span><span class="p">)</span><span class="o">**</span><span class="p">(</span><span class="n">n</span><span class="o">-</span><span class="n">k</span><span class="p">)</span>
	<span class="k">return</span> <span class="n">probability</span></div>
<span class="c1">#from math import factorial</span>
<span class="c1">#print(probability(0.5,10,3))</span>
<span class="c1">#print(probability(0.5,1,1))</span>

<div class="viewcode-block" id="process"><a class="viewcode-back" href="../code.html#alefuncs.process">[docs]</a><span class="k">def</span> <span class="nf">process</span><span class="p">(</span><span class="n">real_dataset</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	Generates, annotates and sorts a controll dataset for the given real dataset.</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="n">gen_control_dataset</span><span class="p">(</span><span class="n">real_dataset</span><span class="p">)</span>
	<span class="n">control_filename</span> <span class="o">=</span> <span class="n">real_dataset</span><span class="p">[:</span><span class="o">-</span><span class="mi">4</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39;_control.txt&#39;</span>

	<span class="c1">#annotate_fusion_genes(real_dataset)</span>
	<span class="n">annotate_fusion_genes</span><span class="p">(</span><span class="n">control_filename</span><span class="p">)</span>

	<span class="n">control_filename</span> <span class="o">=</span> <span class="n">control_filename</span><span class="p">[:</span><span class="o">-</span><span class="mi">4</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39;_annotated.txt&#39;</span>
	<span class="c1">#dataset_filename = real_dataset[:-4]+&#39;_annotated.txt&#39;</span>
	
	<span class="c1">#sort_dataset(dataset_filename)</span>
	<span class="n">sort_dataset</span><span class="p">(</span><span class="n">control_filename</span><span class="p">)</span>

	<span class="nb">print</span><span class="p">(</span><span class="n">real_dataset</span><span class="p">,</span><span class="s1">&#39;processed. All OK.&#39;</span><span class="p">)</span></div>
<span class="c1">#process(&#39;/home/amarcozz/Documents/Projects/Fusion Genes/Scripts/clinvarCnv-DeletionsOnly.txt&#39;)</span>
<span class="c1"># folder = &#39;/home/amarcozz/Documents/Projects/Fusion Genes/Scripts/test_datasets/random&#39;</span>
<span class="c1"># for item in list_of_files(folder,&#39;txt&#39;):</span>
<span class="c1">#   process(item)</span>


<div class="viewcode-block" id="query_encode"><a class="viewcode-back" href="../code.html#alefuncs.query_encode">[docs]</a><span class="k">def</span> <span class="nf">query_encode</span><span class="p">(</span><span class="n">chromosome</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	Queries ENCODE via http://promoter.bx.psu.edu/ENCODE/search_human.php</span>
<span class="sd">	Parses the output and returns a dictionary of CIS elements found and the relative location.</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="kn">import</span> <span class="nn">re</span>
	<span class="kn">from</span> <span class="nn">urllib.request</span> <span class="k">import</span> <span class="n">urlopen</span>
	<span class="kn">from</span> <span class="nn">bs4</span> <span class="k">import</span> <span class="n">BeautifulSoup</span>

	<span class="c1">## Regex setup</span>
	<span class="n">re1</span><span class="o">=</span><span class="s1">&#39;(chr</span><span class="si">{}</span><span class="s1">)&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chromosome</span><span class="p">)</span> <span class="c1"># The specific chromosome</span>
	<span class="n">re2</span><span class="o">=</span><span class="s1">&#39;(:)&#39;</span>    <span class="c1"># Any Single Character &#39;:&#39;</span>
	<span class="n">re3</span><span class="o">=</span><span class="s1">&#39;(</span><span class="se">\\</span><span class="s1">d+)&#39;</span> <span class="c1"># Integer</span>
	<span class="n">re4</span><span class="o">=</span><span class="s1">&#39;(-)&#39;</span>    <span class="c1"># Any Single Character &#39;-&#39;</span>
	<span class="n">re5</span><span class="o">=</span><span class="s1">&#39;(</span><span class="se">\\</span><span class="s1">d+)&#39;</span> <span class="c1"># Integer</span>
	<span class="n">rg</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="n">re1</span><span class="o">+</span><span class="n">re2</span><span class="o">+</span><span class="n">re3</span><span class="o">+</span><span class="n">re4</span><span class="o">+</span><span class="n">re5</span><span class="p">,</span><span class="n">re</span><span class="o">.</span><span class="n">IGNORECASE</span><span class="o">|</span><span class="n">re</span><span class="o">.</span><span class="n">DOTALL</span><span class="p">)</span>


	<span class="c1">## Query ENCODE</span>
	<span class="n">std_link</span> <span class="o">=</span> <span class="s1">&#39;http://promoter.bx.psu.edu/ENCODE/get_human_cis_region.php?assembly=hg19&amp;&#39;</span>
	<span class="n">query</span> <span class="o">=</span> <span class="n">std_link</span> <span class="o">+</span> <span class="s1">&#39;chr=chr</span><span class="si">{}</span><span class="s1">&amp;start=</span><span class="si">{}</span><span class="s1">&amp;end=</span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chromosome</span><span class="p">,</span><span class="n">start</span><span class="p">,</span><span class="n">end</span><span class="p">)</span>
	<span class="nb">print</span><span class="p">(</span><span class="n">query</span><span class="p">)</span>
	<span class="n">html_doc</span> <span class="o">=</span> <span class="n">urlopen</span><span class="p">(</span><span class="n">query</span><span class="p">)</span>
	<span class="n">html_txt</span> <span class="o">=</span> <span class="n">BeautifulSoup</span><span class="p">(</span><span class="n">html_doc</span><span class="p">,</span> <span class="s1">&#39;html.parser&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">get_text</span><span class="p">()</span>
	<span class="n">data</span> <span class="o">=</span> <span class="n">html_txt</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>


	<span class="c1">## Parse the output</span>
	<span class="n">parsed</span> <span class="o">=</span> <span class="p">{}</span>
	<span class="n">coordinates</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">item_</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">data</span><span class="p">)</span> <span class="k">if</span> <span class="n">item_</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="o">==</span> <span class="s1">&#39;Coordinate&#39;</span><span class="p">]</span>
	<span class="n">elements</span> <span class="o">=</span> <span class="p">[</span><span class="n">data</span><span class="p">[</span><span class="n">i</span><span class="o">-</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;  &#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;: &#39;</span><span class="p">,</span><span class="s1">&#39;&#39;</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">coordinates</span><span class="p">]</span>
	<span class="n">blocks</span> <span class="o">=</span> <span class="p">[</span><span class="n">item</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">data</span> <span class="k">if</span> <span class="n">item</span><span class="p">[:</span><span class="mi">3</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;chr&#39;</span><span class="p">]</span>
	<span class="c1">#if len(elements) == len(blocks):</span>
	<span class="n">i</span> <span class="o">=</span> <span class="mi">0</span>
	<span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">elements</span><span class="p">:</span>
		<span class="n">txt</span> <span class="o">=</span> <span class="n">blocks</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
		<span class="n">m</span> <span class="o">=</span> <span class="n">rg</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="n">txt</span><span class="p">)</span>
		<span class="n">bins</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">item</span><span class="p">)</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">m</span><span class="p">]</span>
		<span class="n">parsed</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="n">item</span><span class="p">:</span><span class="n">bins</span><span class="p">})</span>
		<span class="n">i</span> <span class="o">+=</span> <span class="mi">1</span>
			
	<span class="k">return</span> <span class="n">parsed</span></div>
<span class="c1">#cis_elements = query_encode(2,10000,20000)</span>


<div class="viewcode-block" id="run_perl"><a class="viewcode-back" href="../code.html#alefuncs.run_perl">[docs]</a><span class="k">def</span> <span class="nf">run_perl</span><span class="p">(</span><span class="n">perl_script_file</span><span class="p">,</span> <span class="n">input_perl_script</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	Run an external perl script and return its output</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="kn">from</span> <span class="nn">subprocess</span> <span class="k">import</span> <span class="n">check_output</span>
	<span class="k">return</span> <span class="n">check_output</span><span class="p">([</span><span class="s2">&quot;perl&quot;</span><span class="p">,</span> <span class="n">perl_script_file</span><span class="p">,</span> <span class="n">input_perl_script</span><span class="p">])</span></div>
<span class="c1">#print(run_perl(&#39;FusionGenes_Annotation.pl&#39;, &#39;test_data.txt&#39;))</span>

<div class="viewcode-block" id="run_py"><a class="viewcode-back" href="../code.html#alefuncs.run_py">[docs]</a><span class="k">def</span> <span class="nf">run_py</span><span class="p">(</span><span class="n">code</span><span class="p">,</span> <span class="n">interp</span><span class="o">=</span><span class="s1">&#39;python3&#39;</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;Run an block of python code using the target interpreter.&#39;&#39;&#39;</span>
	<span class="kn">from</span> <span class="nn">subprocess</span> <span class="k">import</span> <span class="n">check_output</span>
	<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;tmp.py&#39;</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
		<span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">code</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">):</span>
			<span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">line</span><span class="o">+</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
	<span class="k">return</span> <span class="n">check_output</span><span class="p">([</span><span class="n">interpr</span><span class="p">,</span> <span class="s1">&#39;tmp.py&#39;</span><span class="p">])</span></div>


<div class="viewcode-block" id="run_pypy"><a class="viewcode-back" href="../code.html#alefuncs.run_pypy">[docs]</a><span class="k">def</span> <span class="nf">run_pypy</span><span class="p">(</span><span class="n">code</span><span class="p">,</span> <span class="n">interpr</span><span class="o">=</span><span class="s1">&#39;pypy3&#39;</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;Run an block of python code with PyPy&#39;&#39;&#39;</span>
	<span class="kn">from</span> <span class="nn">subprocess</span> <span class="k">import</span> <span class="n">check_output</span>
	<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;tmp.py&#39;</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
		<span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">code</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">):</span>
			<span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">line</span><span class="o">+</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
	<span class="k">return</span> <span class="n">check_output</span><span class="p">([</span><span class="n">interpr</span><span class="p">,</span> <span class="s1">&#39;tmp.py&#39;</span><span class="p">])</span></div>



<div class="viewcode-block" id="sequence_from_coordinates"><a class="viewcode-back" href="../code.html#alefuncs.sequence_from_coordinates">[docs]</a><span class="k">def</span> <span class="nf">sequence_from_coordinates</span><span class="p">(</span><span class="n">chromosome</span><span class="p">,</span><span class="n">strand</span><span class="p">,</span><span class="n">start</span><span class="p">,</span><span class="n">end</span><span class="p">):</span> <span class="c1">#beta hg19 only</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	Download the nucleotide sequence from the gene_name.</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="kn">from</span> <span class="nn">Bio</span> <span class="k">import</span> <span class="n">Entrez</span><span class="p">,</span> <span class="n">SeqIO</span>
	<span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s2">&quot;a.marcozzi@umcutrecht.nl&quot;</span> <span class="c1"># Always tell NCBI who you are</span>
	
	<span class="c1">#GRCh37 from http://www.ncbi.nlm.nih.gov/assembly/GCF_000001405.25/#/def_asm_Primary_Assembly</span>
	<span class="n">NCBI_IDS</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;1&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000001.10&#39;</span><span class="p">,</span><span class="s1">&#39;2&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000002.11&#39;</span><span class="p">,</span><span class="s1">&#39;3&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000003.11&#39;</span><span class="p">,</span><span class="s1">&#39;4&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000004.11&#39;</span><span class="p">,</span>
				<span class="s1">&#39;5&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000005.9&#39;</span><span class="p">,</span><span class="s1">&#39;6&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000006.11&#39;</span><span class="p">,</span><span class="s1">&#39;7&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000007.13&#39;</span><span class="p">,</span><span class="s1">&#39;8&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000008.10&#39;</span><span class="p">,</span>
				<span class="s1">&#39;9&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000009.11&#39;</span><span class="p">,</span><span class="s1">&#39;10&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000010.10&#39;</span><span class="p">,</span><span class="s1">&#39;11&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000011.9&#39;</span><span class="p">,</span><span class="s1">&#39;12&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000012.11&#39;</span><span class="p">,</span>
				<span class="s1">&#39;13&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000013.10&#39;</span><span class="p">,</span><span class="s1">&#39;14&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000014.8&#39;</span><span class="p">,</span><span class="s1">&#39;15&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000015.9&#39;</span><span class="p">,</span><span class="s1">&#39;16&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000016.9&#39;</span><span class="p">,</span>
				<span class="s1">&#39;17&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000017.10&#39;</span><span class="p">,</span><span class="s1">&#39;18&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000018.9&#39;</span><span class="p">,</span><span class="s1">&#39;19&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000019.9&#39;</span><span class="p">,</span><span class="s1">&#39;20&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000020.10&#39;</span><span class="p">,</span>
				<span class="s1">&#39;21&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000021.8&#39;</span><span class="p">,</span><span class="s1">&#39;22&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000022.10&#39;</span><span class="p">,</span><span class="s1">&#39;X&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000023.10&#39;</span><span class="p">,</span><span class="s1">&#39;Y&#39;</span><span class="p">:</span><span class="s1">&#39;NC_000024.9&#39;</span><span class="p">}</span>       
  
	<span class="k">try</span><span class="p">:</span>        
		<span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">efetch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s2">&quot;nucleotide&quot;</span><span class="p">,</span> 
							   <span class="nb">id</span><span class="o">=</span><span class="n">NCBI_IDS</span><span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">chromosome</span><span class="p">)],</span> 
							   <span class="n">rettype</span><span class="o">=</span><span class="s2">&quot;fasta&quot;</span><span class="p">,</span> 
							   <span class="n">strand</span><span class="o">=</span><span class="n">strand</span><span class="p">,</span> <span class="c1">#&quot;1&quot; for the plus strand and &quot;2&quot; for the minus strand.</span>
							   <span class="n">seq_start</span><span class="o">=</span><span class="n">start</span><span class="p">,</span>
							   <span class="n">seq_stop</span><span class="o">=</span><span class="n">end</span><span class="p">)</span>
		<span class="n">record</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
		<span class="n">handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
		<span class="n">sequence</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
		<span class="k">return</span> <span class="n">sequence</span>
	<span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
		<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;ValueError: no sequence found in NCBI&#39;</span><span class="p">)</span>
		<span class="k">return</span> <span class="kc">False</span></div>
<span class="c1">#a = sequence_from_coordinates(9,&#39;-&#39;,21967751,21994490)</span>
<span class="c1">#print(a)</span>

<div class="viewcode-block" id="sequence_from_gene"><a class="viewcode-back" href="../code.html#alefuncs.sequence_from_gene">[docs]</a><span class="k">def</span> <span class="nf">sequence_from_gene</span><span class="p">(</span><span class="n">gene_name</span><span class="p">):</span> <span class="c1">#beta</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	Download the nucleotide sequence from the gene_name.</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="kn">from</span> <span class="nn">Bio</span> <span class="k">import</span> <span class="n">Entrez</span><span class="p">,</span> <span class="n">SeqIO</span>
	<span class="kn">from</span> <span class="nn">pyensembl</span> <span class="k">import</span> <span class="n">EnsemblRelease</span>
	<span class="n">data</span> <span class="o">=</span> <span class="n">EnsemblRelease</span><span class="p">(</span><span class="mi">75</span><span class="p">,</span> <span class="n">auto_download</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
	<span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s2">&quot;a.marcozzi@umcutrecht.nl&quot;</span> <span class="c1"># Always tell NCBI who you are</span>
	<span class="n">NCBI_IDS</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;1&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000001&quot;</span><span class="p">,</span> <span class="s1">&#39;2&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000002&quot;</span><span class="p">,</span><span class="s1">&#39;3&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000003&quot;</span><span class="p">,</span><span class="s1">&#39;4&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000004&quot;</span><span class="p">,</span>
				<span class="s1">&#39;5&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000005&quot;</span><span class="p">,</span><span class="s1">&#39;6&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000006&quot;</span><span class="p">,</span><span class="s1">&#39;7&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000007&quot;</span><span class="p">,</span> <span class="s1">&#39;8&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000008&quot;</span><span class="p">,</span>
				<span class="s1">&#39;9&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000009&quot;</span><span class="p">,</span> <span class="s1">&#39;10&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000010&quot;</span><span class="p">,</span> <span class="s1">&#39;11&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000011&quot;</span><span class="p">,</span> <span class="s1">&#39;12&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000012&quot;</span><span class="p">,</span>
				<span class="s1">&#39;13&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000013&quot;</span><span class="p">,</span><span class="s1">&#39;14&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000014&quot;</span><span class="p">,</span> <span class="s1">&#39;15&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000015&quot;</span><span class="p">,</span> <span class="s1">&#39;16&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000016&quot;</span><span class="p">,</span> 
				<span class="s1">&#39;17&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000017&quot;</span><span class="p">,</span> <span class="s1">&#39;18&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000018&quot;</span><span class="p">,</span> <span class="s1">&#39;19&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000019&quot;</span><span class="p">,</span> <span class="s1">&#39;20&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000020&quot;</span><span class="p">,</span>
				<span class="s1">&#39;21&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000021&quot;</span><span class="p">,</span> <span class="s1">&#39;22&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000022&quot;</span><span class="p">,</span> <span class="s1">&#39;X&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000023&quot;</span><span class="p">,</span> <span class="s1">&#39;Y&#39;</span><span class="p">:</span><span class="s2">&quot;NC_000024&quot;</span><span class="p">}</span>
	
	<span class="n">gene_obj</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">genes_by_name</span><span class="p">(</span><span class="n">gene_name</span><span class="p">)</span>
	<span class="n">target_chromosome</span> <span class="o">=</span> <span class="n">NCBI_IDS</span><span class="p">[</span><span class="n">gene_obj</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">contig</span><span class="p">]</span>
	<span class="n">seq_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">gene_obj</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">start</span><span class="p">)</span>
	<span class="n">seq_stop</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">gene_obj</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">end</span><span class="p">)</span>
	<span class="n">strand</span> <span class="o">=</span> <span class="mi">1</span> <span class="k">if</span> <span class="n">gene_obj</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="s1">&#39;+&#39;</span> <span class="k">else</span> <span class="mi">2</span>
		
	<span class="k">try</span><span class="p">:</span>
			 
		<span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">efetch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s2">&quot;nucleotide&quot;</span><span class="p">,</span> 
							   <span class="nb">id</span><span class="o">=</span><span class="n">target_chromosome</span><span class="p">,</span> 
							   <span class="n">rettype</span><span class="o">=</span><span class="s2">&quot;fasta&quot;</span><span class="p">,</span> 
							   <span class="n">strand</span><span class="o">=</span><span class="n">strand</span><span class="p">,</span> <span class="c1">#&quot;1&quot; for the plus strand and &quot;2&quot; for the minus strand.</span>
							   <span class="n">seq_start</span><span class="o">=</span><span class="n">seq_start</span><span class="p">,</span>
							   <span class="n">seq_stop</span><span class="o">=</span><span class="n">seq_stop</span><span class="p">)</span>
		<span class="n">record</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
		<span class="n">handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
		<span class="n">sequence</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
		<span class="k">return</span> <span class="n">sequence</span>

	<span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
		<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;ValueError: no sequence found in NCBI&#39;</span><span class="p">)</span>
		<span class="k">return</span> <span class="kc">False</span></div>

<div class="viewcode-block" id="sortby_chr"><a class="viewcode-back" href="../code.html#alefuncs.sortby_chr">[docs]</a><span class="k">def</span> <span class="nf">sortby_chr</span><span class="p">(</span><span class="n">string</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	Helps to sort datasets grouped by ChrA/B.</span>
<span class="sd">	To use with sorted().</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="c1"># since the ChrA/B value is a string, when sorting by chr may return [&#39;1&#39;,&#39;10&#39;,&#39;11&#39;...&#39;2&#39;,&#39;20&#39;...&#39;3&#39;...&#39;X&#39;,&#39;Y&#39;]</span>
	<span class="c1"># instead I want sorted() to return [&#39;1&#39;,&#39;2&#39;,...&#39;9&#39;,&#39;10&#39;,&#39;11&#39;...&#39;X&#39;,&#39;Y&#39;]</span>
	<span class="k">if</span> <span class="n">string</span> <span class="o">==</span> <span class="s1">&#39;X&#39;</span><span class="p">:</span>
		<span class="k">return</span> <span class="mi">23</span>
	<span class="k">elif</span> <span class="n">string</span> <span class="o">==</span> <span class="s1">&#39;Y&#39;</span><span class="p">:</span>
		<span class="k">return</span> <span class="mi">24</span>
	<span class="k">else</span><span class="p">:</span>
		<span class="k">return</span> <span class="nb">int</span><span class="p">(</span><span class="n">string</span><span class="p">)</span></div>
<span class="c1"># prova = [&#39;1&#39;,&#39;10&#39;,&#39;11&#39;,&#39;9&#39;,&#39;2&#39;,&#39;20&#39;,&#39;3&#39;,&#39;X&#39;,&#39;Y&#39;]</span>
<span class="c1"># print(&#39;sorted()&#39;, sorted(prova))</span>
<span class="c1"># print(&#39;sortby_chr()&#39;, sorted(prova, key=sortby_chr))</span>

<div class="viewcode-block" id="sort_dataset"><a class="viewcode-back" href="../code.html#alefuncs.sort_dataset">[docs]</a><span class="k">def</span> <span class="nf">sort_dataset</span><span class="p">(</span><span class="n">dataset_file</span><span class="p">,</span> <span class="n">overwrite</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	Sort a dataset by ChrA. It helps during plotting</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="kn">from</span> <span class="nn">operator</span> <span class="k">import</span> <span class="n">itemgetter</span>
	<span class="n">text</span> <span class="o">=</span> <span class="p">[]</span>
	<span class="n">header_counter</span> <span class="o">=</span> <span class="mi">0</span>
	<span class="n">header</span> <span class="o">=</span> <span class="kc">False</span>
	<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Sorting...&#39;</span><span class="p">)</span>
	<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">dataset_file</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">infile</span><span class="p">:</span>
		<span class="c1">#lines = infile.readlines()</span>
		<span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">infile</span><span class="p">:</span>
			<span class="n">list_</span> <span class="o">=</span> <span class="n">line_to_list</span><span class="p">(</span><span class="n">line</span><span class="p">,</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
			<span class="k">if</span> <span class="n">line</span><span class="p">[:</span><span class="mi">2</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;##&#39;</span><span class="p">:</span>
				<span class="n">header</span> <span class="o">=</span> <span class="n">list_</span>
				<span class="n">header_counter</span> <span class="o">+=</span> <span class="mi">1</span>
			<span class="k">else</span><span class="p">:</span>
				<span class="n">text</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">list_</span><span class="p">)</span>
	<span class="c1">#checkpoint</span>
	<span class="k">if</span> <span class="n">header</span> <span class="o">==</span> <span class="kc">False</span> <span class="ow">or</span> <span class="n">header_counter</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
		<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Something is wrong with the header line...&#39;</span><span class="p">,</span> <span class="n">header_counter</span><span class="p">,</span> <span class="n">header</span><span class="p">)</span>
		<span class="k">return</span> <span class="kc">None</span>     
	<span class="c1"># sort by the second element of the list i.e. &#39;ChrA&#39;</span>
	<span class="n">text</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">sortby_chr</span><span class="p">(</span><span class="n">itemgetter</span><span class="p">(</span><span class="mi">1</span><span class="p">)(</span><span class="n">x</span><span class="p">)))</span> 
	<span class="c1"># Write output</span>
	<span class="k">if</span> <span class="n">overwrite</span> <span class="o">==</span> <span class="kc">False</span><span class="p">:</span>
		<span class="n">outfile</span> <span class="o">=</span> <span class="n">dataset_file</span><span class="p">[:</span><span class="o">-</span><span class="mi">4</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39;_sorted.txt&#39;</span>
	<span class="k">else</span><span class="p">:</span>
		<span class="n">outfile</span> <span class="o">=</span> <span class="n">dataset_files</span>
	<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">outfile</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">outfile</span><span class="p">:</span>
		<span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">list_to_line</span><span class="p">(</span><span class="n">header</span><span class="p">,</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
		<span class="k">for</span> <span class="n">list_</span> <span class="ow">in</span> <span class="n">text</span><span class="p">:</span>
			<span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">list_to_line</span><span class="p">(</span><span class="n">list_</span><span class="p">,</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
	<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Done!&#39;</span><span class="p">)</span></div>
<span class="c1"># sort_dataset(&#39;test_data.txt&#39;)</span>
<span class="c1"># folder = &#39;/home/amarcozz/Documents/Projects/Fusion Genes/Scripts/test_datasets/public&#39;</span>
<span class="c1"># for item in list_of_files(folder, &#39;txt&#39;):</span>
<span class="c1">#   sort_dataset(item)</span>
<span class="c1"># sort_dataset(&#39;/home/amarcozz/Documents/Projects/Fusion Genes/Scripts/test_datasets/public/annotated/dgvMerged-DeletionsOnly_annotated.txt&#39;)</span>

<div class="viewcode-block" id="split_fasta_file"><a class="viewcode-back" href="../code.html#alefuncs.split_fasta_file">[docs]</a><span class="k">def</span> <span class="nf">split_fasta_file</span><span class="p">(</span><span class="n">infile</span><span class="p">):</span> <span class="c1">#beta</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	Split a fasta file containing multiple sequences</span>
<span class="sd">	into multiple files containing one sequence each.</span>
<span class="sd">	One sequence per file.</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="n">flag</span> <span class="o">=</span> <span class="kc">False</span>
	<span class="n">length</span> <span class="o">=</span> <span class="mi">0</span>
	<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">infile</span><span class="p">,</span><span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
		<span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">f</span><span class="p">:</span>
			<span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;&gt;&#39;</span><span class="p">):</span>
				<span class="k">if</span> <span class="n">flag</span> <span class="o">==</span> <span class="kc">False</span><span class="p">:</span>
					<span class="n">flag</span> <span class="o">=</span> <span class="kc">True</span>
					<span class="n">outfile</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="si">{}</span><span class="s1">.txt&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">line</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span><span class="o">.</span><span class="n">strip</span><span class="p">())</span>
					<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;writing </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">outfile</span><span class="p">))</span>
					<span class="n">lines</span> <span class="o">=</span> <span class="p">[</span><span class="n">line</span><span class="p">]</span>
				<span class="k">else</span><span class="p">:</span>
					<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">outfile</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">out</span><span class="p">:</span>
						<span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">lines</span><span class="p">:</span>
							<span class="n">out</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">_</span><span class="p">)</span>
						<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1"> bases written&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">length</span><span class="p">))</span>
						<span class="n">length</span> <span class="o">=</span> <span class="mi">0</span>

					<span class="n">outfile</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="si">{}</span><span class="s1">.txt&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">line</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span><span class="o">.</span><span class="n">strip</span><span class="p">())</span>
					<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;writing </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">outfile</span><span class="p">))</span>
					<span class="n">lines</span> <span class="o">=</span> <span class="p">[</span><span class="n">line</span><span class="p">]</span>
			<span class="k">else</span><span class="p">:</span>
				<span class="n">lines</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
				<span class="n">length</span> <span class="o">+=</span> <span class="nb">len</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">())</span>

		<span class="c1">#Write last file        </span>
		<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">outfile</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">out</span><span class="p">:</span>
			<span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">lines</span><span class="p">:</span>
				<span class="n">out</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">_</span><span class="p">)</span>
		<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1"> bases written&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">length</span><span class="p">))</span>    </div>

<div class="viewcode-block" id="substract_datasets"><a class="viewcode-back" href="../code.html#alefuncs.substract_datasets">[docs]</a><span class="k">def</span> <span class="nf">substract_datasets</span><span class="p">(</span><span class="n">infile_1</span><span class="p">,</span> <span class="n">infile_2</span><span class="p">,</span> <span class="n">outfile</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	Takes two files containing tab delimited data, comapares them and return a file</span>
<span class="sd">	containing the data that is present only in infile_2 but not in infile_1.</span>
<span class="sd">	The variable by_column is an int that indicates which column to use</span>
<span class="sd">	as data reference for the comparison.</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="n">header2</span> <span class="o">=</span> <span class="kc">False</span>
	<span class="n">comment_line</span> <span class="o">=</span> <span class="s1">&#39;# dataset generated by substracting &#39;</span> <span class="o">+</span> <span class="n">infile_1</span> <span class="o">+</span> <span class="s1">&#39; to &#39;</span> <span class="o">+</span> <span class="n">infile_2</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span>

	<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">infile_1</span><span class="p">)</span> <span class="k">as</span> <span class="n">infile_1</span><span class="p">:</span>
		<span class="n">lines_1</span> <span class="o">=</span> <span class="n">infile_1</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
	<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">infile_2</span><span class="p">)</span> <span class="k">as</span> <span class="n">infile_2</span><span class="p">:</span>
		<span class="n">lines_2</span> <span class="o">=</span> <span class="n">infile_2</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>

	<span class="n">row_to_removes</span> <span class="o">=</span> <span class="p">[]</span>
	<span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">lines_1</span><span class="p">:</span>
		<span class="k">if</span> <span class="n">line</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;#&#39;</span><span class="p">:</span> <span class="c1"># skips comments</span>
			<span class="k">if</span> <span class="n">header</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
				<span class="n">header2</span> <span class="o">=</span> <span class="kc">True</span> <span class="c1"># to use for the second file</span>
				<span class="n">header</span> <span class="o">=</span> <span class="kc">False</span> <span class="c1"># set back header to false since the first line will be skipped</span>
				<span class="n">first_line</span> <span class="o">=</span> <span class="n">line</span>
				<span class="k">pass</span>
			<span class="k">else</span><span class="p">:</span>
				<span class="n">item</span> <span class="o">=</span> <span class="n">line_to_list</span><span class="p">(</span><span class="n">line</span><span class="p">,</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
				<span class="n">row_to_removes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">item</span><span class="p">)</span>

	<span class="n">result_list</span> <span class="o">=</span> <span class="p">[]</span>
	<span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">lines_2</span><span class="p">:</span>
		<span class="k">if</span> <span class="n">line</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;#&#39;</span><span class="p">:</span> <span class="c1"># skips comments</span>
			<span class="k">if</span> <span class="n">header2</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
				<span class="n">header2</span> <span class="o">=</span> <span class="kc">False</span> <span class="c1"># set back header to false since the first line will be skipped</span>
				<span class="k">pass</span>
			<span class="k">else</span><span class="p">:</span>
				<span class="n">item</span> <span class="o">=</span> <span class="n">line_to_list</span><span class="p">(</span><span class="n">line</span><span class="p">,</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
				<span class="k">if</span> <span class="n">item</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">row_to_removes</span><span class="p">:</span>
					<span class="n">result_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">item</span><span class="p">)</span>
	
	<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">outfile</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">outfile</span><span class="p">:</span>
		<span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">comment_line</span><span class="p">)</span>
		<span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">first_line</span><span class="p">)</span>
		<span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">result_list</span><span class="p">:</span>
			<span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">list_to_line</span><span class="p">(</span><span class="n">item</span><span class="p">,</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
	<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;substraction of two datasets DONE&#39;</span><span class="p">)</span></div>
<span class="c1"># substract_datasets(&#39;dataset_1_b.txt&#39;, &#39;dataset_1.txt&#39;, &#39;dataset_1-1b.txt&#39;, header=True)</span>

<div class="viewcode-block" id="yield_file"><a class="viewcode-back" href="../code.html#alefuncs.yield_file">[docs]</a><span class="k">def</span> <span class="nf">yield_file</span><span class="p">(</span><span class="n">filepath</span><span class="p">):</span>
	<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">	A simple generator that yield the lines of a file.</span>
<span class="sd">	Good to read large file without running out of memory.</span>
<span class="sd">	&#39;&#39;&#39;</span>
	<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
		<span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">f</span><span class="p">:</span>
			<span class="k">yield</span> <span class="n">line</span></div>
<span class="c1"># for line in yield_file(&#39;GRCh37_hg19_variants_2014-10-16.txt&#39;):</span>
<span class="c1">#   print(line[:20])</span>
</pre></div>

          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">alefuncs a documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="index.html" >Module code</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2017, Alessio Marcozzi.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.5.2.
    </div>
  </body>
</html>